// @module-federation/vite/lib/index.mjs
import { createRequire } from "node:module";
import * as fs$1 from "fs";
import fs, { existsSync, mkdirSync, readFileSync, readdirSync, realpathSync, statSync, writeFile, writeFileSync } from "fs";
import { createRequire as createRequire$1 } from "module";
import * as path$1 from "pathe";
import path, { basename, dirname, join, parse, resolve } from "pathe";
import { normalizeOptions } from "@module-federation/sdk";
import { consumeTypesAPI, generateTypesAPI, isTSProject, normalizeConsumeTypesOptions, normalizeDtsOptions, normalizeGenerateTypesOptions } from "@module-federation/dts-plugin";
import { rpc } from "@module-federation/dts-plugin/core";
import { fileURLToPath } from "url";
import { init, parse as parse$1 } from "es-module-lexer";
//#region \0rolldown/runtime.js
var __require = /* @__PURE__ */ createRequire(import.meta.url);
//#endregion
//#region src/utils/codeRewriter.ts
const BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
var CodeRewriter = class {
	replacements = [];
	constructor(original) {
		this.original = original;
	}
	overwrite(start, end, content) {
		if (start < 0 || end < start || end > this.original.length) throw new Error(`Invalid overwrite range: ${start}-${end}`);
		this.replacements.push({
			start,
			end,
			content
		});
	}
	toString() {
		return applyReplacements(this.original, this.getSortedReplacements()).code;
	}
	generateMap(source = "") {
		const { code, replacements } = applyReplacements(this.original, this.getSortedReplacements());
		return {
			version: 3,
			sources: [source],
			sourcesContent: [this.original],
			names: [],
			mappings: generateLineMappings(code, this.original, replacements)
		};
	}
	getSortedReplacements() {
		return [...this.replacements].sort((a, b) => a.start - b.start || a.end - b.end);
	}
};
function createSourceMap(code, source = "") {
	return {
		version: 3,
		sources: [source],
		sourcesContent: [code],
		names: [],
		mappings: generateLineMappings(code, code, [])
	};
}
function applyReplacements(original, replacements) {
	let code = "";
	let cursor = 0;
	let delta = 0;
	const applied = [];
	for (const replacement of replacements) {
		if (replacement.start < cursor) throw new Error("Overlapping overwrite ranges are not supported");
		code += original.slice(cursor, replacement.start);
		const generatedStart = replacement.start + delta;
		code += replacement.content;
		const generatedEnd = generatedStart + replacement.content.length;
		applied.push({
			...replacement,
			generatedStart,
			generatedEnd
		});
		cursor = replacement.end;
		delta += replacement.content.length - (replacement.end - replacement.start);
	}
	code += original.slice(cursor);
	return {
		code,
		replacements: applied
	};
}
function generateLineMappings(generated, original, replacements) {
	const generatedLineStarts = getLineStarts(generated);
	const originalLineStarts = getLineStarts(original);
	let previousOriginalLine = 0;
	let previousOriginalColumn = 0;
	let mappings = "";
	generatedLineStarts.forEach((generatedOffset, lineIndex) => {
		if (lineIndex > 0) mappings += ";";
		const originalOffset = generatedOffsetToOriginalOffset(generatedOffset, replacements);
		const originalLine = findLine(originalLineStarts, originalOffset);
		const originalColumn = originalOffset - originalLineStarts[originalLine];
		mappings += encodeSegment([
			0,
			0,
			originalLine - previousOriginalLine,
			originalColumn - previousOriginalColumn
		]);
		previousOriginalLine = originalLine;
		previousOriginalColumn = originalColumn;
	});
	return mappings;
}
function generatedOffsetToOriginalOffset(offset, replacements) {
	let delta = 0;
	for (const replacement of replacements) {
		if (offset < replacement.generatedStart) break;
		if (offset < replacement.generatedEnd) return replacement.start;
		delta += replacement.content.length - (replacement.end - replacement.start);
	}
	return offset - delta;
}
function getLineStarts(code) {
	const starts = [0];
	for (let i = 0; i < code.length; i++) if (code.charCodeAt(i) === 10) starts.push(i + 1);
	return starts;
}
function findLine(lineStarts, offset) {
	let low = 0;
	let high = lineStarts.length - 1;
	while (low <= high) {
		const mid = low + high >> 1;
		if (lineStarts[mid] <= offset) low = mid + 1;
		else high = mid - 1;
	}
	return Math.max(0, high);
}
function encodeSegment(values) {
	return values.map(encodeVlq).join("");
}
function encodeVlq(value) {
	let vlq = value < 0 ? (-value << 1) + 1 : value << 1;
	let encoded = "";
	do {
		let digit = vlq & 31;
		vlq >>>= 5;
		if (vlq > 0) digit |= 32;
		encoded += BASE64_CHARS[digit];
	} while (vlq > 0);
	return encoded;
}
//#endregion
//#region src/utils/mapCodeToCodeWithSourcemap.ts
async function mapCodeToCodeWithSourcemap(code) {
	const resolvedCode = await code;
	if (resolvedCode === void 0) return;
	return {
		code: resolvedCode,
		map: createSourceMap(resolvedCode)
	};
}
//#endregion
//#region src/utils/htmlEntryUtils.ts
function sanitizeDevEntryPath(devEntryPath) {
	return devEntryPath.replace(/\\\\?/g, "/");
}
/**
* Rewrites entry module script tags to point at an external wrapper module.
* The wrapper can then sequence federation init before the app entry without
* relying on CSP-breaking inline `<script type="module">`.
*/
function rewriteEntryScripts(html, createProxySrc) {
	return html.replace(/<script\b(?=[^>]*\btype=["']module["'])(?=[^>]*\bsrc=["'][^"']+["'])([^>]*)>/gi, (match, attrs) => {
		const srcMatch = attrs.match(/\bsrc=["']([^"']+)["']/i);
		if (!srcMatch) return match;
		const originalSrc = srcMatch[1];
		if (originalSrc.includes("@vite/client")) return match;
		const proxySrc = createProxySrc(originalSrc);
		return match.replace(srcMatch[0], `src=${JSON.stringify(proxySrc)}`);
	});
}
function injectEntryScript(html, initSrc) {
	const src = sanitizeDevEntryPath(initSrc);
	return html.replace("<head>", `<head><script type="module" src=${JSON.stringify(src)}><\/script>`);
}
//#endregion
//#region src/utils/logger.ts
const MODULE_FEDERATION_LOG_PREFIX = "[Module Federation]";
function formatModuleFederationMessage(message) {
	return `${MODULE_FEDERATION_LOG_PREFIX} ${message}`;
}
function createModuleFederationError(message) {
	return new Error(formatModuleFederationMessage(message));
}
function toConsoleArgs(message, rest = []) {
	if (typeof message === "string") return [formatModuleFederationMessage(message), ...rest];
	if (message === void 0) return [MODULE_FEDERATION_LOG_PREFIX, ...rest];
	return [
		MODULE_FEDERATION_LOG_PREFIX,
		message,
		...rest
	];
}
const moduleFederationConsole = {
	log(message, ...rest) {
		console.log(...toConsoleArgs(message, rest));
	},
	warn(message, ...rest) {
		console.warn(...toConsoleArgs(message, rest));
	},
	error(message, ...rest) {
		console.error(...toConsoleArgs(message, rest));
	}
};
moduleFederationConsole.log;
const mfWarn = moduleFederationConsole.warn;
const mfError = moduleFederationConsole.error;
//#endregion
//#region src/utils/packageUtils.ts
const dependencyPresenceCache = /* @__PURE__ */ new Map();
let packageDetectionCwd;
function getDependencyCacheKey(cwd, dependencyName) {
	return `${cwd}:${dependencyName}`;
}
function setPackageDetectionCwd(cwd) {
	packageDetectionCwd = cwd;
}
function getPackageDetectionCwd() {
	return packageDetectionCwd || process.cwd();
}
const DEFAULT_EXPORT_CONDITIONS = [
	"browser",
	"import",
	"module",
	"default",
	"require"
];
function resolveExportsEntry(exportsField, conditions = DEFAULT_EXPORT_CONDITIONS) {
	if (typeof exportsField === "string") return exportsField;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const rootExport = record["."];
	if (rootExport) return resolveExportsEntry(rootExport);
	for (const condition of conditions) {
		const target = resolveExportsEntry(record[condition], conditions);
		if (target) return target;
	}
	for (const target of Object.values(record)) {
		const resolved = resolveExportsEntry(target, conditions);
		if (resolved) return resolved;
	}
}
function getPackageExportsTarget(pkg, packageName, exportsField) {
	if (typeof exportsField === "string") return pkg === packageName ? exportsField : void 0;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const subpath = pkg === packageName ? "." : `.${pkg.slice(packageName.length)}`;
	if (subpath !== ".") return record[subpath];
	return record["."] ?? (!Object.keys(record).some((key) => key.startsWith(".")) ? record : void 0);
}
/**
* Escaping rules:
* Convert using the format __${mapping}__, where _ and $ are not allowed in npm package names but can be used in variable names.
*  @ => 1
*  / => 2
*  - => 3
*  . => 4
*/
/**
* Encodes a package name into a valid file name.
* @param {string} name - The package name, e.g., "@scope/xx-xx.xx".
* @returns {string} - The encoded file name.
*/
function packageNameEncode(name) {
	if (typeof name !== "string") throw createModuleFederationError("A string package name is required");
	return name.replace(/@/g, "_mf_0_").replace(/\//g, "_mf_1_").replace(/-/g, "_mf_2_").replace(/\./g, "_mf_3_");
}
/**
* Decodes an encoded file name back to the original package name.
* @param {string} encoded - The encoded file name, e.g., "_mf_0_scope_mf_1_xx_mf_2_xx_mf_3_xx".
* @returns {string} - The decoded package name.
*/
function packageNameDecode(encoded) {
	if (typeof encoded !== "string") throw createModuleFederationError("A string encoded file name is required");
	return encoded.replace(/_mf_0_/g, "@").replace(/_mf_1_/g, "/").replace(/_mf_2_/g, "-").replace(/_mf_3_/g, ".");
}
/**
* Removes any subpath from an npm package specifier and returns the package name only.
* @param {string} packageString - The package specifier, e.g., "@scope/pkg/runtime" or "react/jsx-runtime".
* @returns {string} - The base npm package name.
*/
function getPackageName(packageString) {
	const match = packageString.match(/^(?:@[^/]+\/)?[^/]+/);
	return match ? match[0] : packageString;
}
function getPackageNameFromNodeModulePath(source) {
	const normalized = source.replace(/\\/g, "/");
	const nodeModulesIndex = normalized.lastIndexOf("/node_modules/");
	if (nodeModulesIndex < 0) return;
	const parts = normalized.slice(nodeModulesIndex + 14).split("/");
	if (!parts[0]) return;
	if (parts[0].startsWith("@")) return parts[1] ? `${parts[0]}/${parts[1]}` : void 0;
	return parts[0];
}
function getInstalledPackageJson(pkg, opts) {
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	const tryReadPackageJson = (packageJsonPath) => {
		if (!existsSync(packageJsonPath)) return void 0;
		try {
			return {
				path: packageJsonPath,
				dir: path.dirname(packageJsonPath),
				packageJson: JSON.parse(readFileSync(packageJsonPath, "utf-8"))
			};
		} catch {
			return;
		}
	};
	const findPackageInPnpmStore = (startDir) => {
		let currentDir = startDir;
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const pnpmStoreDir = path.join(currentDir, "node_modules", ".pnpm");
			if (existsSync(pnpmStoreDir)) try {
				for (const entry of readdirSync(pnpmStoreDir, { withFileTypes: true })) {
					if (!entry.isDirectory()) continue;
					const candidate = tryReadPackageJson(path.join(pnpmStoreDir, entry.name, "node_modules", packageName, "package.json"));
					if (candidate?.packageJson.name === packageName) return candidate;
				}
			} catch {}
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
	};
	try {
		const projectRequire = createRequire$1(new URL(`file://${path.join(cwd, "package.json")}`));
		let resolvedPath;
		if (opts?.fromResolvedEntry) resolvedPath = opts.fromResolvedEntry;
		else try {
			resolvedPath = projectRequire.resolve(pkg);
		} catch {
			resolvedPath = projectRequire.resolve(packageName);
		}
		let currentDir = path.dirname(resolvedPath);
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const packageJsonPath = path.join(currentDir, "package.json");
			if (existsSync(packageJsonPath)) {
				const packageJsonContent = readFileSync(packageJsonPath, "utf-8");
				try {
					const packageJson = JSON.parse(packageJsonContent);
					if (packageJson.name === packageName) return {
						path: packageJsonPath,
						dir: currentDir,
						packageJson
					};
				} catch (error) {
					if (!(error instanceof SyntaxError)) throw error;
				}
			}
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
	} catch {
		let currentDir = cwd;
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const directCandidate = tryReadPackageJson(path.join(currentDir, "node_modules", packageName, "package.json"));
			if (directCandidate?.packageJson.name === packageName) return directCandidate;
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
		return findPackageInPnpmStore(cwd);
	}
}
function getInstalledPackageEntry(pkg, opts) {
	const installed = getInstalledPackageJson(pkg, opts);
	if (!installed) return void 0;
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	if (pkg !== packageName && opts?.resolveSubpathWithRequire !== false) try {
		return createRequire$1(new URL(`file://${path.join(cwd, "package.json")}`)).resolve(pkg);
	} catch {}
	const packageJson = installed.packageJson;
	const explicitEntry = resolveExportsEntry(getPackageExportsTarget(pkg, packageName, packageJson.exports), opts?.conditions) || (typeof packageJson.module === "string" ? packageJson.module : void 0) || (typeof packageJson.main === "string" ? packageJson.main : void 0) || "index.js";
	return path.join(installed.dir, explicitEntry);
}
/**
* Detect whether the current runtime is Vite 8+ by checking for a Vite version flag
* on the plugin hook context, with Rolldown metadata kept as a compatibility fallback.
*/
function getIsRolldown(ctx) {
	const viteVersion = ctx?.meta?.viteVersion;
	const viteMajor = Number(String(viteVersion ?? "").split(".")[0]);
	return Number.isFinite(viteMajor) && viteMajor >= 8 || !!ctx?.meta?.rolldownVersion;
}
function hasPackageDependency(dependencyName, cwd = packageDetectionCwd || process.cwd()) {
	const cacheKey = getDependencyCacheKey(cwd, dependencyName);
	const cached = dependencyPresenceCache.get(cacheKey);
	if (cached !== void 0) return cached;
	try {
		const packageJson = JSON.parse(readFileSync(path.join(cwd, "package.json"), "utf8"));
		const hasDependency = [
			packageJson.dependencies,
			packageJson.devDependencies,
			packageJson.peerDependencies,
			packageJson.optionalDependencies
		].some((deps) => !!deps?.[dependencyName]);
		dependencyPresenceCache.set(cacheKey, hasDependency);
		return hasDependency;
	} catch {
		dependencyPresenceCache.set(cacheKey, false);
		return false;
	}
}
//#endregion
//#region src/utils/normalizeModuleFederationOptions.ts
const INTERNAL_NAME_PREFIX = "__mfe_internal__";
function toInternalModuleFederationName(name) {
	return name.startsWith(INTERNAL_NAME_PREFIX) ? name : `${INTERNAL_NAME_PREFIX}${name}`;
}
function warnOnReservedInternalNamePrefix(name, kind) {
	if (!name.startsWith(INTERNAL_NAME_PREFIX)) return;
	mfWarn(`Reserved internal ${kind} prefix "${INTERNAL_NAME_PREFIX}" detected in public ${kind} "${name}". This prefix is reserved for internal module federation names and may cause conflicts.`);
}
function normalizeExposesItem(item) {
	let importPath = "";
	if (typeof item === "string") importPath = item;
	if (typeof item === "object") importPath = item.import;
	return { import: importPath };
}
function normalizeExposes(exposes) {
	if (!exposes) return {};
	const res = {};
	Object.keys(exposes).forEach((key) => {
		res[key] = normalizeExposesItem(exposes[key]);
	});
	return res;
}
function normalizeRemotes(remotes) {
	if (!remotes) return {};
	const result = {};
	if (typeof remotes === "object") Object.keys(remotes).forEach((key) => {
		result[key] = normalizeRemoteItem(key, remotes[key]);
	});
	return result;
}
function normalizeRemoteItem(key, remote) {
	warnOnReservedInternalNamePrefix(key, "remoteAlias");
	if (typeof remote === "string") {
		const separatorIndex = remote.startsWith("@") ? remote.indexOf("@", 1) : remote.indexOf("@");
		let entryGlobalName;
		let entry;
		if (separatorIndex > 0) {
			entryGlobalName = remote.slice(0, separatorIndex);
			entry = remote.slice(separatorIndex + 1);
		} else {
			entryGlobalName = remote;
			entry = remote;
		}
		return {
			type: "var",
			name: key,
			internalName: toInternalModuleFederationName(key),
			entry,
			entryGlobalName,
			shareScope: "default"
		};
	}
	return Object.assign({
		type: "var",
		name: key,
		internalName: toInternalModuleFederationName(key),
		shareScope: "default",
		entryGlobalName: key
	}, {
		...remote,
		internalName: toInternalModuleFederationName(remote.name || key)
	});
}
/**
* Tries to find the package.json's version of a shared package
* if `package.json` is not declared in `exports`
* @param {string} sharedName
* @returns {string | undefined}
*/
function searchPackageVersion(sharedName) {
	const version = getInstalledPackageJson(sharedName, { packageName: sharedName })?.packageJson.version;
	return typeof version === "string" ? version : void 0;
}
function inferVersionFromRequiredVersion(requiredVersion) {
	if (!requiredVersion) return void 0;
	return requiredVersion.match(/\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?/)?.[0];
}
function getLitExportSubpathShares(sharedName) {
	if (sharedName !== "lit") return [];
	const exportsField = getInstalledPackageJson(sharedName, { packageName: sharedName })?.packageJson.exports;
	if (!exportsField || typeof exportsField === "string") return [];
	return Object.keys(exportsField).filter((key) => key.startsWith("./") && key !== "." && !key.includes("*")).map((key) => `${sharedName}/${key.slice(2)}`);
}
function normalizeShareItem(key, shareItem) {
	let version;
	const isImportFalse = typeof shareItem === "object" && shareItem.import === false;
	try {
		try {
			version = __require(path$1.join(getPackageName(key), "package.json")).version;
		} catch (e1) {
			try {
				version = __require(path$1.join(process.cwd(), "node_modules", getPackageName(key), "package.json")).version;
			} catch (e2) {
				version = searchPackageVersion(key);
				if (!version && !isImportFalse) mfError(e1);
			}
		}
	} catch (e) {
		if (!isImportFalse) mfError(`Unexpected error resolving version for ${key}:`, e);
	}
	if (typeof shareItem === "string") return {
		name: shareItem,
		version,
		scope: "default",
		from: "",
		shareConfig: {
			import: void 0,
			singleton: false,
			requiredVersion: version ? `^${version}` : "*"
		}
	};
	return {
		name: key,
		from: "",
		version: shareItem.version || inferVersionFromRequiredVersion(shareItem.requiredVersion) || version,
		scope: shareItem.shareScope || "default",
		shareConfig: {
			import: shareItem.import,
			singleton: shareItem.singleton || false,
			requiredVersion: shareItem.requiredVersion || (version ? `^${version}` : "*"),
			strictVersion: !!shareItem.strictVersion
		}
	};
}
function normalizeShared(shared) {
	explicitSharedKeys = /* @__PURE__ */ new Set();
	if (!shared) {
		const result = {};
		const packageJsonPath = path$1.join(getPackageDetectionCwd(), "package.json");
		try {
			const packageJson = JSON.parse(fs$1.readFileSync(packageJsonPath, "utf-8"));
			if (packageJson.name === "@module-federation/vite") return result;
			Object.keys(packageJson.dependencies || {}).filter((key) => shouldAutoShareDependency(key)).forEach((key) => {
				result[key] = normalizeShareItem(key, {
					name: key,
					import: void 0,
					singleton: true
				});
			});
		} catch {
			return result;
		}
		return result;
	}
	const result = {};
	const sourceEntries = [];
	if (Array.isArray(shared)) shared.forEach((key) => {
		result[key] = normalizeShareItem(key, key);
		explicitSharedKeys.add(key);
		sourceEntries.push([key, key]);
	});
	else if (typeof shared === "object") Object.keys(shared).forEach((key) => {
		const value = shared[key];
		result[key] = normalizeShareItem(key, value);
		explicitSharedKeys.add(key);
		sourceEntries.push([key, value]);
	});
	sourceEntries.forEach(([key, value]) => {
		for (const subpathShare of getLitExportSubpathShares(key)) {
			if (result[subpathShare]) continue;
			result[subpathShare] = normalizeShareItem(subpathShare, value);
		}
	});
	return result;
}
function shouldAutoShareDependency(key) {
	const pkg = getInstalledPackageJson(key)?.packageJson;
	if (!pkg) return false;
	if (!pkg.browser && !pkg.exports && typeof pkg.module !== "string") return false;
	if (key === "@module-federation/vite") return false;
	const entry = [
		pkg.module,
		pkg.main,
		typeof pkg.exports === "string" ? pkg.exports : void 0
	].find((value) => typeof value === "string")?.replace(/\\/g, "/");
	if (key === "nuxt" || key === "nuxt-nightly") return false;
	if (key.includes("nuxt") && entry?.endsWith("/dist/module.mjs")) return false;
	return true;
}
function normalizeLibrary(library) {
	if (!library) return void 0;
	return library;
}
function normalizeManifest(manifest) {
	if (manifest === void 0) return;
	if (typeof manifest === "boolean") return manifest;
	return {
		...manifest,
		fileName: manifest.fileName || "mf-manifest.json"
	};
}
let config;
let explicitSharedKeys = /* @__PURE__ */ new Set();
function getNormalizeModuleFederationOptions() {
	return config;
}
function isExplicitSharedKey(key) {
	return explicitSharedKeys.has(key);
}
function getNormalizeShareItem(key) {
	const options = getNormalizeModuleFederationOptions();
	return options.shared[key] || options.shared[getPackageName(key)] || options.shared[getPackageName(key) + "/"];
}
function normalizeModuleFederationOptions(options) {
	warnOnReservedInternalNamePrefix(options.name, "containerName");
	if (options.virtualModuleDir && options.virtualModuleDir.includes("/")) throw createModuleFederationError(`Invalid virtualModuleDir: "${options.virtualModuleDir}". The virtualModuleDir option cannot contain slashes (/). Please use a single directory name like '__mf__virtual__your_app_name'.`);
	return config = {
		exposes: normalizeExposes(options.exposes),
		filename: options.filename || "remoteEntry-[hash]",
		internalName: toInternalModuleFederationName(options.name),
		library: normalizeLibrary(options.library),
		name: options.name,
		remotes: normalizeRemotes(options.remotes),
		runtime: options.runtime,
		shareScope: options.shareScope || "default",
		shared: normalizeShared(options.shared),
		runtimePlugins: options.runtimePlugins || [],
		implementation: options.implementation || __require.resolve("@module-federation/runtime"),
		manifest: normalizeManifest(options.manifest),
		dev: options.dev,
		dts: options.dts,
		getPublicPath: options.getPublicPath,
		publicPath: options.publicPath,
		shareStrategy: options.shareStrategy || "version-first",
		ignoreOrigin: options.ignoreOrigin || false,
		virtualModuleDir: options.virtualModuleDir || "__mf__virtual",
		hostInitInjectLocation: options.hostInitInjectLocation || "html",
		bundleAllCSS: options.bundleAllCSS || false,
		moduleParseTimeout: options.moduleParseTimeout || 10,
		moduleParseIdleTimeout: options.moduleParseIdleTimeout,
		varFilename: options.varFilename,
		target: options.target
	};
}
//#endregion
//#region src/utils/VirtualModule.ts
/**
* Initialize virtual module infrastructure BEFORE VirtualModule class is used.
* This must be called in the config hook to ensure the directory exists
* before Vite's optimization phase.
*/
function initVirtualModuleInfrastructure(root, virtualModuleDir = "__mf__virtual") {
	const virtualPackagePath = join(join(root, "node_modules"), virtualModuleDir);
	mkdirSync(virtualPackagePath, { recursive: true });
	writeFileSync(join(virtualPackagePath, "empty.js"), "");
	writeFileSync(join(virtualPackagePath, "package.json"), JSON.stringify({
		name: virtualModuleDir,
		main: "empty.js"
	}));
}
let rootDir;
function findNodeModulesDir(root = process.cwd()) {
	let currentDir = root;
	while (currentDir !== parse(currentDir).root) {
		const nodeModulesPath = join(currentDir, "node_modules");
		if (existsSync(nodeModulesPath)) return nodeModulesPath;
		currentDir = dirname(currentDir);
	}
	return "";
}
let cachedNodeModulesDir;
function getNodeModulesDir() {
	if (!cachedNodeModulesDir) cachedNodeModulesDir = findNodeModulesDir(rootDir);
	return cachedNodeModulesDir;
}
function getSuffix(name) {
	const base = basename(name);
	const dotIndex = base.lastIndexOf(".");
	if (dotIndex > 0 && dotIndex < base.length - 1) return base.slice(dotIndex);
	return ".js";
}
const patternMap = {};
const cacheMap = {};
/**
* Physically generate files as virtual modules under node_modules/__mf__virtual/*
*/
function assertModuleFound(tag, str = "") {
	const module = VirtualModule.findModule(tag, str);
	if (!module) throw createModuleFederationError(`Module Federation shared module '${str}' not found. Please ensure it's installed as a dependency in your package.json.`);
	return module;
}
var VirtualModule = class {
	name;
	tag;
	suffix;
	inited = false;
	/**
	* Set the root path for finding node_modules
	* @param root - Root path
	*/
	static setRoot(root) {
		rootDir = root;
		cachedNodeModulesDir = void 0;
	}
	/**
	* Ensure virtual package directory exists
	*/
	static ensureVirtualPackageExists() {
		const nodeModulesDir = getNodeModulesDir();
		const { virtualModuleDir } = getNormalizeModuleFederationOptions();
		const virtualPackagePath = resolve(nodeModulesDir, virtualModuleDir);
		mkdirSync(virtualPackagePath, { recursive: true });
		writeFileSync(resolve(virtualPackagePath, "empty.js"), "");
		writeFileSync(resolve(virtualPackagePath, "package.json"), JSON.stringify({
			name: virtualModuleDir,
			main: "empty.js"
		}));
	}
	static findModule(tag, str = "") {
		if (!patternMap[tag]) patternMap[tag] = new RegExp(`(.*${packageNameEncode(tag)}(.+?)${packageNameEncode(tag)}.*)`);
		const moduleName = (str.match(patternMap[tag]) || [])[2];
		if (moduleName) return cacheMap[tag][packageNameDecode(moduleName)];
	}
	constructor(name, tag = "__mf_v__", suffix = "") {
		this.name = name;
		this.tag = tag;
		this.suffix = suffix || getSuffix(name);
		if (!cacheMap[this.tag]) cacheMap[this.tag] = {};
		cacheMap[this.tag][this.name] = this;
	}
	getPath() {
		return resolve(getNodeModulesDir(), this.getImportId());
	}
	getImportId() {
		const { internalName: mfName, virtualModuleDir } = getNormalizeModuleFederationOptions();
		return `${virtualModuleDir}/${packageNameEncode(`${mfName}${this.tag}${this.name}${this.tag}`)}${this.suffix}`;
	}
	writeSync(code, force) {
		if (!force && this.inited) return;
		if (!this.inited) this.inited = true;
		const path = this.getPath();
		mkdirSync(dirname(path), { recursive: true });
		writeFileSync(path, code);
	}
	write(code) {
		const path = this.getPath();
		mkdirSync(dirname(path), { recursive: true });
		writeFile(path, code, function() {});
	}
};
//#endregion
//#region src/utils/serializeRuntimeOptions.ts
/**
* Serializes a JavaScript object into a string of source code that can be evaluated.
* This function is used to create runtime plugin options without relying solely on JSON.stringify,
* allowing support for non-JSON types like RegExp, Date, Map, Set, and Functions.
* It also safely handles circular references.
*
* @param {Record<string, unknown>} options - The options object to serialize.
* @returns {string} The resulting JavaScript source code string.
*/
function serializeRuntimeOptions(options) {
	const seenObjects = /* @__PURE__ */ new WeakSet();
	/**
	* Recursive inner function to serialize any value into a source code string.
	*/
	function valueToCode(val) {
		if (val === null) return "null";
		const type = typeof val;
		if (type === "string") return JSON.stringify(val);
		if (type === "number" || type === "boolean") return String(val);
		if (type === "undefined") return "undefined";
		if (type === "symbol") {
			const desc = val.description ?? "";
			return `Symbol(${JSON.stringify(desc)})`;
		}
		if (type === "function") return val.toString();
		if (val instanceof Date) return `new Date(${JSON.stringify(val.toISOString())})`;
		if (val instanceof RegExp) return `new RegExp(${JSON.stringify(val.source)}, ${JSON.stringify(val.flags)})`;
		if (type === "object") {
			if (seenObjects.has(val)) return `"__circular__"`;
			seenObjects.add(val);
		}
		if (Array.isArray(val)) return `[${val.map(valueToCode).join(", ")}]`;
		if (val instanceof Map) return `new Map([${Array.from(val.entries()).map(([k, v]) => `[${valueToCode(k)}, ${valueToCode(v)}]`).join(", ")}])`;
		if (val instanceof Set) return `new Set([${Array.from(val.values()).map(valueToCode).join(", ")}])`;
		if (type === "object") {
			const properties = [];
			for (const key in val) if (Object.prototype.hasOwnProperty.call(val, key)) properties.push(`${JSON.stringify(key)}: ${valueToCode(val[key])}`);
			return `{${properties.join(", ")}}`;
		}
		return JSON.stringify(String(val));
	}
	const topLevelProps = [];
	for (const key in options) if (Object.prototype.hasOwnProperty.call(options, key)) topLevelProps.push(`${JSON.stringify(key)}: ${valueToCode(options[key])}`);
	return `{${topLevelProps.join(", ")}}`;
}
//#endregion
//#region src/virtualModules/virtualExposes.ts
const EXPOSES_CSS_MAP_PLACEHOLDER = "__MF_EXPOSES_CSS_MAP__";
function getExposesCssMapPlaceholder() {
	return EXPOSES_CSS_MAP_PLACEHOLDER;
}
function getVirtualExposesId(options) {
	return `virtual:mf-exposes:${`${options.internalName}__${options.filename}`.replace(/[^a-zA-Z0-9_-]/g, "_")}`;
}
function generateExposes(options) {
	return `
    const cssAssetMap = ${JSON.stringify(options.bundleAllCSS ? EXPOSES_CSS_MAP_PLACEHOLDER : {})};
    const injectedCssHrefs = new Set();
    let exposeLoadQueue = Promise.resolve();

    async function importExposedModule(loader) {
      const currentLoad = exposeLoadQueue.then(loader, loader);
      exposeLoadQueue = currentLoad.then(
        () => undefined,
        () => undefined
      );
      return currentLoad;
    }

    async function injectCssAssets(exposeKey) {
      if (typeof document === "undefined") {
        return;
      }

      // Replaced at build time with expose -> css asset paths.
      const cssAssets = cssAssetMap[exposeKey] || [];

      await Promise.all(
        cssAssets.map((cssAsset) => {
          const href = new URL(cssAsset, import.meta.url).href;

          // Same expose can be resolved multiple times in one page.
          if (injectedCssHrefs.has(href)) {
            return Promise.resolve();
          }
          injectedCssHrefs.add(href);

          const existingLink = document.querySelector(
            \`link[rel="stylesheet"][data-mf-href="\${href}"]\`
          );
          if (existingLink) {
            return Promise.resolve();
          }

          return new Promise((resolve, reject) => {
            const link = document.createElement("link");
            link.rel = "stylesheet";
            link.href = href;
            link.setAttribute("data-mf-href", href);
            link.onload = () => resolve();
            link.onerror = () => reject(new Error(\`[Module Federation] Failed to load CSS asset: \${href}\`));
            document.head.appendChild(link);
          });
        })
      );
    }

    export default {
    ${Object.keys(options.exposes).map((key) => {
		return `
        ${JSON.stringify(key)}: async () => {
          await injectCssAssets(${JSON.stringify(key)})
          const importModule = await importExposedModule(
            () => import(${JSON.stringify(options.exposes[key].import)})
          )
          const exportModule = {}
          Object.assign(exportModule, importModule)
          Object.defineProperty(exportModule, "__esModule", {
            value: true,
            enumerable: false
          })
          return exportModule
        }
      `;
	}).join(",")}
  }
  `;
}
//#endregion
//#region src/virtualModules/virtualRuntimeInitStatus.ts
const virtualRuntimeInitStatus = new VirtualModule("runtimeInit");
const MODULE_CACHE_GLOBAL_KEY = "__mf_module_cache__";
function getRuntimeInitGlobalKey() {
	return `__mf_init__${virtualRuntimeInitStatus.getImportId()}__`;
}
function getDeferredInitPromiseCode() {
	return `let initResolve, initReject;
  const initPromise = new Promise((re, rj) => {
    initResolve = re;
    initReject = rj;
  });`;
}
function getSsrNoopResolveCode() {
	return `if (typeof window === 'undefined') {
    initResolve({
      loadRemote: function() { return Promise.resolve(undefined); },
      loadShare: function() { return Promise.resolve(undefined); },
    });
  }`;
}
function getRuntimeInitStateBootstrapCode(options) {
	return `
const ${options.globalKeyVar} = ${JSON.stringify(getRuntimeInitGlobalKey())};
let ${options.stateVar} = globalThis[${options.globalKeyVar}];
if (!${options.stateVar}) {
  ${getDeferredInitPromiseCode()}
  ${options.stateVar} = globalThis[${options.globalKeyVar}] = {
    initPromise,
    initResolve,
    initReject,
  };
  ${getSsrNoopResolveCode()}
}
const ${options.exposedConst} = ${options.stateVar}.${options.exposedProperty};
`;
}
function getRuntimeInitBootstrapCode() {
	return `
const globalKey = ${JSON.stringify(getRuntimeInitGlobalKey())};
const moduleCacheGlobalKey = ${JSON.stringify(MODULE_CACHE_GLOBAL_KEY)};
globalThis[moduleCacheGlobalKey] ||= { share: {}, remote: {} };
globalThis[moduleCacheGlobalKey].share ||= {};
globalThis[moduleCacheGlobalKey].remote ||= {};
if (!globalThis[globalKey]) {
  ${getDeferredInitPromiseCode()}
  globalThis[globalKey] = {
    initPromise,
    initResolve,
    initReject,
    moduleCache: globalThis[moduleCacheGlobalKey],
  };
  ${getSsrNoopResolveCode()}
}
globalThis[globalKey].moduleCache ||= globalThis[moduleCacheGlobalKey];
globalThis[globalKey].moduleCache.share ||= {};
globalThis[globalKey].moduleCache.remote ||= {};
`;
}
function getRuntimeModuleCacheBootstrapCode() {
	return `
const __mfCacheGlobalKey = ${JSON.stringify(MODULE_CACHE_GLOBAL_KEY)};
globalThis[__mfCacheGlobalKey] ||= { share: {}, remote: {} };
globalThis[__mfCacheGlobalKey].share ||= {};
globalThis[__mfCacheGlobalKey].remote ||= {};
const __mfModuleCache = globalThis[__mfCacheGlobalKey];
`;
}
function getRuntimeInitResolveBootstrapCode() {
	return getRuntimeInitStateBootstrapCode({
		globalKeyVar: "__mfResolveGlobalKey",
		stateVar: "__mfResolveState",
		exposedConst: "initResolve",
		exposedProperty: "initResolve"
	});
}
function writeRuntimeInitStatus(command) {
	const exportStatement = command === "build" ? `const { initPromise, initResolve, initReject, moduleCache } = globalThis[globalKey];
export { initPromise, initResolve, initReject, moduleCache };` : `module.exports = globalThis[globalKey];`;
	virtualRuntimeInitStatus.writeSync(`
${getRuntimeInitBootstrapCode()}
${exportStatement}
`);
}
//#endregion
//#region src/virtualModules/virtualShared_preBuild.ts
/**
* Even the resolveId hook cannot interfere with vite pre-build,
* and adding query parameter virtual modules will also fail.
* You can only proxy to the real file through alias
*/
/**
* shared will be proxied:
* 1. __prebuild__: export shareModule (pre-built source code of modules such as vue, react, etc.)
* 2. __loadShare__: load shareModule (mfRuntime.loadShare('vue'))
*/
const JS_IDENTIFIER_REGEX = /* @__PURE__ */ new RegExp("^[$_\\p{ID_Start}][$_\\u200C\\u200D\\p{ID_Continue}]*$", "u");
function escapeGeneratedStringLiteral(value) {
	return JSON.stringify(value).replace(/[<>\u2028\u2029]/g, (char) => {
		switch (char) {
			case "<": return "\\u003C";
			case ">": return "\\u003E";
			case "\u2028": return "\\u2028";
			case "\u2029": return "\\u2029";
			default: return char;
		}
	});
}
function isValidJsIdentifier(name) {
	return JS_IDENTIFIER_REGEX.test(name);
}
function isValidEsmExportName(name) {
	return !!name && name !== "default" && name !== "__esModule" && isValidJsIdentifier(name);
}
const JS_IDENTIFIER_PATTERN = `[\$_\\p{ID_Start}][\$_\\u200C\\u200D\\p{ID_Continue}]*`;
const localRequire = createRequire$1(import.meta.url);
function resolvePackageEntryFromProjectRoot(pkg) {
	try {
		return createRequire$1(new URL(`file://${path.join(getPackageDetectionCwd(), "package.json")}`)).resolve(pkg);
	} catch {
		return;
	}
}
function getPackageEsmEntryPath(pkg) {
	return getInstalledPackageEntry(pkg, {
		conditions: [
			"browser",
			"import",
			"module",
			"default"
		],
		resolveSubpathWithRequire: false
	}) || resolvePackageEntryFromProjectRoot(pkg);
}
function getEsmNamedExports(pkg) {
	let source = "";
	let entryPath;
	try {
		entryPath = getPackageEsmEntryPath(pkg);
		if (!entryPath) return [];
		const { initSync, parse } = localRequire("es-module-lexer");
		initSync();
		source = readFileSync(entryPath, "utf-8");
		const [, exports] = parse(source, entryPath);
		const names = exports.map((item) => item.n).filter((name) => isValidEsmExportName(name));
		const regexNames = getNamedExportsViaRegex(source, entryPath);
		const filteredNames = names.filter((name) => name !== "type" || regexNames.includes(name));
		if (filteredNames.length > 0) return Array.from(new Set([...filteredNames, ...regexNames]));
		return regexNames;
	} catch {
		return source ? getNamedExportsViaRegex(source, entryPath) : [];
	}
}
function resolveRelativeModule(filePath, specifier) {
	const dir = path.dirname(filePath);
	const exact = path.resolve(dir, specifier);
	if (existsSync(exact) && !statSync(exact).isDirectory()) return exact;
	const extensions = [
		".ts",
		".tsx",
		".js",
		".jsx",
		".mjs",
		".mts"
	];
	for (const ext of extensions) {
		const candidate = path.resolve(dir, specifier + ext);
		if (existsSync(candidate) && !statSync(candidate).isDirectory()) return candidate;
	}
	const resolved = path.resolve(dir, specifier);
	for (const ext of extensions) {
		const candidate = path.join(resolved, "index" + ext);
		if (existsSync(candidate)) return candidate;
	}
}
function getNamedExportsViaRegex(source, filePath, visited) {
	const names = /* @__PURE__ */ new Set();
	visited = visited || /* @__PURE__ */ new Set();
	if (filePath) visited.add(filePath);
	const declRegex = new RegExp(`export\\s+(?:async\\s+)?(?:function(?:\\*\\s*|\\s+\\*?\\s*)|const\\s+|let\\s+|var\\s+|class\\s+)(${JS_IDENTIFIER_PATTERN})`, "gu");
	let match;
	while ((match = declRegex.exec(source)) !== null) {
		const name = match[1];
		if (isValidEsmExportName(name)) names.add(name);
	}
	const listRegex = /export\s*\{([^}]+)\}/g;
	const typeOnlySpecifierRegex = new RegExp(`^type\\s+${JS_IDENTIFIER_PATTERN}(?:\\s+as\\s+${JS_IDENTIFIER_PATTERN})?$`, "u");
	const exportSpecifierRegex = new RegExp(`(?:\\S+\\s+as\\s+)?(${JS_IDENTIFIER_PATTERN})$`, "u");
	while ((match = listRegex.exec(source)) !== null) {
		const specifiers = match[1].split(",");
		for (const specifier of specifiers) {
			const trimmed = specifier.trim();
			if (typeOnlySpecifierRegex.test(trimmed)) continue;
			const asMatch = trimmed.match(exportSpecifierRegex);
			if (!asMatch) continue;
			const name = asMatch[1];
			if (isValidEsmExportName(name)) names.add(name);
		}
	}
	if (filePath) {
		const starExportRegex = /export\s+\*\s+from\s+['"]([^'"]+)['"]/g;
		while ((match = starExportRegex.exec(source)) !== null) {
			const specifier = match[1];
			if (!specifier.startsWith(".")) continue;
			const resolvedPath = resolveRelativeModule(filePath, specifier);
			if (!resolvedPath || visited.has(resolvedPath)) continue;
			try {
				const reExportNames = getNamedExportsViaRegex(readFileSync(resolvedPath, "utf-8"), resolvedPath, visited);
				for (const name of reExportNames) names.add(name);
			} catch {}
		}
	}
	return Array.from(names);
}
function getPackageNamedExports(pkg) {
	try {
		const mod = createRequire$1(new URL(`file://${path.join(getPackageDetectionCwd(), "package.json")}`))(pkg);
		return Object.keys(mod).filter((k) => isValidEsmExportName(k));
	} catch {
		return getEsmNamedExports(pkg);
	}
}
function getLocalProviderImportPath(pkg) {
	try {
		const resolved = createRequire$1(new URL(`file://${path.join(getPackageDetectionCwd(), "package.json")}`)).resolve(pkg);
		return isWorkspaceFilePath(resolved) ? resolved : void 0;
	} catch {
		const resolved = getInstalledPackageEntry(pkg, {
			conditions: [
				"browser",
				"import",
				"module",
				"default"
			],
			resolveSubpathWithRequire: false
		});
		return isWorkspaceFilePath(resolved) ? resolved : void 0;
	}
}
function getProjectResolvedImportPath(pkg) {
	if (pkg === getPackageName(pkg)) {
		const esmEntry = getPackageEsmEntryPath(pkg);
		if (esmEntry) return esmEntry;
	}
	try {
		return createRequire$1(new URL(`file://${path.join(getPackageDetectionCwd(), "package.json")}`)).resolve(pkg);
	} catch {
		return;
	}
}
function isWorkspaceFilePath(resolved) {
	if (!resolved) return false;
	let realResolved = resolved;
	try {
		realResolved = realpathSync.native(resolved);
	} catch {}
	return !realResolved.includes("/node_modules/") && !realResolved.includes("\\node_modules\\");
}
function isWorkspacePackageEntry(pkg, resolved) {
	if (!resolved || !path.isAbsolute(resolved) || !isWorkspaceFilePath(resolved)) return false;
	return !!getInstalledPackageJson(pkg, {
		packageName: getPackageName(pkg),
		fromResolvedEntry: resolved
	});
}
function tryResolveImportFromPackageRoot(pkg, root) {
	try {
		return createRequire$1(new URL(`file://${path.join(root, "package.json")}`)).resolve(pkg);
	} catch {
		return;
	}
}
function getConcreteSharedImportSource(pkg, shareItem) {
	const configuredImport = shareItem?.shareConfig.import;
	if (typeof configuredImport === "string") return configuredImport;
	const projectRoot = getPackageDetectionCwd();
	if (tryResolveImportFromPackageRoot(pkg, projectRoot)) return;
	let currentDir = path.dirname(projectRoot);
	while (currentDir !== path.dirname(currentDir)) {
		const resolved = tryResolveImportFromPackageRoot(pkg, currentDir);
		if (resolved) return resolved;
		currentDir = path.dirname(currentDir);
	}
	return tryResolveImportFromPackageRoot(pkg, currentDir);
}
const preBuildCacheMap = {};
const preBuildShareItemMap = {};
const PREBUILD_TAG = "__prebuild__";
function writePreBuildLibPath(pkg, shareItem) {
	if (!preBuildCacheMap[pkg]) preBuildCacheMap[pkg] = new VirtualModule(pkg, PREBUILD_TAG, ".js");
	preBuildShareItemMap[pkg] = shareItem;
	const importSource = getConcreteSharedImportSource(pkg, shareItem) || pkg;
	if (pkg === "react/jsx-dev-runtime") {
		preBuildCacheMap[pkg].writeSync(`
    import __mfPrebuildDefault from ${escapeGeneratedStringLiteral(importSource)};
    import * as __mfPrebuildNamespace from ${escapeGeneratedStringLiteral(importSource)};
    const __mfPrebuildExports = __mfPrebuildDefault ?? __mfPrebuildNamespace;
    export const Fragment = __mfPrebuildExports.Fragment;
    export const jsxDEV = __mfPrebuildExports.jsxDEV;
    export default __mfPrebuildExports;
  `, true);
		return;
	}
	if (pkg === "react/jsx-runtime") {
		preBuildCacheMap[pkg].writeSync(`
    import __mfPrebuildDefault from ${escapeGeneratedStringLiteral(importSource)};
    import * as __mfPrebuildNamespace from ${escapeGeneratedStringLiteral(importSource)};
    const __mfPrebuildExports = __mfPrebuildDefault ?? __mfPrebuildNamespace;
    export const Fragment = __mfPrebuildExports.Fragment;
    export const jsx = __mfPrebuildExports.jsx;
    export const jsxs = __mfPrebuildExports.jsxs;
    export default __mfPrebuildExports;
  `, true);
		return;
	}
	preBuildCacheMap[pkg].writeSync(`
    import * as __mfPrebuildExports from ${escapeGeneratedStringLiteral(importSource)};
    export * from ${escapeGeneratedStringLiteral(importSource)};
    export default __mfPrebuildExports.default ?? __mfPrebuildExports;
  `, true);
}
function getPreBuildLibImportId(pkg) {
	if (!preBuildCacheMap[pkg]) preBuildCacheMap[pkg] = new VirtualModule(pkg, PREBUILD_TAG, ".js");
	return preBuildCacheMap[pkg].getImportId();
}
function getPreBuildShareItem(pkg) {
	return preBuildShareItemMap[pkg];
}
function getSharedImportSource(pkg, shareItem) {
	return getConcreteSharedImportSource(pkg, shareItem) || getPreBuildLibImportId(pkg);
}
const LOAD_SHARE_TAG = "__loadShare__";
const loadShareCacheMap = {};
function getLoadShareImportId(pkg, _isRolldown) {
	if (!loadShareCacheMap[pkg]) loadShareCacheMap[pkg] = new VirtualModule(pkg, LOAD_SHARE_TAG, ".mjs");
	return loadShareCacheMap[pkg].getImportId();
}
function getLoadShareModulePath(pkg, isRolldown) {
	if (!loadShareCacheMap[pkg]) getLoadShareImportId(pkg, isRolldown);
	return loadShareCacheMap[pkg].getPath();
}
function writeLoadShareModule(pkg, shareItem, command, _isRolldown) {
	if (!loadShareCacheMap[pkg]) loadShareCacheMap[pkg] = new VirtualModule(pkg, LOAD_SHARE_TAG, ".mjs");
	const importLine = getRuntimeModuleCacheBootstrapCode();
	if (shareItem.shareConfig.import === false) {
		const namedExports = getPackageNamedExports(pkg);
		let exportLine;
		if (namedExports.length > 0) exportLine = `export default exportModule.default ?? exportModule;\n    ${`const { ${namedExports.map((name, i) => `${name}: __mf_${i}`).join(", ")} } = exportModule;`}\n    ${`export { ${namedExports.map((name, i) => `__mf_${i} as ${name}`).join(", ")} };`}`;
		else {
			mfWarn(`Shared dependency "${pkg}" has import: false but is not installed locally.\n  Named imports (e.g. import { ... } from '${pkg}') will not work in production builds.\n  Install it as a devDependency to enable named export detection.`);
			exportLine = "export default exportModule.default ?? exportModule";
		}
		loadShareCacheMap[pkg].writeSync(`
    ${importLine}
    const exportModule = __mfModuleCache.share[${escapeGeneratedStringLiteral(pkg)}]
    if (exportModule === undefined) {
      throw new Error("[Module Federation] Shared module ${pkg} was imported before federation bootstrap finished.")
    }
    ${exportLine}
  `, true);
		return;
	}
	const concreteSharedImportSource = getConcreteSharedImportSource(pkg, shareItem);
	const sharedImportSource = concreteSharedImportSource || getPreBuildLibImportId(pkg);
	const devImportSource = concreteSharedImportSource || pkg;
	const localProviderPath = getLocalProviderImportPath(pkg);
	const isWorkspacePackage = isWorkspacePackageEntry(pkg, localProviderPath) || isWorkspacePackageEntry(pkg, concreteSharedImportSource);
	const lazyLocalFallbackSource = concreteSharedImportSource || localProviderPath || sharedImportSource;
	const skipServePrebuildWarmup = command !== "build" && (pkg === "lit" || pkg.startsWith("lit/"));
	const usesLazyLocalFallback = isWorkspacePackage && shareItem.shareConfig.singleton === true;
	const namedExports = getPackageNamedExports(pkg);
	let exportLine;
	if (namedExports.length > 0) exportLine = `export default exportModule.default ?? exportModule;\n    ${`const { ${namedExports.map((name, i) => `${name}: __mf_${i}`).join(", ")} } = exportModule;`}\n    ${`export { ${namedExports.map((name, i) => `__mf_${i} as ${name}`).join(", ")} };`}`;
	else if (usesLazyLocalFallback) exportLine = `export default exportModule.default ?? exportModule`;
	else exportLine = `export default exportModule.default ?? exportModule\n    export * from ${escapeGeneratedStringLiteral(sharedImportSource)}`;
	const prebuildImportLine = usesLazyLocalFallback || isWorkspacePackage && command !== "build" ? "" : `import * as __mfLocalShare from ${escapeGeneratedStringLiteral(skipServePrebuildWarmup ? devImportSource : sharedImportSource)};`;
	const devDynamicImportLine = isWorkspacePackage ? "" : command !== "build" && !skipServePrebuildWarmup ? `;() => import(${escapeGeneratedStringLiteral(devImportSource)}).catch(() => {});` : "";
	loadShareCacheMap[pkg].writeSync(`
    ${prebuildImportLine}
    ${devDynamicImportLine}
    ${importLine}
    let exportModule = __mfModuleCache.share[${escapeGeneratedStringLiteral(pkg)}]
    if (exportModule === undefined) {
      ${usesLazyLocalFallback ? `exportModule = await import(${escapeGeneratedStringLiteral(lazyLocalFallbackSource)});
      __mfModuleCache.share[${escapeGeneratedStringLiteral(pkg)}] = exportModule;` : `exportModule = __mfLocalShare;
      __mfModuleCache.share[${escapeGeneratedStringLiteral(pkg)}] = exportModule;`}
    }
    ${exportLine}
  `, true);
}
//#endregion
//#region src/virtualModules/virtualRemoteEntry.ts
let usedShares = /* @__PURE__ */ new Set();
function getUsedShares() {
	return usedShares;
}
function addUsedShares(pkg) {
	usedShares.add(pkg);
}
const LOCAL_SHARED_IMPORT_MAP_ID = "virtual:mf-localSharedImportMap";
function getLocalSharedImportMapPath() {
	const { internalName, name } = getNormalizeModuleFederationOptions();
	return `${LOCAL_SHARED_IMPORT_MAP_ID}:${packageNameEncode(internalName || name)}`;
}
function getResolvedLocalSharedImportMapId() {
	return `\0${getLocalSharedImportMapPath()}`;
}
let invalidateLocalSharedImportMap;
function setLocalSharedImportMapInvalidator(invalidator) {
	invalidateLocalSharedImportMap = invalidator;
}
let prevLocalSharedImportMapContent;
function writeLocalSharedImportMap() {
	const nextContent = generateLocalSharedImportMap();
	if (prevLocalSharedImportMapContent !== nextContent) {
		prevLocalSharedImportMapContent = nextContent;
		invalidateLocalSharedImportMap?.();
	}
}
function shouldUseDirectReactImport() {
	const isVinext = hasPackageDependency("vinext");
	const isAstro = hasPackageDependency("astro");
	return isVinext || isAstro;
}
function getLocalSharedPackagePath(pkg, shareItem) {
	if (shouldUseDirectReactImport() && pkg === "react") return "react";
	return getConcreteSharedImportSource(pkg, shareItem) || getLocalProviderImportPath(pkg) || getSharedImportSource(pkg, shareItem);
}
function getDirectSharedCacheSeedImportPath(pkg, shareItem) {
	return getConcreteSharedImportSource(pkg, shareItem) || getProjectResolvedImportPath(pkg) || getLocalProviderImportPath(pkg) || pkg;
}
function generateLocalSharedImportMap() {
	const useDirectReactImport = shouldUseDirectReactImport();
	const options = getNormalizeModuleFederationOptions();
	return `
    import {loadShare} from "@module-federation/runtime";
    const importMap = {
      ${Array.from(getUsedShares()).sort().map((pkg) => {
		const shareItem = getNormalizeShareItem(pkg);
		return `
        ${JSON.stringify(pkg)}: async () => {
          ${shareItem?.shareConfig.import === false ? `throw new Error(\`[Module Federation] Shared module '\${${JSON.stringify(pkg)}}' must be provided by host\`);` : `let pkg = await import(${JSON.stringify(getLocalSharedPackagePath(pkg, shareItem))});
            return pkg;`}
        }
      `;
	}).join(",")}
    }
      const usedShared = {
      ${Array.from(getUsedShares()).sort().map((key) => {
		const shareItem = getNormalizeShareItem(key);
		if (!shareItem) return null;
		return `
          ${JSON.stringify(key)}: {
            name: ${JSON.stringify(key)},
            version: ${JSON.stringify(shareItem.version)},
            scope: [${JSON.stringify(shareItem.scope)}],
            loaded: false,
            from: ${JSON.stringify(options.internalName)},
            async get () {
              if (${shareItem.shareConfig.import === false}) {
                throw new Error(\`[Module Federation] Shared module '\${${JSON.stringify(key)}}' must be provided by host\`);
              }
              usedShared[${JSON.stringify(key)}].loaded = true
              const {${JSON.stringify(key)}: pkgDynamicImport} = importMap
              const res = await pkgDynamicImport()
              const exportModule = ${JSON.stringify(useDirectReactImport)} && ${JSON.stringify(key)} === "react"
                ? (res?.default ?? res)
                : {...res}
              // All npm packages pre-built by vite will be converted to esm
              Object.defineProperty(exportModule, "__esModule", {
                value: true,
                enumerable: false
              })
              return function () {
                return exportModule
              }
            },
            shareConfig: {
              singleton: ${shareItem.shareConfig.singleton},
              requiredVersion: ${JSON.stringify(shareItem.shareConfig.requiredVersion)},
              ${shareItem.shareConfig.import === false ? "import: false," : ""}
            }
          }
        `;
	}).filter((x) => x !== null).join(",")}
    }
      const usedRemotes = [${Object.keys(getUsedRemotesMap()).map((key) => {
		const remote = options.remotes[key];
		if (!remote) return null;
		return `
                {
                  entryGlobalName: ${JSON.stringify(remote.entryGlobalName)},
                  name: ${JSON.stringify(remote.name)},
                  type: ${JSON.stringify(remote.type)},
                  entry: ${JSON.stringify(remote.entry)},
                  shareScope: ${JSON.stringify(remote.shareScope ?? "default")},
                }
          `;
	}).filter((x) => x !== null).join(",")}
      ]
      export {
        usedShared,
        usedRemotes
      }
      `;
}
function generateUsedSharedPreloadConfig() {
	return `{
      ${getOrderedUsedShares().map((pkg) => {
		const shareItem = getShareItemForPreload(pkg);
		if (!shareItem) return null;
		return `${JSON.stringify(pkg)}: {
            shareConfig: {
              singleton: ${shareItem.shareConfig.singleton},
              requiredVersion: ${JSON.stringify(shareItem.shareConfig.requiredVersion)},
              ${shareItem.shareConfig.import === false ? "import: false," : ""}
            }
          }`;
	}).filter((item) => item !== null).join(",\n")}
    }`;
}
function getOrderedUsedShares() {
	const shares = new Set(getUsedShares());
	try {
		Object.keys(getNormalizeModuleFederationOptions().shared).forEach((pkg) => {
			if (!pkg.endsWith("/")) shares.add(pkg);
		});
	} catch {}
	return Array.from(shares).sort((a, b) => {
		const priority = (pkg) => pkg === "react" ? 0 : pkg === "react-dom" ? 1 : pkg.startsWith("react/") ? 2 : 3;
		return priority(a) - priority(b) || a.localeCompare(b);
	});
}
function getShareItemForPreload(pkg) {
	const shared = getNormalizeModuleFederationOptions().shared;
	const wildcardKey = `${pkg.startsWith("@") ? pkg.split("/").slice(0, 2).join("/") : pkg.split("/")[0]}/`;
	if (isExplicitSharedKey(pkg)) return shared[pkg];
	if (isExplicitSharedKey(wildcardKey)) return shared[wildcardKey];
}
function generateSharedCacheSeedItem(pkg, importPath) {
	return `if (__mfModuleCache.share[${JSON.stringify(pkg)}] === undefined) {
        const mod = await import(${JSON.stringify(importPath)});
        const exportModule = ${JSON.stringify(shouldUseDirectReactImport())} && ${JSON.stringify(pkg)} === "react"
          ? (mod?.default ?? mod)
          : {...mod};
        Object.defineProperty(exportModule, "__esModule", {
          value: true,
          enumerable: false
        });
        __mfModuleCache.share[${JSON.stringify(pkg)}] = exportModule;
      }`;
}
function generateDirectSharedCacheSeedCode(command = "build") {
	return getOrderedUsedShares().map((pkg) => {
		const shareItem = getShareItemForPreload(pkg);
		if (!shareItem || shareItem.shareConfig.import === false) return null;
		return generateSharedCacheSeedItem(pkg, command === "serve" ? getLocalSharedPackagePath(pkg, shareItem) : getDirectSharedCacheSeedImportPath(pkg, shareItem));
	}).filter((item) => item !== null).join("\n");
}
function getBrowserImportPath(importPath) {
	if (/^(?:[a-zA-Z]:[\\/]|\/)/.test(importPath) && !importPath.startsWith("/@")) return `/@fs/${importPath}`;
	return importPath;
}
function getHostAutoInitSharedSeedItems() {
	return getOrderedUsedShares().map((pkg) => ({
		pkg,
		shareItem: getShareItemForPreload(pkg)
	})).filter(({ shareItem }) => shareItem?.shareConfig.import === false).sort((a, b) => {
		const priority = (pkg) => pkg === "vue" ? 0 : pkg === "pinia" ? 1 : 2;
		const aIsLocal = !!getLocalProviderImportPath(a.pkg);
		const bIsLocal = !!getLocalProviderImportPath(b.pkg);
		return priority(a.pkg) - priority(b.pkg) || Number(aIsLocal) - Number(bIsLocal) || a.pkg.localeCompare(b.pkg);
	});
}
function generateHostAutoInitSharedCacheSeedCode() {
	return getHostAutoInitSharedSeedItems().map(({ pkg, shareItem }) => {
		if (!shareItem) return null;
		return generateSharedCacheSeedItem(pkg, getBrowserImportPath(getDirectSharedCacheSeedImportPath(pkg, shareItem)));
	}).filter((item) => item !== null).join("\n");
}
const REMOTE_ENTRY_ID = "virtual:mf-REMOTE_ENTRY_ID";
function getRemoteEntryId(options) {
	return `${REMOTE_ENTRY_ID}:${`${options.internalName}__${options.filename}`.replace(/[^a-zA-Z0-9_-]/g, "_")}`;
}
function generateRemoteEntry(options, virtualExposesId = getVirtualExposesId(options), command = "build") {
	const pluginImportNames = options.runtimePlugins.map((p, i) => {
		if (typeof p === "string") return [
			`$runtimePlugin_${i}`,
			`import $runtimePlugin_${i} from "${p}";`,
			`undefined`
		];
		else return [
			`$runtimePlugin_${i}`,
			`import $runtimePlugin_${i} from "${p[0]}";`,
			serializeRuntimeOptions(p[1])
		];
	});
	return `
  // Shim Vue HMR runtime for dev-compiled components loaded by a non-Vite host.
  // When a remote is served by a Vite dev server, Vue's SFC compiler injects HMR
  // hooks that reference __VUE_HMR_RUNTIME__. This global only exists on pages
  // served by Vite's client runtime. When a production host loads the remote,
  // the HMR calls would throw. This no-op shim prevents that.
  if (typeof __VUE_HMR_RUNTIME__ === 'undefined') {
    globalThis.__VUE_HMR_RUNTIME__ = { createRecord() {}, rerender() {}, reload() {} };
  }
  import {createInstance, loadRemote} from "@module-federation/runtime";
  ${pluginImportNames.map((item) => item[1]).join("\n")}
  ${command === "build" ? getRuntimeInitResolveBootstrapCode() : getRuntimeInitBootstrapCode() + "\n  const { initResolve } = globalThis[globalKey];"}
  ${getRuntimeModuleCacheBootstrapCode()}
  const initTokens = {}
  const shareScopeName = ${JSON.stringify(options.shareScope)}
  const mfName = ${JSON.stringify(options.internalName)}
  let runtimeInstance
  let localSharedImportMapPromise
  let exposesMapPromise
  const shouldRetrySharedInitError = ${command !== "build"} && ((error) => {
    const message = String((error && error.message) || error || '');
    return message.includes('Importing a module script failed') ||
      message.includes('Failed to fetch') ||
      message.includes('Load failed') ||
      message.includes('Outdated Optimize Dep');
  });
  const waitSharedInitRetry = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
  async function retrySharedInit(fn) {
    for (let attempt = 0; ; attempt++) {
      try {
        return await fn();
      } catch (e) {
        const canRetry = typeof shouldRetrySharedInitError === 'function' && shouldRetrySharedInitError(e);
        if (!canRetry || attempt >= 19) throw e;
        await waitSharedInitRetry(250);
      }
    }
  }

  async function getLocalSharedImportMap() {
    if (!localSharedImportMapPromise) {
      localSharedImportMapPromise = retrySharedInit(() => import("${getLocalSharedImportMapPath()}"))
        .catch((e) => { localSharedImportMapPromise = undefined; throw e; });
    }
    return localSharedImportMapPromise
  }

  async function getExposesMap() {
    if (!exposesMapPromise) {
      exposesMapPromise = retrySharedInit(() => import("${virtualExposesId}"))
        .then((mod) => mod.default ?? mod)
        .catch((e) => { exposesMapPromise = undefined; throw e; });
    }
    return exposesMapPromise
  }

  async function init(shared = {}, initScope = []) {
    const {usedShared, usedRemotes} = await getLocalSharedImportMap()
    ${generateDirectSharedCacheSeedCode(command)}
    const runtimeOptions = {
      name: mfName,
      remotes: usedRemotes,
      shared: usedShared,
      plugins: [${pluginImportNames.map((item) => `${item[0]}(${item[2]})`).join(", ")}],
      ${options.shareStrategy ? `shareStrategy: '${options.shareStrategy}'` : ""}
    };
    if (!runtimeInstance) {
      runtimeInstance = createInstance(runtimeOptions);
    } else {
      runtimeInstance.initOptions(runtimeOptions);
    }
    const initRes = runtimeInstance;
    // handling circular init calls
    var initToken = initTokens[shareScopeName];
    if (!initToken)
      initToken = initTokens[shareScopeName] = { from: mfName };
    if (initScope.indexOf(initToken) >= 0) return;
    initScope.push(initToken);
    initRes.initShareScopeMap('${options.shareScope}', shared);
    initResolve(initRes)
    try {
      await retrySharedInit(async () => {
        await Promise.all(await initRes.initializeSharing('${options.shareScope}', {
          strategy: '${options.shareStrategy}',
          from: "build",
          initScope
        }));
      });
    } catch (e) {
      console.error('[Module Federation]', e)
    }
    return initRes
  }

  async function getExposes(moduleName) {
    const exposesMap = await getExposesMap()
    if (!(moduleName in exposesMap)) throw new Error(\`[Module Federation] Module \${moduleName} does not exist in container.\`)
    return (exposesMap[moduleName])().then(res => () => res)
  }
  export {
      init,
      getExposes as get
  }
  `;
}
const hostAutoInitModule = new VirtualModule("hostAutoInit", "__H_A_I__");
let currentHostAutoInitRemoteEntryId = REMOTE_ENTRY_ID;
let currentHostAutoInitCommand = "build";
function generateHostAutoInitCode(remoteEntryImport, _command = "build") {
	return `
    ${getRuntimeModuleCacheBootstrapCode()}
    let hostInitPromise;
    async function initHost() {
      if (!hostInitPromise) {
        hostInitPromise = (async () => {
          ${generateHostAutoInitSharedCacheSeedCode()}
          const remoteEntry = await import(${remoteEntryImport});
          const runtime = await remoteEntry.init();
          const usedShared = ${generateUsedSharedPreloadConfig()};
          for (const [pkg, share] of Object.entries(usedShared)) {
            if (__mfModuleCache.share[pkg] !== undefined) {
              continue;
            }
            await runtime.loadShare(pkg, {
              customShareInfo: { shareConfig: share.shareConfig }
            }).then((factory) => {
              const mod = typeof factory === "function" ? factory() : factory;
              return Promise.resolve(mod).then((resolved) => {
                __mfModuleCache.share[pkg] = resolved;
              });
            });
          }
          const __mfRemotePreloads = [];
          await Promise.all(__mfRemotePreloads);
          return runtime;
        })();
      }
      return hostInitPromise;
    }
    hostInitPromise = initHost();
    export { initHost, hostInitPromise };
    `;
}
function writeHostAutoInit(remoteEntryId = REMOTE_ENTRY_ID, command = "build") {
	currentHostAutoInitRemoteEntryId = remoteEntryId;
	currentHostAutoInitCommand = command;
	hostAutoInitModule.writeSync(generateHostAutoInitCode(JSON.stringify(remoteEntryId), command), true);
}
function refreshHostAutoInit() {
	try {
		writeHostAutoInit(currentHostAutoInitRemoteEntryId, currentHostAutoInitCommand);
	} catch {}
}
function getHostAutoInitImportId() {
	return hostAutoInitModule.getImportId();
}
function getHostAutoInitPath() {
	return hostAutoInitModule.getPath();
}
//#endregion
//#region src/virtualModules/virtualRemotes.ts
const cacheRemoteMap = {};
const LOAD_REMOTE_TAG = "__loadRemote__";
function getRemoteVirtualModule(remote, command) {
	if (!cacheRemoteMap[remote]) {
		cacheRemoteMap[remote] = new VirtualModule(remote, LOAD_REMOTE_TAG, ".mjs");
		cacheRemoteMap[remote].writeSync(generateRemotes(remote, command));
	}
	return cacheRemoteMap[remote];
}
const usedRemotesMap = {};
function addUsedRemote(remoteKey, remoteModule) {
	if (!usedRemotesMap[remoteKey]) usedRemotesMap[remoteKey] = /* @__PURE__ */ new Set();
	usedRemotesMap[remoteKey].add(remoteModule);
}
function getUsedRemotesMap() {
	return usedRemotesMap;
}
function generateRemotes(id, command) {
	const useReactProxy = command === "serve" && hasPackageDependency("react");
	const reactImportLine = useReactProxy ? `import * as __mfReact from "react";` : "";
	const importLine = command === "build" ? `${getRuntimeModuleCacheBootstrapCode()}
    import { hostInitPromise as __mfHostInitPromise } from ${JSON.stringify(getHostAutoInitPath())};` : `${getRuntimeInitBootstrapCode()}
    const { initPromise, moduleCache: __mfModuleCache } = globalThis[globalKey];`;
	const exportLine = command === "serve" ? `if (__mfRemotePending) {
  const mod = await __mfRemotePending;
  if (mod !== undefined) exportModule = mod;
}
export const __moduleExports = exportModule;
export const __mf_remote_pending = Promise.resolve(exportModule);
export default exportModule?.__esModule ? exportModule.default : exportModule.default ?? exportModule` : command === "build" ? `if (__mfRemotePending) {
  const mod = await __mfRemotePending;
  if (mod !== undefined) exportModule = mod;
}
export const __moduleExports = exportModule;
export const __mf_remote_pending = Promise.resolve(exportModule);
export default exportModule?.__esModule ? exportModule.default : exportModule.default ?? exportModule` : "export default exportModule";
	return `
    ${reactImportLine}
    ${importLine}
    ${command !== "build" ? `
    function __mfCreateRemoteProxy(pendingPromise) {
      const listeners = new Set();
      pendingPromise?.finally(() => {
        for (const listener of listeners) listener();
      });
      const getModule = () => __mfModuleCache.remote[${JSON.stringify(id)}];
      const proxyTarget = function (...args) {
        ${useReactProxy ? `const [, setVersion] = __mfReact.useState(0);
        __mfReact.useEffect(() => {
          const listener = () => setVersion((value) => value + 1);
          listeners.add(listener);
          if (getModule()) listener();
          return () => listeners.delete(listener);
        }, []);` : ""}
        const mod = getModule();
        const fn = mod && (mod.default ?? mod);
        if (fn !== undefined && fn !== null) {
          ${useReactProxy ? `return __mfReact.createElement(fn, args[0]);` : `return fn.apply(this, args);`}
        }
        ${useReactProxy ? `return null;` : `throw pendingPromise;`}
      };
      return new Proxy(proxyTarget, {
        get(_target, prop) {
          if (prop === "__mf_is_remote_proxy") return true;
          if (prop === "__esModule") return true;
          if (prop === "then") return undefined;
          const mod = getModule();
          if (mod) {
            return prop in mod ? mod[prop] : mod.default?.[prop];
          }
          ${useReactProxy ? `return undefined;` : `throw pendingPromise;`}
        },
        ownKeys() {
          const mod = getModule();
          const keys = new Set(mod ? Reflect.ownKeys(mod) : []);
          // Proxy invariant: must include non-configurable target own keys
          for (const k of Reflect.ownKeys(proxyTarget)) {
            const d = Object.getOwnPropertyDescriptor(proxyTarget, k);
            if (d && !d.configurable) keys.add(k);
          }
          return Array.from(keys);
        },
        getOwnPropertyDescriptor(_target, prop) {
          // Proxy invariant: non-configurable target props must be reported accurately
          const targetDesc = Object.getOwnPropertyDescriptor(proxyTarget, prop);
          if (targetDesc && !targetDesc.configurable) return targetDesc;
          const mod = getModule();
          if (!mod) return undefined;
          return Object.getOwnPropertyDescriptor(mod, prop) || {
            configurable: true,
            enumerable: true,
            value: mod[prop],
          };
        },
        apply(target, thisArg, args) {
          return target.apply(thisArg, args);
        }
      });
    }` : ""}
    let __mfRemotePending;
    let exportModule = __mfModuleCache.remote[${JSON.stringify(id)}]
    if (exportModule === undefined) {
      ${command !== "build" ? `const pendingKey = ${JSON.stringify(`__mf_pending__${id}`)};
      if (!__mfModuleCache.remote[pendingKey]) {
        __mfModuleCache.remote[pendingKey] = initPromise
          .then((runtime) => runtime.loadRemote(${JSON.stringify(id)}))
          .then((mod) => {
            __mfModuleCache.remote[${JSON.stringify(id)}] = mod;
            delete __mfModuleCache.remote[pendingKey];
            return mod;
          })
          .catch(() => {
            delete __mfModuleCache.remote[pendingKey];
          });
      }` : ""}
      ${command !== "build" ? `__mfRemotePending = __mfModuleCache.remote[pendingKey];
      exportModule = __mfCreateRemoteProxy(__mfRemotePending);` : `const pendingKey = ${JSON.stringify(`__mf_pending__${id}`)};
      if (!__mfModuleCache.remote[pendingKey]) {
        __mfModuleCache.remote[pendingKey] = __mfHostInitPromise
          .then((runtime) => runtime.loadRemote(${JSON.stringify(id)}))
          .then((mod) => {
            __mfModuleCache.remote[${JSON.stringify(id)}] = mod;
            delete __mfModuleCache.remote[pendingKey];
            return mod;
          })
          .catch((error) => {
            delete __mfModuleCache.remote[pendingKey];
            throw error;
          });
      }
      __mfRemotePending = __mfModuleCache.remote[pendingKey];
      exportModule = {};`}
    }
    ${exportLine}
  `;
}
//#endregion
//#region src/plugins/pluginAddEntry.ts
function getFirstHtmlEntryFile(entryFiles) {
	return entryFiles.find((file) => file.endsWith(".html"));
}
const addEntry = ({ entryName, entryPath, fileName, inject = "entry", forceClientInjected }) => {
	const DEV_HTML_PROXY_PREFIX = "virtual:mf-html-entry-proxy?";
	const ENTRY_BOOTSTRAP_QUERY = "?mf-entry-bootstrap";
	const waitsForInit = entryName === "hostInit";
	const getEntryPath = () => typeof entryPath === "function" ? entryPath() : entryPath;
	let devEntryPath = "";
	let entryFiles = [];
	let htmlFilePath;
	let _command;
	let emitFileId;
	let viteConfig;
	let clientInjected = forceClientInjected ?? false;
	let emittedFileName;
	function skipSvelteKitSsrBuild() {
		return (_command === "build" || viteConfig?.command === "build") && viteConfig?.build?.ssr && hasPackageDependency("@sveltejs/kit");
	}
	function isSvelteKitServerModule(id) {
		return hasPackageDependency("@sveltejs/kit") && (id.includes(".svelte-kit/generated/") || id.includes("/@sveltejs/kit/src/runtime/server/"));
	}
	function rewriteSvelteKitInlineStart(html, initPath) {
		return html.replace(/<script>([\s\S]*?)<\/script>/gi, (scriptTag, body) => {
			if (!body.includes("kit.start(app, element);") || !body.includes("Promise.all([")) return scriptTag;
			const blockStart = body.indexOf("{");
			const blockEnd = body.lastIndexOf("}");
			if (blockStart === -1 || blockEnd <= blockStart) return scriptTag;
			return `<script>${body.slice(0, blockStart + 1) + `
(async () => {
  await import(${JSON.stringify(initPath)}).then(({ initHost }) => initHost());
` + body.slice(blockStart + 1, blockEnd) + `
})();
` + body.slice(blockEnd)}<\/script>`;
		});
	}
	function walkFiles(dir, predicate) {
		if (!fs$1.existsSync(dir)) return [];
		return fs$1.readdirSync(dir, { withFileTypes: true }).flatMap((entry) => {
			const entryPath = path$1.join(dir, entry.name);
			if (entry.isDirectory()) return walkFiles(entryPath, predicate);
			return entry.isFile() && predicate(entry.name) ? [entryPath] : [];
		});
	}
	function walkHtmlFiles(dir) {
		return walkFiles(dir, (fileName) => fileName.endsWith(".html"));
	}
	function toRelativeImport(fromFile, targetFile) {
		const relative = path$1.relative(path$1.dirname(fromFile), targetFile).replace(/\\/g, "/");
		return relative.startsWith(".") ? relative : `./${relative}`;
	}
	function patchSvelteKitStaticHtml() {
		const buildDir = path$1.resolve(viteConfig.root, "build");
		let initFile = emittedFileName ? path$1.resolve(buildDir, emittedFileName) : void 0;
		if (!initFile || !fs$1.existsSync(initFile)) initFile = walkFiles(buildDir, (fileName) => fileName.endsWith(".js")).find((file) => {
			const code = fs$1.readFileSync(file, "utf-8");
			return code.includes("hostInitPromise") && code.includes("initHost");
		});
		if (!initFile) return false;
		let patched = false;
		for (const htmlFile of walkHtmlFiles(buildDir)) {
			const html = fs$1.readFileSync(htmlFile, "utf-8");
			const rewritten = rewriteSvelteKitInlineStart(html, toRelativeImport(htmlFile, initFile));
			if (rewritten !== html) {
				fs$1.writeFileSync(htmlFile, rewritten);
				patched = true;
			}
		}
		return patched;
	}
	function getBootstrapSource(initSrc, entrySrc, useSystemImportFallback = false) {
		const remotePreloads = Object.entries(getUsedRemotesMap()).flatMap(([remoteKey, remotes]) => Array.from(remotes).filter((remote) => remote !== remoteKey)).sort().map((remote) => `runtime.loadRemote(${JSON.stringify(remote)})`).join(",");
		const importHelper = useSystemImportFallback ? `const __mfImport = (src) =>
  globalThis.System && typeof globalThis.System.import === 'function'
    ? globalThis.System.import(src)
    : import(src);
` : "";
		const importExpression = (src) => useSystemImportFallback ? `__mfImport(${JSON.stringify(src)})` : `import(${JSON.stringify(src)})`;
		return `${getRuntimeModuleCacheBootstrapCode()}
${importHelper}(async () => {
  const { initHost } = await ${importExpression(initSrc)};
  const runtime = await initHost();
  const __mfRemotePreloads = [${remotePreloads}];
  await Promise.all(__mfRemotePreloads);
})().then(() => ${importExpression(entrySrc)});
`;
	}
	function getSystemBootstrapSource(initSrc, entrySrc) {
		return getBootstrapSource(initSrc, entrySrc, true);
	}
	function injectHtml() {
		return inject === "html" && (htmlFilePath || hasPackageDependency("@sveltejs/kit"));
	}
	function injectEntry() {
		if (inject === "html" && hasPackageDependency("@sveltejs/kit")) return false;
		return inject === "entry" || !htmlFilePath;
	}
	return [{
		name: "add-entry",
		apply: "serve",
		config(_config, { command }) {
			_command = command;
		},
		configResolved(config) {
			viteConfig = config;
			const resolvedEntryPath = getEntryPath();
			if (resolvedEntryPath.startsWith("virtual:mf")) devEntryPath = config.base + "@id/" + resolvedEntryPath;
			else {
				const normalized = resolvedEntryPath.replace(/\\\\?/g, "/");
				const root = config.root.replace(/\\\\?/g, "/").replace(/\/$/, "");
				const relativePath = normalized.startsWith(root + "/") ? normalized.slice(root.length) : "/" + normalized.replace(/^[A-Za-z]:[\\/]/, "");
				devEntryPath = config.base + relativePath.replace(/^\//, "");
			}
		},
		configureServer(server) {
			server.middlewares.use((req, _res, next) => {
				if (!fileName) {
					next();
					return;
				}
				if (req.url && req.url.startsWith((viteConfig.base + fileName).replace(/^\/?/, "/"))) req.url = devEntryPath;
				next();
			});
		},
		transformIndexHtml: {
			order: "pre",
			handler(c) {
				if (!injectHtml()) return;
				clientInjected = true;
				const base = viteConfig.base.replace(/\/$/, "");
				const stripBase = (p) => base && p.startsWith(base + "/") ? p.slice(base.length) : p;
				const html = rewriteEntryScripts(c, (originalSrc) => {
					return `/@id/${DEV_HTML_PROXY_PREFIX}${new URLSearchParams({
						init: sanitizeDevEntryPath(stripBase(devEntryPath)),
						entry: sanitizeDevEntryPath(stripBase(originalSrc))
					}).toString()}`;
				});
				return html === c ? injectEntryScript(c, stripBase(devEntryPath)) : html;
			}
		},
		resolveId(id) {
			if (id.startsWith(DEV_HTML_PROXY_PREFIX)) return id;
		},
		load(id) {
			if (!id.startsWith(DEV_HTML_PROXY_PREFIX)) return;
			const params = new URLSearchParams(id.slice(28));
			const initSrc = params.get("init");
			const entrySrc = params.get("entry");
			if (!initSrc || !entrySrc) return;
			return getBootstrapSource(initSrc, entrySrc);
		},
		transform(code, id) {
			if (id.includes("node_modules") || inject !== "html" || htmlFilePath) return;
			if (id.includes(".svelte-kit") && id.includes("internal.js")) return code.replace(/<head>/g, "<head><script type=\\\"module\\\" src=\\\"" + sanitizeDevEntryPath(devEntryPath) + "\\\"><\/script>");
		}
	}, {
		name: "add-entry",
		enforce: "post",
		configResolved(config) {
			viteConfig = config;
			const inputOptions = config.build.rollupOptions.input;
			if (!inputOptions) htmlFilePath = path$1.resolve(config.root, "index.html");
			else if (typeof inputOptions === "string") entryFiles = [inputOptions];
			else if (Array.isArray(inputOptions)) entryFiles = inputOptions;
			else if (typeof inputOptions === "object") entryFiles = Object.values(inputOptions);
			if (entryFiles.length > 0) htmlFilePath = getFirstHtmlEntryFile(entryFiles);
			if (_command === "serve" && htmlFilePath && fs$1.existsSync(htmlFilePath)) {
				const htmlContent = fs$1.readFileSync(htmlFilePath, "utf-8");
				const scriptRegex = /<script\s+[^>]*src=["']([^"']+)["'][^>]*>/gi;
				let match;
				while ((match = scriptRegex.exec(htmlContent)) !== null) entryFiles.push(match[1]);
			}
		},
		buildStart() {
			if (_command === "serve") return;
			if (skipSvelteKitSsrBuild()) return;
			const hasHash = fileName?.includes?.("[hash");
			const emitFileOptions = {
				name: entryName,
				type: "chunk",
				id: getEntryPath(),
				preserveSignature: "strict"
			};
			if (!hasHash) emitFileOptions.fileName = fileName;
			emitFileId = this.emitFile(emitFileOptions);
			if (htmlFilePath && fs$1.existsSync(htmlFilePath)) {
				const htmlContent = fs$1.readFileSync(htmlFilePath, "utf-8");
				const scriptRegex = /<script\s+[^>]*src=["']([^"']+)["'][^>]*>/gi;
				let match;
				while ((match = scriptRegex.exec(htmlContent)) !== null) entryFiles.push(match[1]);
			}
		},
		generateBundle(_options, bundle) {
			if (skipSvelteKitSsrBuild()) return;
			if (!injectHtml()) return;
			const file = this.getFileName(emitFileId);
			emittedFileName = file;
			const resolvePath = (htmlFileName) => {
				if (!viteConfig.experimental?.renderBuiltUrl) return viteConfig.base + file;
				const result = viteConfig.experimental.renderBuiltUrl(file, {
					hostId: htmlFileName,
					hostType: "html",
					type: "asset",
					ssr: false
				});
				if (typeof result === "string") return result;
				if (result && typeof result === "object") {
					if ("runtime" in result) {
						mfWarn("renderBuiltUrl returned runtime code for HTML injection. Runtime code cannot be used in <script src=\"\">. Falling back to base path.");
						return viteConfig.base + file;
					}
					if (result.relative) return file;
				}
				return viteConfig.base + file;
			};
			let bootstrapIndex = 0;
			for (const fileName in bundle) if (fileName.endsWith(".html")) {
				let htmlAsset = bundle[fileName];
				if (htmlAsset.type === "chunk") return;
				let htmlContent = htmlAsset.source.toString() || "";
				const initPath = resolvePath(fileName);
				const scriptRegex = /<script\b(?=[^>]*\btype=["']module["'])(?=[^>]*\bsrc=["']([^"']+)["'])[^>]*>\s*<\/script>/gi;
				let rewritten = false;
				htmlContent = htmlContent.replace(scriptRegex, (scriptTag, entrySrc) => {
					rewritten = true;
					const bootstrapFileName = `mf-entry-bootstrap-${bootstrapIndex++}.js`;
					const bootstrapRef = this.emitFile({
						type: "asset",
						fileName: bootstrapFileName,
						source: getSystemBootstrapSource(initPath, entrySrc)
					});
					const bootstrapPath = viteConfig.base + this.getFileName(bootstrapRef);
					return scriptTag.replace(entrySrc, bootstrapPath);
				});
				if (!rewritten) {
					const svelteKitHtml = rewriteSvelteKitInlineStart(htmlContent, initPath);
					if (svelteKitHtml !== htmlContent) {
						htmlAsset.source = svelteKitHtml;
						continue;
					}
					const scriptContent = `
          <script type="module" src="${initPath}"><\/script>
        `;
					htmlContent = htmlContent.replace("<head>", `<head>${scriptContent}`);
				}
				htmlAsset.source = htmlContent;
			}
		},
		closeBundle() {
			if (_command === "serve" || skipSvelteKitSsrBuild()) return;
			let attempts = 0;
			const retry = () => {
				attempts += 1;
				if (!patchSvelteKitStaticHtml() && attempts < 20) setTimeout(retry, 50);
			};
			setTimeout(retry, 0);
		},
		transform(code, id) {
			if (skipSvelteKitSsrBuild()) return;
			if (isSvelteKitServerModule(id)) return;
			if (id.includes(ENTRY_BOOTSTRAP_QUERY)) return;
			const isVinext = hasPackageDependency("vinext");
			if (isVinext && inject === "html" && id.includes("virtual:vite-rsc/remove-duplicate-server-css")) {
				const namespaceReactImport = `import * as React from 'react';`;
				if (code.includes(namespaceReactImport)) return;
				const rewritten = code.replace(/import\s+React\s+from\s+['"]react['"];?/, namespaceReactImport);
				return rewritten === code ? void 0 : mapCodeToCodeWithSourcemap(rewritten);
			}
			if (isVinext && inject === "html" && (id.includes("virtual:vite-rsc/entry-browser") || id.includes("virtual:vinext-app-browser-entry"))) {
				const injection = `import ${JSON.stringify(getEntryPath())};\n`;
				if (code.includes(injection.trim())) {
					clientInjected = true;
					return;
				}
				clientInjected = true;
				return mapCodeToCodeWithSourcemap(injection + code);
			}
			if (injectEntry() && entryFiles.some((file) => id.endsWith(file)) || _command === "serve" && inject === "html" && !isVinext && !clientInjected && !id.startsWith("\0") && !id.includes("node_modules") && /\.(js|ts|mjs|vue|jsx|tsx)(\?|$)/.test(id)) {
				clientInjected = true;
				if (!waitsForInit) return mapCodeToCodeWithSourcemap(`import ${JSON.stringify(getEntryPath())};\n` + code);
				const entrySrc = id.includes("?") ? `${id}&${ENTRY_BOOTSTRAP_QUERY.slice(1)}` : `${id}${ENTRY_BOOTSTRAP_QUERY}`;
				return mapCodeToCodeWithSourcemap(getBootstrapSource(getEntryPath(), entrySrc));
			}
		}
	}];
};
//#endregion
//#region src/plugins/pluginCheckAliasConflicts.ts
/**
* Check if user-defined alias conflicts with shared modules
* This should run after aliasToArrayPlugin to ensure alias is an array
*/
function checkAliasConflicts(options) {
	const { shared = {} } = options;
	const sharedKeys = Object.keys(shared);
	return {
		name: "check-alias-conflicts",
		configResolved(config) {
			if (sharedKeys.length === 0) return;
			const userAliases = config.resolve?.alias || [];
			const conflicts = [];
			const matchesSharedKey = (aliasEntry, sharedKey) => {
				const findPattern = aliasEntry.find;
				if (typeof findPattern === "string") return findPattern === sharedKey || sharedKey.startsWith(findPattern + "/");
				if (findPattern instanceof RegExp) return findPattern.test(sharedKey);
				return false;
			};
			for (const sharedKey of sharedKeys) for (const aliasEntry of userAliases) {
				const replacement = aliasEntry.replacement;
				if (!matchesSharedKey(aliasEntry, sharedKey)) continue;
				if (replacement === "$1") break;
				if (typeof replacement === "string") {
					if (getPackageNameFromNodeModulePath(replacement) === (sharedKey.endsWith("/") ? sharedKey.slice(0, -1) : sharedKey)) continue;
					conflicts.push({
						sharedModule: sharedKey,
						alias: String(aliasEntry.find),
						target: replacement
					});
				}
			}
			if (conflicts.length > 0) {
				mfWarn("Detected alias conflicts with shared modules:");
				conflicts.forEach(({ sharedModule, alias, target }) => {
					mfWarn(`Shared module "${sharedModule}" is aliased by "${alias}" to "${target}"`);
				});
				mfWarn("This may cause runtime errors as the shared module will bypass Module Federation's sharing mechanism.");
			}
		}
	};
}
//#endregion
//#region src/plugins/pluginDevRemoteHmr.ts
const REMOTE_HMR_ENDPOINT = "__mf_hmr";
const REMOTE_HMR_EVENT = "mf:remote-update";
const REMOTE_HMR_CONNECT_RETRY_DELAY_MS = 1e3;
const REMOTE_HMR_CONNECT_MAX_RETRIES = 10;
/**
* Proxy module served for `/@react-refresh` on MF remote dev servers.
* Delegates to the host page's RefreshRuntime instance via
* `window.location.origin`, ensuring a single shared component registry
* across federation boundaries. A `configureServer` middleware is used
* instead of `resolveId` because `@vitejs/plugin-react`'s
* `vite:react-refresh` sub-plugin uses `enforce: 'pre'` and typically
* wins the `resolveId` race.
*/
const REACT_REFRESH_PROXY_MODULE = [
	`const __rt = await import(window.location.origin + '/@react-refresh');`,
	`export const injectIntoGlobalHook = __rt.injectIntoGlobalHook;`,
	`export const register = __rt.register;`,
	`export const createSignatureFunctionForTransform = __rt.createSignatureFunctionForTransform;`,
	`export const registerExportsForReactRefresh = __rt.registerExportsForReactRefresh;`,
	`export const validateRefreshBoundaryAndEnqueueUpdate = __rt.validateRefreshBoundaryAndEnqueueUpdate;`,
	`export const __hmr_import = __rt.__hmr_import;`,
	`export default __rt.default || __rt;`
].join("\n");
function getBasePath$1(base) {
	if (!base) return "/";
	if (base.startsWith("http://") || base.startsWith("https://")) try {
		return new URL(base).pathname || "/";
	} catch {
		return "/";
	}
	return base;
}
function getRemoteHmrPath(base) {
	return `${getBasePath$1(base).replace(/\/?$/, "/")}${REMOTE_HMR_ENDPOINT}`.replace(/\/{2,}/g, "/");
}
function getHmrWsPath(base, hmrPath) {
	const normalizedBase = getBasePath$1(base);
	const normalizedPath = getBasePath$1(hmrPath || "");
	if (!normalizedPath || normalizedPath === "/") return normalizedBase;
	return `${normalizedBase.endsWith("/") ? normalizedBase.slice(0, -1) : normalizedBase}/${normalizedPath.startsWith("/") ? normalizedPath.slice(1) : normalizedPath}`;
}
function shouldIgnoreFile(file, options) {
	return file.includes("/node_modules/") || file.includes("\\node_modules\\") || file.includes(`/${options.virtualModuleDir}/`) || file.includes(`\\${options.virtualModuleDir}\\`) || file.includes("/.vite/") || file.includes("\\.vite\\") || file.includes("/.mf/") || file.includes("\\.mf\\") || file.includes("/mf-manifest.json") || file.includes("\\mf-manifest.json") || file.includes("/mf-stats.json") || file.includes("\\mf-stats.json");
}
function getRemoteHmrWsUrl(server) {
	const hmr = server.config.server.hmr;
	return `${hmr && typeof hmr === "object" && hmr.protocol ? hmr.protocol : server.config.server.https ? "wss" : "ws"}://${hmr && typeof hmr === "object" && hmr.host ? hmr.host : typeof server.config.server.host === "string" && server.config.server.host !== "0.0.0.0" ? server.config.server.host : "localhost"}:${hmr && typeof hmr === "object" && (hmr.clientPort || hmr.port) ? hmr.clientPort || hmr.port : server.config.server.port}${getHmrWsPath(server.config.base, hmr && typeof hmr === "object" ? hmr.path : "")}?token=${server.config.webSocketToken}`;
}
function getLocalFallbackOrigin(server) {
	return `${server.config.server.https ? "https" : "http"}://${typeof server.config.server.host === "string" && server.config.server.host !== "0.0.0.0" && server.config.server.host !== "::" ? server.config.server.host : "localhost"}:${server.config.server.port || 5173}`;
}
function getRemoteHmrEndpoint(remoteEntry, server) {
	try {
		const remoteManifestUrl = new URL(remoteEntry, getLocalFallbackOrigin(server));
		remoteManifestUrl.pathname = `/${remoteManifestUrl.pathname.split("/").filter(Boolean).slice(0, -1).join("/")}`;
		if (!remoteManifestUrl.pathname.endsWith("/")) remoteManifestUrl.pathname += "/";
		remoteManifestUrl.search = "";
		remoteManifestUrl.hash = "";
		return new URL(REMOTE_HMR_ENDPOINT, remoteManifestUrl).toString();
	} catch {
		return null;
	}
}
function parseRemoteHmrMessage(rawData) {
	if (typeof rawData !== "string") return null;
	try {
		const parsed = JSON.parse(rawData);
		if (parsed?.type !== "custom" || typeof parsed?.event !== "string") return null;
		return parsed;
	} catch {
		return null;
	}
}
function getStringPreview(value, max = 180) {
	let rawValue = "";
	if (typeof value === "string") rawValue = value;
	else if (value instanceof Error) rawValue = `${value.name}: ${value.message}`;
	else if (typeof value === "object" && value !== null) try {
		rawValue = JSON.stringify(value);
	} catch {}
	return rawValue.slice(0, max);
}
function isRemoteHmrEnabled(dev) {
	return typeof dev === "object" && dev !== null && !!dev.remoteHmr;
}
/**
* Detects whether the Vite plugin pipeline includes a framework with
* cross-federation HMR support (a shared runtime proxy that works
* across module federation boundaries).
*
* Currently only React is supported via the shared /@react-refresh proxy.
*/
function hasCrossFederationHmr(plugins) {
	const supportedPlugins = ["vite:react-refresh", "vite:react-swc:refresh"];
	return plugins.some((p) => supportedPlugins.includes(p.name));
}
function resolveHmrStrategy(dev, plugins) {
	if (typeof dev === "object" && dev !== null && dev.remoteHmr === "full-reload") return "full-reload";
	return hasCrossFederationHmr(plugins) ? "native" : "full-reload";
}
function pluginDevRemoteHmr(options) {
	return {
		name: "module-federation-dev-remote-hmr",
		apply: "serve",
		configureServer(server) {
			if (!isRemoteHmrEnabled(options.dev)) return;
			const isRemote = Object.keys(options.exposes).length > 0;
			const isHost = Object.keys(options.remotes).length > 0;
			const strategy = resolveHmrStrategy(options.dev, server.config.plugins);
			if (isRemote) {
				const endpointPath = getRemoteHmrPath(server.config.base);
				const wsUrl = getRemoteHmrWsUrl(server);
				server.middlewares.use((req, res, next) => {
					if (req.url?.replace(/\?.*$/, "") !== "/@react-refresh") return next();
					res.setHeader("Content-Type", "application/javascript; charset=utf-8");
					res.setHeader("Access-Control-Allow-Origin", "*");
					res.end(REACT_REFRESH_PROXY_MODULE);
				});
				server.middlewares.use((req, res, next) => {
					if (req.url?.replace(/\?.*/, "") !== endpointPath) {
						next();
						return;
					}
					res.setHeader("Content-Type", "application/json");
					res.setHeader("Access-Control-Allow-Origin", "*");
					res.end(JSON.stringify({
						remote: options.name,
						event: REMOTE_HMR_EVENT,
						wsUrl
					}));
				});
				const broadcast = (file) => {
					if (strategy === "native") return;
					if (shouldIgnoreFile(file, options)) return;
					server.ws.send({
						type: "custom",
						event: REMOTE_HMR_EVENT,
						data: {
							remote: options.name,
							file,
							ts: Date.now()
						}
					});
				};
				server.watcher.on("change", broadcast);
				server.watcher.on("add", broadcast);
				server.watcher.on("unlink", broadcast);
				server.httpServer?.once("close", () => {
					server.watcher.off("change", broadcast);
					server.watcher.off("add", broadcast);
					server.watcher.off("unlink", broadcast);
				});
			}
			if (isHost) {
				const connections = [];
				const reconnectTimers = /* @__PURE__ */ new Map();
				let isTearingDown = false;
				const clearReconnectTimer = (remoteName) => {
					const timer = reconnectTimers.get(remoteName);
					if (!timer) return;
					clearTimeout(timer);
					reconnectTimers.delete(remoteName);
				};
				const scheduleReconnect = (remoteName, remote, attempt, reason) => {
					if (isTearingDown) return;
					if (attempt >= REMOTE_HMR_CONNECT_MAX_RETRIES) {
						mfWarn(`Remote "${remoteName}" full HMR reconnect skipped after ${REMOTE_HMR_CONNECT_MAX_RETRIES} attempts: ${reason}`);
						return;
					}
					clearReconnectTimer(remoteName);
					const timer = setTimeout(() => {
						reconnectTimers.delete(remoteName);
						connectRemote(remoteName, remote, attempt + 1);
					}, REMOTE_HMR_CONNECT_RETRY_DELAY_MS);
					reconnectTimers.set(remoteName, timer);
				};
				const connectRemote = async (remoteName, remote, attempt = 0) => {
					if (isTearingDown) return;
					const endpoint = getRemoteHmrEndpoint(remote.entry, server);
					if (!endpoint) {
						mfWarn(`Failed to build HMR endpoint URL for remote "${remoteName}"`);
						return;
					}
					try {
						const metadataResponse = await fetch(endpoint);
						if (!metadataResponse.ok) {
							mfWarn(`Failed to fetch remote HMR metadata from "${remoteName}": ${metadataResponse.status}`);
							scheduleReconnect(remoteName, remote, attempt, `HTTP ${metadataResponse.status}`);
							return;
						}
						const metadata = await metadataResponse.json();
						if (metadata.event !== REMOTE_HMR_EVENT || !metadata.wsUrl) {
							mfWarn(`Remote "${remoteName}" returned unexpected HMR metadata shape`);
							return;
						}
						const ws = new WebSocket(metadata.wsUrl, "vite-hmr");
						ws.onmessage = (rawEvent) => {
							if (strategy === "native") return;
							const message = parseRemoteHmrMessage(rawEvent.data);
							if (!message || message.event !== REMOTE_HMR_EVENT) return;
							server.ws.send({ type: "full-reload" });
						};
						ws.onopen = () => clearReconnectTimer(remoteName);
						ws.onerror = (error) => mfWarn(`Remote HMR socket error for "${remoteName}":`, error);
						ws.onclose = () => scheduleReconnect(remoteName, remote, attempt, "socket closed");
						connections.push(ws);
					} catch (error) {
						mfWarn(`Failed to connect remote HMR for "${remoteName}" on attempt ${attempt + 1}: ${getStringPreview(error)}`);
						scheduleReconnect(remoteName, remote, attempt, getStringPreview(error));
					}
				};
				const teardown = () => {
					isTearingDown = true;
					reconnectTimers.forEach((timer) => clearTimeout(timer));
					reconnectTimers.clear();
					connections.forEach((connection) => {
						if (connection.readyState !== connection.CLOSING && connection.readyState !== connection.CLOSED) connection.close();
					});
					connections.length = 0;
				};
				for (const [remoteName, remote] of Object.entries(options.remotes)) connectRemote(remoteName, remote);
				const triggerHostReload = (file) => {
					if (strategy === "native") return;
					if (shouldIgnoreFile(file, options)) return;
					server.ws.send({ type: "full-reload" });
				};
				server.watcher.on("change", triggerHostReload);
				server.watcher.on("add", triggerHostReload);
				server.watcher.on("unlink", triggerHostReload);
				server.httpServer?.once("close", teardown);
			}
		}
	};
}
//#endregion
//#region src/plugins/pluginDts.ts
const DEFAULT_DEV_OPTIONS = {
	disableLiveReload: true,
	disableHotTypesReload: false,
	disableDynamicRemoteTypeHints: false
};
const DYNAMIC_HINTS_PLUGIN = "@module-federation/dts-plugin/dynamic-remote-type-hints-plugin";
const getIPv4 = () => process.env["FEDERATION_IPV4"] || "127.0.0.1";
const DEV_TYPES_FOLDER = ".dev-server";
const DEFAULT_PUBLIC_TYPES_FOLDER = "@mf-types";
const forkDevWorkerPath = __require.resolve("@module-federation/dts-plugin/dist/fork-dev-worker.js");
var DevWorker = class {
	worker = rpc.createRpcWorker(forkDevWorkerPath, {}, void 0, false);
	constructor(options) {
		this.worker.connect(options);
	}
	update() {
		this.worker.process?.send?.({
			type: rpc.RpcGMCallTypes.CALL,
			id: this.worker.id,
			args: [void 0, "update"]
		});
	}
	exit() {
		this.worker.terminate();
	}
};
const normalizeDevOptions = (dev) => {
	if (dev === false) return false;
	if (dev === true || typeof dev === "undefined") return { ...DEFAULT_DEV_OPTIONS };
	return {
		...DEFAULT_DEV_OPTIONS,
		...dev
	};
};
const buildDtsModuleFederationConfig = (options) => {
	const exposes = {};
	Object.entries(options.exposes).forEach(([key, value]) => {
		if (value.import) exposes[key] = value.import;
	});
	const remotes = {};
	Object.entries(options.remotes).forEach(([key, remote]) => {
		if (!remote.entry) return;
		remotes[key] = `${remote.entryGlobalName?.startsWith("http") || remote.entryGlobalName?.includes(".json") ? remote.name || key : remote.entryGlobalName || remote.name || key}@${remote.entry}`;
	});
	return {
		...options,
		exposes,
		remotes
	};
};
const resolveOutputDir = (config) => {
	const { outDir } = config.build;
	if (path$1.isAbsolute(outDir)) return path$1.relative(config.root, outDir);
	return outDir;
};
const ensureRuntimePlugin = (options, pluginId) => {
	if (!options.runtimePlugins.some((plugin) => {
		if (typeof plugin === "string") return plugin === pluginId;
		return plugin[0] === pluginId;
	})) options.runtimePlugins.push(pluginId);
};
const getExposeImportPaths = (options) => {
	return Object.values(options.exposes).map((value) => {
		return value.import;
	}).filter((value) => Boolean(value));
};
const usesVueSfcExposes = (options) => {
	return getExposeImportPaths(options).some((value) => value.endsWith(".vue"));
};
const resolveDtsPluginOptions = (dts, options, context) => {
	if (dts === false) return false;
	const inferredGenerateTypesDefaults = { generateAPITypes: true };
	if (usesVueSfcExposes(options) && hasPackageDependency("vue-tsc", context)) inferredGenerateTypesDefaults.compilerInstance = "vue-tsc";
	if (dts === true || typeof dts === "undefined") return { generateTypes: inferredGenerateTypesDefaults };
	const generateTypes = dts.generateTypes;
	return {
		...dts,
		generateTypes: generateTypes === false ? false : {
			...inferredGenerateTypesDefaults,
			...generateTypes === true || typeof generateTypes === "undefined" ? {} : generateTypes
		}
	};
};
const getBasePath = (base) => {
	if (base.startsWith("http://") || base.startsWith("https://")) return new URL(base).pathname.replace(/\/$/, "") || "/";
	return base.replace(/\/$/, "") || "/";
};
const joinBaseAndAsset = (base, assetFileName) => {
	const basePath = getBasePath(base);
	return `${basePath === "/" ? "" : basePath}/${assetFileName}`.replace(/\/{2,}/g, "/");
};
const getDevDtsAssetPaths = (options) => {
	const { outputDir, publicTypesFolder, root, base } = options;
	return {
		apiFilePath: path$1.resolve(root, outputDir, `${DEV_TYPES_FOLDER}.d.ts`),
		apiRequestPath: joinBaseAndAsset(base, `${publicTypesFolder}.d.ts`),
		zipFilePath: path$1.resolve(root, outputDir, `${DEV_TYPES_FOLDER}.zip`),
		zipRequestPath: joinBaseAndAsset(base, `${publicTypesFolder}.zip`)
	};
};
const createDevDtsAssetMiddleware = (assetPaths) => {
	return (req, res, next) => {
		const requestPath = req.url?.split("?")[0];
		const isZipRequest = requestPath === assetPaths.zipRequestPath;
		const isApiRequest = requestPath === assetPaths.apiRequestPath;
		if (!isZipRequest && !isApiRequest) {
			next();
			return;
		}
		const filePath = isZipRequest ? assetPaths.zipFilePath : assetPaths.apiFilePath;
		if (!fs.existsSync(filePath)) {
			res.statusCode = 404;
			res.end();
			return;
		}
		res.statusCode = 200;
		res.setHeader("Content-Type", isZipRequest ? "application/x-gzip" : "application/typescript");
		if (req.method === "HEAD") {
			res.end();
			return;
		}
		const stream = fs.createReadStream(filePath);
		stream.on("error", () => {
			if (!res.headersSent) res.statusCode = 500;
			res.end();
		});
		res.on("close", () => {
			stream.destroy();
		});
		stream.pipe(res);
	};
};
const normalizeDevDtsOptions = (dts, context) => {
	return normalizeOptions(isTSProject(dts, context), {
		generateTypes: { compileInChildProcess: true },
		consumeTypes: { consumeAPITypes: true },
		extraOptions: {},
		displayErrorInTerminal: typeof dts === "object" && dts ? dts.displayErrorInTerminal : void 0
	}, "mfOptions.dts")(dts);
};
const logDtsError = (error, dtsOptions) => {
	if (dtsOptions === false) return;
	if (typeof dtsOptions === "object" && dtsOptions && dtsOptions.displayErrorInTerminal === false) return;
	mfError(error);
};
function pluginDts(options) {
	if (options.dts === false) return [];
	const baseDtsModuleFederationConfig = buildDtsModuleFederationConfig(options);
	const getDtsModuleFederationConfig = (context) => ({
		...baseDtsModuleFederationConfig,
		dts: resolveDtsPluginOptions(options.dts, options, context)
	});
	let resolvedConfig;
	let devWorker;
	let normalizedDevOptions;
	let hasGeneratedBundle = false;
	return [{
		name: "module-federation-dts-dev",
		apply: "serve",
		config(config) {
			normalizedDevOptions = normalizeDevOptions(options.dev);
			if (!normalizedDevOptions) return;
			if (normalizedDevOptions.disableDynamicRemoteTypeHints) return;
			ensureRuntimePlugin(options, DYNAMIC_HINTS_PLUGIN);
			const define = config.define ? { ...config.define } : {};
			if (!("FEDERATION_IPV4" in define)) define.FEDERATION_IPV4 = JSON.stringify(getIPv4());
			config.define = define;
		},
		configResolved(config) {
			resolvedConfig = config;
		},
		configureServer(server) {
			if (!normalizedDevOptions || !resolvedConfig) return;
			const devOptions = normalizedDevOptions;
			if (devOptions.disableDynamicRemoteTypeHints && devOptions.disableHotTypesReload && devOptions.disableLiveReload) return;
			if (!options.name) throw createModuleFederationError("name is required if you want to enable dev server!");
			const outputDir = resolveOutputDir(resolvedConfig);
			const dtsModuleFederationConfig = getDtsModuleFederationConfig(resolvedConfig.root);
			const normalizedDtsOptions = normalizeDevDtsOptions(dtsModuleFederationConfig.dts, resolvedConfig.root);
			if (typeof normalizedDtsOptions !== "object") return;
			const normalizedGenerateTypes = normalizeOptions(Boolean(normalizedDtsOptions), { compileInChildProcess: true }, "mfOptions.dts.generateTypes")(normalizedDtsOptions.generateTypes);
			const remote = normalizedGenerateTypes === false ? void 0 : {
				implementation: normalizedDtsOptions.implementation,
				context: resolvedConfig.root,
				outputDir,
				moduleFederationConfig: { ...dtsModuleFederationConfig },
				hostRemoteTypesFolder: normalizedGenerateTypes.typesFolder || DEFAULT_PUBLIC_TYPES_FOLDER,
				...normalizedGenerateTypes,
				typesFolder: DEV_TYPES_FOLDER
			};
			if (remote) server.middlewares.use(createDevDtsAssetMiddleware(getDevDtsAssetPaths({
				outputDir,
				publicTypesFolder: remote.hostRemoteTypesFolder || DEFAULT_PUBLIC_TYPES_FOLDER,
				root: resolvedConfig.root,
				base: resolvedConfig.base
			})));
			if (remote && !remote.tsConfigPath && normalizedDtsOptions.tsConfigPath) remote.tsConfigPath = normalizedDtsOptions.tsConfigPath;
			const normalizedConsumeTypes = normalizeOptions(Boolean(normalizedDtsOptions), { consumeAPITypes: true }, "mfOptions.dts.consumeTypes")(normalizedDtsOptions.consumeTypes);
			const host = normalizedConsumeTypes === false ? void 0 : {
				implementation: normalizedDtsOptions.implementation,
				context: resolvedConfig.root,
				moduleFederationConfig: dtsModuleFederationConfig,
				typesFolder: normalizedConsumeTypes.typesFolder || "@mf-types",
				abortOnError: false,
				...normalizedConsumeTypes
			};
			const extraOptions = normalizedDtsOptions.extraOptions || {};
			if (!remote && !host && devOptions.disableLiveReload) return;
			const startDevWorker = async () => {
				let remoteTypeUrls;
				if (host) remoteTypeUrls = await new Promise((resolve) => {
					consumeTypesAPI({
						host,
						extraOptions,
						displayErrorInTerminal: normalizedDtsOptions.displayErrorInTerminal
					}, resolve);
				});
				devWorker = new DevWorker({
					name: options.name,
					remote,
					host: host ? {
						...host,
						remoteTypeUrls
					} : void 0,
					extraOptions,
					disableLiveReload: devOptions.disableLiveReload,
					disableHotTypesReload: devOptions.disableHotTypesReload
				});
				const update = () => devWorker?.update();
				server.watcher.on("change", update);
				server.watcher.on("add", update);
				server.watcher.on("unlink", update);
				server.httpServer?.once("close", () => {
					devWorker?.exit();
					server.watcher.off("change", update);
					server.watcher.off("add", update);
					server.watcher.off("unlink", update);
				});
			};
			startDevWorker().catch((error) => {
				logDtsError(error, normalizedDtsOptions);
			});
		}
	}, {
		name: "module-federation-dts-build",
		apply: "build",
		configResolved(config) {
			resolvedConfig = config;
		},
		async generateBundle() {
			if (hasGeneratedBundle) return;
			hasGeneratedBundle = true;
			if (!resolvedConfig) return;
			let normalizedDtsOptions;
			try {
				normalizedDtsOptions = normalizeDtsOptions(getDtsModuleFederationConfig(resolvedConfig.root), resolvedConfig.root);
			} catch (error) {
				logDtsError(error, options.dts);
				return;
			}
			if (typeof normalizedDtsOptions !== "object") return;
			const context = resolvedConfig.root;
			const outputDir = resolveOutputDir(resolvedConfig);
			let consumeOptions;
			try {
				consumeOptions = normalizeConsumeTypesOptions({
					context,
					dtsOptions: normalizedDtsOptions,
					pluginOptions: getDtsModuleFederationConfig(resolvedConfig.root)
				});
			} catch (error) {
				logDtsError(error, normalizedDtsOptions);
				return;
			}
			if (consumeOptions?.host?.typesOnBuild) try {
				await consumeTypesAPI(consumeOptions);
			} catch (error) {
				logDtsError(error, normalizedDtsOptions);
			}
			let generateOptions;
			try {
				generateOptions = normalizeGenerateTypesOptions({
					context,
					outputDir,
					dtsOptions: normalizedDtsOptions,
					pluginOptions: getDtsModuleFederationConfig(resolvedConfig.root)
				});
			} catch (error) {
				logDtsError(error, normalizedDtsOptions);
				return;
			}
			if (!generateOptions) return;
			try {
				await generateTypesAPI({ dtsManagerOptions: generateOptions });
			} catch (error) {
				logDtsError(error, normalizedDtsOptions);
			}
		}
	}];
}
//#endregion
//#region src/virtualModules/index.ts
function initVirtualModules(command, remoteEntryId) {
	writeLocalSharedImportMap();
	writeHostAutoInit(remoteEntryId, command);
	writeRuntimeInitStatus(command);
}
//#endregion
//#region src/utils/bundleHelpers.ts
function isOutputChunk$1(chunk) {
	return chunk.type === "chunk";
}
function escapeRegExp(value) {
	return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function getProxyBaseName(fileName) {
	return fileName.replace(/^.*\//, "").replace(/\.js$/, "").replace(/-[A-Za-z0-9_-]+$/, "");
}
function extractFunctionDeclaration(code, functionName) {
	const funcRe = new RegExp(`function\\s+${functionName}\\s*\\([^)]*\\)\\s*\\{`);
	const funcStart = code.search(funcRe);
	if (funcStart < 0) return;
	let depth = 0;
	for (let i = code.indexOf("{", funcStart); i < code.length; i++) if (code[i] === "{") depth++;
	else if (code[i] === "}") {
		depth--;
		if (depth === 0) return code.slice(funcStart, i + 1);
	}
}
/**
* Resolve the local alias for a non-inlineable proxy binding.
* If Rollup's deconflict renamed the alias but didn't update references
* in the code body, fall back to proxyLocal so they stay in sync.
*/
function resolveProxyAlias(binding, proxyLocal, code, fullImport, claimedLocals = /* @__PURE__ */ new Set()) {
	const codeWithoutImport = code.replace(fullImport, "");
	const escapedLocal = binding.local.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
	const localUsedInCode = new RegExp(`\\b${escapedLocal}\\b`).test(codeWithoutImport);
	const claimedImportLocals = /* @__PURE__ */ new Set();
	const importRe = /import\s*\{([^}]+)\}\s*from\s*["'][^"']+["']\s*;?/g;
	let match;
	while ((match = importRe.exec(codeWithoutImport)) !== null) for (const spec of match[1].split(",")) {
		const parts = spec.trim().split(/\s+as\s+/);
		claimedImportLocals.add((parts[1] || parts[0]).trim());
	}
	const local = !localUsedInCode && !claimedLocals.has(proxyLocal) && !claimedImportLocals.has(proxyLocal) ? proxyLocal : binding.local;
	return {
		imported: binding.imported,
		local
	};
}
function collectLoadShareProxyChunks(bundle, loadShareTag) {
	const proxyChunks = /* @__PURE__ */ new Map();
	for (const [fileName, chunk] of Object.entries(bundle)) {
		if (!isOutputChunk$1(chunk)) continue;
		if (fileName.includes(loadShareTag) && fileName.includes("commonjs-proxy")) proxyChunks.set(fileName, {
			code: chunk.code,
			fileName
		});
	}
	return proxyChunks;
}
function collectSystemProxyInfos(proxyChunks, loadShareTag) {
	const systemProxyInfo = /* @__PURE__ */ new Map();
	for (const [proxyFileName, proxyInfo] of Array.from(proxyChunks.entries())) {
		const depsMatch = proxyInfo.code.match(/System\.register\(\[([\s\S]*?)\]/);
		if (!depsMatch) continue;
		const loadShareDep = Array.from(depsMatch[1].matchAll(/["']([^"']+)["']/g)).map((m) => m[1]).find((dep) => dep.includes(loadShareTag) && !dep.includes("commonjs-proxy"));
		if (!loadShareDep) continue;
		const loadShareBindings = {};
		for (const m of proxyInfo.code.matchAll(/([A-Za-z_$][\w$]*)\s*=\s*module\d+\.([A-Za-z_$][\w$]*)/g)) loadShareBindings[m[1]] = m[2];
		const exportMap = {};
		const objectExportMatch = proxyInfo.code.match(/exports\(\s*\{([\s\S]*?)\}\s*\)/);
		if (objectExportMatch) for (const m of objectExportMatch[1].matchAll(/([A-Za-z_$][\w$]*)\s*:\s*([A-Za-z_$][\w$]*)/g)) {
			const [, exported, local] = m;
			const funcBody = extractFunctionDeclaration(proxyInfo.code, local);
			if (funcBody) exportMap[exported] = {
				type: "helper",
				code: funcBody
			};
		}
		for (const m of proxyInfo.code.matchAll(/exports\(\s*["']([^"']+)["']\s*,([\s\S]*?)\);/g)) {
			const exported = m[1];
			const expression = m[2];
			for (const [local, exportName] of Object.entries(loadShareBindings)) if (new RegExp(`\\b${local}\\b`).test(expression)) {
				exportMap[exported] = {
					type: "reexport",
					exportName
				};
				break;
			}
		}
		if (Object.keys(exportMap).length > 0) systemProxyInfo.set(proxyFileName, {
			loadShareDep,
			exportMap
		});
	}
	return systemProxyInfo;
}
function rewriteEsmProxyConsumers(code, proxyChunks) {
	let nextCode = code;
	const claimedLocals = /* @__PURE__ */ new Set();
	for (const [proxyFileName, proxyInfo] of Array.from(proxyChunks.entries())) {
		const proxyBaseName = getProxyBaseName(proxyFileName);
		const importMatch = new RegExp(`import\\s*\\{([^}]+)\\}\\s*from\\s*["']([^"']*${escapeRegExp(proxyBaseName)}[^"']*)["']\\s*;?`).exec(nextCode);
		if (!importMatch) continue;
		const fullImport = importMatch[0];
		const bindings = importMatch[1].split(",").map((s) => {
			const parts = s.trim().split(/\s+as\s+/);
			return {
				imported: parts[0].trim(),
				local: (parts[1] || parts[0]).trim()
			};
		});
		const exportMapMatch = proxyInfo.code.match(/export\s*\{([^}]+)\}/);
		if (!exportMapMatch) continue;
		const exportMap = {};
		for (const entry of exportMapMatch[1].split(",")) {
			const parts = entry.trim().split(/\s+as\s+/);
			if (parts.length === 2) exportMap[parts[1].trim()] = parts[0].trim();
		}
		const inlineable = [];
		const nonInlineable = [];
		const pendingLocals = new Set(bindings.map((binding) => binding.local));
		for (const b of bindings) {
			pendingLocals.delete(b.local);
			const proxyLocal = exportMap[b.imported];
			if (!proxyLocal) {
				claimedLocals.add(b.local);
				nonInlineable.push(b);
				continue;
			}
			const funcBody = extractFunctionDeclaration(proxyInfo.code, proxyLocal);
			if (funcBody) {
				inlineable.push({
					local: b.local,
					funcBody: funcBody.replace(new RegExp(`function\\s+${proxyLocal}\\s*\\(`), `function ${b.local}(`)
				});
				claimedLocals.add(b.local);
			} else {
				const unavailableLocals = new Set(claimedLocals);
				pendingLocals.forEach((local) => unavailableLocals.add(local));
				const resolvedBinding = resolveProxyAlias(b, proxyLocal, nextCode, fullImport, unavailableLocals);
				claimedLocals.add(resolvedBinding.local);
				nonInlineable.push(resolvedBinding);
			}
		}
		const hasRenamedAlias = nonInlineable.some((b) => bindings.find((ob) => ob.imported === b.imported)?.local !== b.local);
		if (inlineable.length === 0 && !hasRenamedAlias) continue;
		let replacement = "";
		if (nonInlineable.length > 0) replacement = `import{${nonInlineable.map((b) => b.imported === b.local ? b.imported : `${b.imported} as ${b.local}`).join(",")}}from"${importMatch[2]}";`;
		replacement += inlineable.map((f) => f.funcBody).join("");
		nextCode = nextCode.replace(fullImport, () => replacement);
	}
	return nextCode;
}
function rewriteSystemProxyConsumers(code, systemProxyInfo) {
	if (!code.includes("System.register(")) return code;
	let nextCode = code;
	for (const [proxyFileName, proxyInfo] of Array.from(systemProxyInfo.entries())) {
		const proxyBaseName = getProxyBaseName(proxyFileName);
		const depMatch = new RegExp(`["']([^"']*${escapeRegExp(proxyBaseName)}[^"']*)["']`).exec(nextCode);
		if (!depMatch) continue;
		let setterIndex = 0;
		const depListMatch = nextCode.match(/System\.register\(\[([\s\S]*?)\]/);
		if (depListMatch) setterIndex = Array.from(depListMatch[1].matchAll(/["']([^"']+)["']/g)).map((m) => m[1]).findIndex((dep) => dep.includes(proxyBaseName));
		if (setterIndex < 0) continue;
		const settersStart = nextCode.indexOf("setters: [");
		if (settersStart < 0) continue;
		const setterMatch = Array.from(nextCode.slice(settersStart).matchAll(/\((module\d+)\)\s*=>\s*\{([\s\S]*?)\}/g))[setterIndex];
		if (!setterMatch) continue;
		const [fullSetter, moduleLocal, setterBody] = setterMatch;
		const helpersToInline = [];
		const nextSetterBody = setterBody.replace(new RegExp(`([A-Za-z_$][\\w$]*)\\s*=\\s*${moduleLocal}\\.([A-Za-z_$][\\w$]*);?`, "g"), (assignment, local, imported) => {
			const mapped = proxyInfo.exportMap[imported];
			if (!mapped) return assignment;
			if (mapped.type === "helper") {
				helpersToInline.push(mapped.code.replace(new RegExp(`function\\s+${imported}\\s*\\(`), `function ${local}(`));
				return "";
			}
			return `${local} = ${moduleLocal}.${mapped.exportName};`;
		});
		if (nextSetterBody === setterBody && helpersToInline.length === 0) continue;
		const nextSetter = fullSetter.replace(setterBody, () => nextSetterBody);
		nextCode = nextCode.replace(fullSetter, () => nextSetter);
		nextCode = nextCode.replace(depMatch[0], JSON.stringify(proxyInfo.loadShareDep));
		if (helpersToInline.length > 0) nextCode = nextCode.replace("execute: (function() {", () => {
			return `execute: (function() {${helpersToInline.join("")}`;
		});
	}
	return nextCode;
}
function findRemoteEntryFile(filename, bundle) {
	for (const [_, fileData] of Object.entries(bundle)) if (filename.replace(/[\[\]]/g, "_").replace(/\.[^/.]+$/, "") === fileData.name || fileData.name === "remoteEntry") return fileData.fileName;
}
//#endregion
//#region src/utils/cssModuleHelpers.ts
const ASSET_TYPES = ["js", "css"];
const LOAD_TIMINGS = ["sync", "async"];
const JS_EXTENSIONS = [
	".ts",
	".tsx",
	".jsx",
	".mjs",
	".cjs"
];
/**
* Creates an empty asset map structure for tracking JS and CSS assets
* @returns Initialized asset map with sync/async arrays for JS and CSS
*/
const createEmptyAssetMap = () => ({
	js: {
		sync: [],
		async: []
	},
	css: {
		sync: [],
		async: []
	}
});
/**
* Tracks an asset in the preload map with deduplication
* @param map - The preload map to update
* @param key - The module key to track under
* @param fileName - The asset filename to track
* @param isAsync - Whether the asset is loaded async
* @param type - The asset type ('js' or 'css')
*/
const trackAsset = (map, key, fileName, isAsync, type) => {
	if (!map[key]) map[key] = createEmptyAssetMap();
	const target = isAsync ? map[key][type].async : map[key][type].sync;
	if (!target.includes(fileName)) target.push(fileName);
};
/**
* Checks if a file is a CSS file by extension
* @param fileName - The filename to check
* @returns True if file has a CSS extension (.css, .scss, .less)
*/
const isCSSFile = (fileName) => {
	return fileName.endsWith(".css") || fileName.endsWith(".scss") || fileName.endsWith(".less");
};
/**
* Collects all CSS assets from the bundle
* @param bundle - The Rollup output bundle
* @returns Set of CSS asset filenames
*/
const collectCssAssets = (bundle) => {
	const cssAssets = /* @__PURE__ */ new Set();
	for (const [fileName, fileData] of Object.entries(bundle)) if (fileData.type === "asset" && isCSSFile(fileName)) cssAssets.add(fileName);
	return cssAssets;
};
/**
* Checks if a chunk contains CSS modules (e.g. .css, .vanilla.css, .scss, .less)
* by scanning its module list
*/
const chunkContainsCssModules = (modules) => {
	for (const modulePath of Object.keys(modules)) if (isCSSFile(modulePath)) return true;
	return false;
};
/**
* Processes module assets and tracks them in the files map
* @param bundle - The Rollup output bundle
* @param filesMap - The preload map to populate
* @param moduleMatcher - Function that matches module paths to keys
*/
const processModuleAssets = (bundle, filesMap, moduleMatcher, options = {}) => {
	const bundleCssAssets = collectCssAssets(bundle);
	for (const [fileName, fileData] of Object.entries(bundle)) {
		if (fileData.type !== "chunk") continue;
		if (!fileData.modules) continue;
		for (const modulePath of Object.keys(fileData.modules)) {
			const comparableModulePath = options.root ? path.resolve(options.root, modulePath) : modulePath;
			const comparableModulePaths = [comparableModulePath];
			if (options.stripKnownJsExtensions) {
				const ext = path.extname(comparableModulePath);
				if (JS_EXTENSIONS.includes(ext)) comparableModulePaths.push(path.join(path.dirname(comparableModulePath), path.basename(comparableModulePath, ext)));
			}
			const matchKey = comparableModulePaths.map(moduleMatcher).find(Boolean);
			if (!matchKey) continue;
			trackAsset(filesMap, matchKey, fileName, false, "js");
			let foundCssViaMetadata = false;
			if (fileData.viteMetadata?.importedCss?.size) for (const cssFile of Array.from(fileData.viteMetadata.importedCss)) {
				trackAsset(filesMap, matchKey, cssFile, false, "css");
				foundCssViaMetadata = true;
			}
			if (!foundCssViaMetadata && chunkContainsCssModules(fileData.modules)) for (const cssAsset of Array.from(bundleCssAssets)) trackAsset(filesMap, matchKey, cssAsset, false, "css");
			if (fileData.dynamicImports) for (const dynamicImport of fileData.dynamicImports) {
				if (!bundle[dynamicImport]) continue;
				trackAsset(filesMap, matchKey, dynamicImport, true, isCSSFile(dynamicImport) ? "css" : "js");
			}
		}
	}
};
/**
* Adds global CSS assets to all module exports
* @param filesMap - The preload map to update
* @param cssAssets - Set of CSS asset filenames to add
*/
const addCssAssetsToAllExports = (filesMap, cssAssets) => {
	Object.keys(filesMap).forEach((key) => {
		cssAssets.forEach((cssAsset) => {
			trackAsset(filesMap, key, cssAsset, false, "css");
		});
	});
};
/**
* Deduplicates assets in the files map
* @param filesMap - The preload map to deduplicate
* @returns New deduplicated preload map
*/
const deduplicateAssets = (filesMap) => {
	const result = {};
	for (const [key, assetMaps] of Object.entries(filesMap)) {
		result[key] = createEmptyAssetMap();
		for (const type of ASSET_TYPES) for (const timing of LOAD_TIMINGS) result[key][type][timing] = Array.from(new Set(assetMaps[type][timing]));
	}
	return result;
};
/**
* Builds a mapping between module files and their share keys
* @param shareKeys - Set of share keys to map
* @param resolveFn - Function to resolve module paths
* @returns Map of file paths to their corresponding share keys
*/
const buildFileToShareKeyMap = async (shareKeys, resolveFn) => {
	const fileToShareKey = /* @__PURE__ */ new Map();
	const resolutions = await Promise.all(Array.from(shareKeys).map((shareKey) => resolveFn(getPreBuildLibImportId(shareKey)).then((resolution) => ({
		shareKey,
		file: resolution?.id?.split("?")[0]
	})).catch(() => null)));
	for (const resolution of resolutions) if (resolution?.file) fileToShareKey.set(resolution.file, resolution.shareKey);
	return fileToShareKey;
};
//#endregion
//#region src/utils/pathNormalization.ts
const COMMON_SHARED_SUBPATHS = {
	react: ["react/jsx-runtime", "react/jsx-dev-runtime"],
	"react-dom": [
		"react-dom/client",
		"react-dom/server",
		"react-dom/server.browser"
	],
	"solid-js": [
		"solid-js/web",
		"solid-js/store",
		"solid-js/html",
		"solid-js/h"
	]
};
function removeTrailingSlash(value) {
	return value.endsWith("/") ? value.slice(0, -1) : value;
}
function ensureTrailingSlash(value) {
	return `${removeTrailingSlash(value)}/`;
}
function normalizeNodeModulePath(source) {
	return source.replace(/\\/g, "/").replace(/\?.*$/, "");
}
function isNodeModulePath(source) {
	return source.includes("/node_modules/") || source.includes("\\node_modules\\");
}
function filterId(id) {
	return typeof id === "string" && !id.includes("\0");
}
function getMatchingNodeModuleSubpath(source, candidates) {
	const normalized = normalizeNodeModulePath(source);
	return [...candidates].sort((a, b) => b.length - a.length).find((candidate) => normalized.includes(`/node_modules/${candidate}/`) || normalized.includes(`/node_modules/${candidate}.`));
}
function getCommonSharedSubpaths(sharedKey) {
	return COMMON_SHARED_SUBPATHS[removeTrailingSlash(sharedKey)] || [];
}
function getCommonSharedSubpathFromNodeModulePath(source, sharedKey) {
	return getMatchingNodeModuleSubpath(source, getCommonSharedSubpaths(removeTrailingSlash(sharedKey)));
}
/**
* Resolves the public path for remote entries
* @param options - Module Federation options
* @param viteBase - Vite's base config value
* @param originalBase - Original base config before any transformations
* @returns The resolved public path
*/
function resolvePublicPath(options, viteBase, originalBase) {
	if (options.publicPath && options.publicPath !== "auto") return options.publicPath;
	if (originalBase === "") return "auto";
	if (viteBase) return ensureTrailingSlash(viteBase);
	return "auto";
}
//#endregion
//#region src/plugins/pluginMFManifest.ts
/**
* Resolves the build version for the module federation manifest.
*
* Priority:
* 1. `MF_BUILD_VERSION` environment variable (set by CI or manually)
* 2. Falls back to `'1.0.0'` to preserve backward compatibility
*
* This mirrors the behavior of the webpack/rspack plugins via
* `getBuildVersion()` from `@module-federation/managers`.
*/
function getBuildVersion() {
	return process.env["MF_BUILD_VERSION"] ?? "1.0.0";
}
const Manifest = () => {
	const mfOptions = getNormalizeModuleFederationOptions();
	const { name, filename, getPublicPath, manifest: manifestOptions, varFilename } = mfOptions;
	let mfManifestName = manifestOptions === true ? "mf-manifest.json" : typeof manifestOptions === "object" ? path$1.join(manifestOptions?.filePath || "", manifestOptions?.fileName || "mf-manifest.json") : void 0;
	let mfManifestStatsName = mfManifestName ? getStatsFileName(mfManifestName) : void 0;
	const isConsumerProject = Object.keys(mfOptions.exposes).length === 0;
	let disableAssetsAnalyze = false;
	const getDefaultDisableAssetsAnalyze = (command) => command === "serve" && isConsumerProject && (typeof manifestOptions !== "object" || !Object.prototype.hasOwnProperty.call(manifestOptions, "disableAssetsAnalyze"));
	const getConfiguredDisableAssetsAnalyze = (command) => {
		if (typeof manifestOptions === "object" && manifestOptions !== null) {
			if (Object.prototype.hasOwnProperty.call(manifestOptions, "disableAssetsAnalyze")) return manifestOptions.disableAssetsAnalyze === true;
		}
		return getDefaultDisableAssetsAnalyze(command);
	};
	let root;
	let remoteEntryFile;
	let publicPath;
	let _command;
	let _originalConfigBase;
	let viteConfig;
	return [{
		name: "module-federation-manifest",
		apply: "serve",
		configResolved(config) {
			viteConfig = config;
		},
		configureServer(server) {
			server.middlewares.use((req, res, next) => {
				if (!mfManifestName) {
					next();
					return;
				}
				if (req.url?.replace(/\?.*/, "") === (viteConfig.base + mfManifestName).replace(/^\/?/, "/")) {
					res.setHeader("Content-Type", "application/json");
					res.setHeader("Access-Control-Allow-Origin", "*");
					res.end(JSON.stringify({
						...generateMFManifest({}, disableAssetsAnalyze),
						id: name,
						name,
						metaData: {
							name,
							type: "app",
							buildInfo: {
								buildVersion: getBuildVersion(),
								buildName: name
							},
							remoteEntry: {
								name: filename,
								path: "",
								type: "module"
							},
							ssrRemoteEntry: {
								name: filename,
								path: "",
								type: "module"
							},
							varRemoteEntry: varFilename ? {
								name: varFilename,
								path: "",
								type: "var"
							} : void 0,
							types: {
								path: "",
								name: ""
							},
							globalName: name,
							pluginVersion: "0.2.5",
							publicPath
						}
					}));
				} else next();
			});
		}
	}, {
		name: "module-federation-manifest",
		enforce: "post",
		config(config, { command }) {
			_command = command;
			if (!config.build) config.build = {};
			if (!config.build.manifest) config.build.manifest = config.build.manifest || !!mfManifestName;
			disableAssetsAnalyze = getConfiguredDisableAssetsAnalyze(command);
			_originalConfigBase = config.base;
		},
		configResolved(config) {
			root = config.root;
			let base = config.base;
			if (_command === "serve") base = (config.server.origin || "") + config.base;
			publicPath = mfOptions.publicPath === "auto" ? "auto" : resolvePublicPath(mfOptions, base, _originalConfigBase);
		},
		async generateBundle(_options, bundle) {
			if (!mfManifestName) return;
			let filesMap = {};
			const foundRemoteEntryFile = findRemoteEntryFile(mfOptions.filename, bundle);
			if (foundRemoteEntryFile) remoteEntryFile = foundRemoteEntryFile;
			const allCssAssets = mfOptions.bundleAllCSS && !disableAssetsAnalyze ? collectCssAssets(bundle) : /* @__PURE__ */ new Set();
			if (!disableAssetsAnalyze) {
				const exposesModules = Object.keys(mfOptions.exposes).map((item) => mfOptions.exposes[item].import);
				processModuleAssets(bundle, filesMap, (modulePath) => {
					return exposesModules.find((exposeModule) => {
						return modulePath === path$1.resolve(root, exposeModule);
					});
				}, {
					root,
					stripKnownJsExtensions: true
				});
				const fileToShareKey = await buildFileToShareKeyMap(getUsedShares(), this.resolve.bind(this));
				processModuleAssets(bundle, filesMap, (modulePath) => fileToShareKey.get(modulePath));
				if (mfOptions.bundleAllCSS) addCssAssetsToAllExports(filesMap, allCssAssets);
				filesMap = deduplicateAssets(filesMap);
			}
			this.emitFile({
				type: "asset",
				fileName: mfManifestName,
				source: JSON.stringify(generateMFManifest(filesMap, disableAssetsAnalyze))
			});
			if (mfManifestStatsName) this.emitFile({
				type: "asset",
				fileName: mfManifestStatsName,
				source: JSON.stringify(generateMFStats(filesMap, bundle, disableAssetsAnalyze))
			});
		}
	}];
	/**
	* Generates the final manifest JSON structure
	* @param preloadMap - Map of module assets to include
	* @returns Complete manifest object
	*/
	function generateMFManifest(preloadMap, disableAssetsAnalyze = false) {
		const options = getNormalizeModuleFederationOptions();
		const { name, varFilename } = options;
		const remoteEntry = {
			name: remoteEntryFile,
			path: "",
			type: "module"
		};
		const varRemoteEntry = varFilename ? {
			name: varFilename,
			path: "",
			type: "module"
		} : void 0;
		const remotes = Array.from(Object.entries(getUsedRemotesMap())).flatMap(([remoteKey, modules]) => Array.from(modules).map((moduleKey) => ({
			federationContainerName: options.remotes[remoteKey].entry,
			moduleName: moduleKey.replace(remoteKey, "").replace("/", ""),
			alias: remoteKey,
			entry: "*"
		})));
		const shared = Array.from(getUsedShares()).flatMap((shareKey) => {
			const shareItem = getNormalizeShareItem(shareKey);
			if (!shareItem) return [];
			const assets = preloadMap[shareKey] || createEmptyAssetMap();
			return [{
				id: `${name}:${shareKey}`,
				name: shareKey,
				version: shareItem.version,
				singleton: shareItem.shareConfig.singleton,
				requiredVersion: shareItem.shareConfig.requiredVersion,
				assets: {
					js: {
						async: assets.js.async,
						sync: assets.js.sync
					},
					css: {
						async: assets.css.async,
						sync: assets.css.sync
					}
				}
			}];
		});
		const exposes = Object.entries(options.exposes).map(([key, value]) => {
			const formatKey = key.replace("./", "");
			const assets = preloadMap[value.import] || createEmptyAssetMap();
			return {
				id: `${name}:${formatKey}`,
				name: formatKey,
				assets: {
					js: {
						async: assets.js.async,
						sync: assets.js.sync
					},
					css: {
						async: assets.css.async,
						sync: assets.css.sync
					}
				},
				path: key
			};
		});
		return {
			id: name,
			name,
			metaData: {
				name,
				type: "app",
				buildInfo: {
					buildVersion: getBuildVersion(),
					buildName: name
				},
				remoteEntry,
				ssrRemoteEntry: remoteEntry,
				varRemoteEntry,
				types: {
					path: "",
					name: ""
				},
				globalName: name,
				pluginVersion: "0.2.5",
				...!!getPublicPath ? { getPublicPath } : { publicPath }
			},
			...disableAssetsAnalyze ? {} : { shared },
			remotes,
			...disableAssetsAnalyze ? {} : { exposes }
		};
	}
	function generateMFStats(preloadMap, bundle, disableAssetsAnalyze = false) {
		const baseManifest = generateMFManifest(preloadMap, disableAssetsAnalyze);
		const bundleSummary = Object.entries(bundle).map(([fileName, chunkOrAsset]) => ({
			fileName,
			type: chunkOrAsset.type,
			isEntry: chunkOrAsset.isEntry || false,
			size: typeof chunkOrAsset.code === "string" ? chunkOrAsset.code.length : chunkOrAsset.source?.length || void 0
		}));
		return {
			...baseManifest,
			buildOutput: bundleSummary,
			...disableAssetsAnalyze ? {} : { assetAnalysis: preloadMap }
		};
	}
};
function getStatsFileName(manifestFileName) {
	const parsed = path$1.parse(manifestFileName);
	const fileExt = parsed.ext || ".json";
	const baseName = parsed.ext ? parsed.name : parsed.base;
	const fileName = `${baseName === "mf-manifest" ? "mf" : baseName}-stats${fileExt}`;
	return parsed.dir ? path$1.join(parsed.dir, fileName) : fileName;
}
//#endregion
//#region src/plugins/pluginModuleParseEnd.ts
let _resolve = null;
let _parseTimeout = null;
let parsePromise = Promise.resolve(1);
let exposesParseEnd = false;
let expectsExposesParseEnd = false;
let parseStartSet = /* @__PURE__ */ new Set();
let parseEndSet = /* @__PURE__ */ new Set();
function clearParseTimeout() {
	if (_parseTimeout) {
		clearTimeout(_parseTimeout);
		_parseTimeout = null;
	}
}
function resetParseState() {
	clearParseTimeout();
	exposesParseEnd = false;
	expectsExposesParseEnd = false;
	parseStartSet = /* @__PURE__ */ new Set();
	parseEndSet = /* @__PURE__ */ new Set();
	parsePromise = new Promise((resolve) => {
		_resolve = (v) => {
			clearParseTimeout();
			resolve(v);
		};
	});
}
function setParseTimeout(timeout) {
	if (!_parseTimeout) _parseTimeout = setTimeout(() => {
		mfWarn(`Parse timeout (${timeout}s) - forcing resolve`);
		_resolve?.(1);
	}, timeout * 1e3);
}
function resetIdleTimeout(timeout) {
	clearParseTimeout();
	_parseTimeout = setTimeout(() => {
		mfWarn(`moduleParseIdleTimeout: no module activity for ${timeout}s, forcing resolve. Some shared/remote dependencies may be missing. Consider increasing moduleParseIdleTimeout.`);
		_resolve?.(1);
	}, timeout * 1e3);
}
function pluginModuleParseEnd_default(excludeFn, options) {
	const idleTimeout = options.moduleParseIdleTimeout;
	return [
		{
			name: "_",
			apply: "serve",
			config() {
				_resolve?.(1);
			}
		},
		{
			enforce: "pre",
			name: "parseStart",
			apply: "build",
			buildStart() {
				resetParseState();
				if (idleTimeout) resetIdleTimeout(idleTimeout);
				else setParseTimeout(options.moduleParseTimeout);
			},
			load(id) {
				if (excludeFn(id)) return;
				if (id === options.virtualExposesId) expectsExposesParseEnd = true;
				parseStartSet.add(id);
			}
		},
		{
			enforce: "post",
			name: "parseEnd",
			apply: "build",
			moduleParsed(module) {
				const id = module.id;
				if (id === options.virtualExposesId) exposesParseEnd = true;
				if (idleTimeout) resetIdleTimeout(idleTimeout);
				if (excludeFn(id)) return;
				parseEndSet.add(id);
				if (parseStartSet.size === parseEndSet.size && (!expectsExposesParseEnd || exposesParseEnd)) _resolve?.(1);
			}
		}
	];
}
//#endregion
//#region src/plugins/pluginProxyRemoteEntry.ts
function pluginProxyRemoteEntry_default({ options, remoteEntryId, virtualExposesId }) {
	let viteConfig, _command, root;
	return {
		name: "proxyRemoteEntry",
		enforce: "post",
		configResolved(config) {
			viteConfig = config;
			root = config.root;
		},
		config(_config, { command }) {
			_command = command;
		},
		async buildStart() {
			if (_command !== "build") return;
			for (const expose of Object.values(options.exposes)) {
				const resolved = await this.resolve(expose.import);
				if (resolved) this.emitFile({
					type: "chunk",
					id: resolved.id
				});
			}
		},
		async resolveId(id, importer) {
			if (id === remoteEntryId) return remoteEntryId;
			if (id === virtualExposesId) return virtualExposesId;
			if (_command === "serve" && id.includes(getHostAutoInitPath())) return id;
			if (importer === remoteEntryId && !id.startsWith(".") && !id.startsWith("/") && !id.startsWith("\0") && !id.startsWith("virtual:")) {
				const importPath = typeof __filename === "string" ? __filename : fileURLToPath(import.meta.url);
				const resolved = await this.resolve(id, importPath, { skipSelf: true });
				if (resolved) return resolved;
			}
		},
		load(id) {
			if (id === remoteEntryId) return parsePromise.then((_) => generateRemoteEntry(options, virtualExposesId, _command));
			if (id === virtualExposesId) return generateExposes(options);
			if (_command === "serve" && id.includes(getHostAutoInitPath())) return id;
		},
		transform(code, id) {
			return mapCodeToCodeWithSourcemap((() => {
				if (!filterId(id)) return;
				if (id.includes(remoteEntryId)) return parsePromise.then((_) => generateRemoteEntry(options, virtualExposesId, _command));
				if (id === virtualExposesId) return generateExposes(options);
				if (id.includes(getHostAutoInitPath())) {
					if (_command === "serve") {
						const host = typeof viteConfig.server?.host === "string" && viteConfig.server.host !== "0.0.0.0" ? viteConfig.server.host : "localhost";
						const publicPath = JSON.stringify(resolvePublicPath(options, viteConfig.base) + options.filename);
						const fallbackOrigin = `//${host}:${viteConfig.server?.port}`;
						const ssrRemoteEntry = "data:text/javascript," + encodeURIComponent("export async function init(){return {loadRemote:async()=>({}),loadShare:async()=>({})}}");
						return `
          const origin = typeof window !== 'undefined' && (${!options.ignoreOrigin}) ? window.origin : ${JSON.stringify(fallbackOrigin)};
          const remoteEntryImport = typeof window !== 'undefined' ? origin + ${publicPath} : ${JSON.stringify(ssrRemoteEntry)};
          ${generateHostAutoInitCode("remoteEntryImport", "serve")}
        `;
					}
					return code;
				}
			})());
		},
		generateBundle(_, bundle) {
			if (_command !== "build") return;
			const filesMap = {};
			const exposeEntries = Object.entries(options.exposes);
			const allCssAssets = options.bundleAllCSS ? collectCssAssets(bundle) : /* @__PURE__ */ new Set();
			processModuleAssets(bundle, filesMap, (modulePath) => {
				return exposeEntries.find(([_, exposeOptions]) => {
					return modulePath === path$1.resolve(root, exposeOptions.import);
				})?.[1].import;
			}, {
				root,
				stripKnownJsExtensions: true
			});
			if (options.bundleAllCSS) addCssAssetsToAllExports(filesMap, allCssAssets);
			const ensureRelativeImportPath = (fromFile, toFile) => {
				let relativePath = path$1.relative(path$1.dirname(fromFile), toFile);
				if (!relativePath.startsWith(".")) relativePath = `./${relativePath}`;
				return relativePath;
			};
			const placeholderValue = getExposesCssMapPlaceholder();
			const placeholderPatterns = [
				JSON.stringify(placeholderValue),
				`'${placeholderValue}'`,
				`\`${placeholderValue}\``
			];
			for (const file of Object.values(bundle)) {
				if (file.type !== "chunk" || !file.code.includes(placeholderValue)) continue;
				const cssAssetMap = exposeEntries.reduce((acc, [exposeKey, expose]) => {
					const assets = filesMap[expose.import] || createEmptyAssetMap();
					acc[exposeKey] = [...assets.css.sync, ...assets.css.async].map((cssAsset) => ensureRelativeImportPath(file.fileName, cssAsset));
					return acc;
				}, {});
				for (const placeholderPattern of placeholderPatterns) file.code = file.code.replace(placeholderPattern, JSON.stringify(cssAssetMap));
			}
		}
	};
}
//#endregion
//#region src/plugins/pluginProxyRemotes.ts
function isNodeModulesImporter(importer) {
	return importer?.includes("/node_modules/") || importer?.includes("\\node_modules\\");
}
function appendAlias(config, alias) {
	config.resolve ??= {};
	const existingAlias = config.resolve.alias;
	if (!existingAlias) {
		config.resolve.alias = [alias];
		return;
	}
	if (Array.isArray(existingAlias)) {
		existingAlias.push(alias);
		return;
	}
	config.resolve.alias = [...Object.entries(existingAlias).map(([find, replacement]) => ({
		find,
		replacement
	})), alias];
}
function pluginProxyRemotes_default(options) {
	let command;
	let root = process.cwd();
	const { remotes } = options;
	function resolveRemoteId(source, importer, remoteName) {
		if (source === remoteName) {
			const installedPackageEntry = getInstalledPackageEntry(source, { cwd: root });
			if (installedPackageEntry && (importer === void 0 || isNodeModulesImporter(importer))) return installedPackageEntry;
		}
		const remoteModule = getRemoteVirtualModule(source, command);
		addUsedRemote(remoteName, source);
		refreshHostAutoInit();
		return remoteModule.getPath();
	}
	return {
		name: "proxyRemotes",
		enforce: "pre",
		config(config, { command: _command }) {
			command = _command;
			root = config.root || process.cwd();
			Object.keys(remotes).forEach((key) => {
				const remote = remotes[key];
				appendAlias(config, {
					find: new RegExp(`^(${remote.name}(\/.*|$))`),
					replacement: "$1"
				});
			});
		},
		resolveId(source, importer) {
			if (!filterId(source)) return;
			for (const remote of Object.values(remotes)) {
				if (source !== remote.name && !source.startsWith(`${remote.name}/`)) continue;
				return resolveRemoteId(source, importer, remote.name);
			}
		}
	};
}
//#endregion
//#region src/utils/PromiseStore.ts
/**
* example:
* const store = new PromiseStore<number>();
* store.get("example").then((result) => {
*  console.log("Result from example:", result); // 42
* });
* setTimeout(() => {
*  store.set("example", Promise.resolve(42));
* }, 2000);
*/
var PromiseStore = class {
	promiseMap = /* @__PURE__ */ new Map();
	resolveMap = /* @__PURE__ */ new Map();
	set(id, promise) {
		if (this.resolveMap.has(id)) {
			promise.then(this.resolveMap.get(id));
			this.resolveMap.delete(id);
		}
		this.promiseMap.set(id, promise);
	}
	get(id) {
		if (this.promiseMap.has(id)) return this.promiseMap.get(id);
		const pendingPromise = new Promise((resolve) => {
			this.resolveMap.set(id, resolve);
		});
		this.promiseMap.set(id, pendingPromise);
		return pendingPromise;
	}
};
//#endregion
//#region src/plugins/pluginProxySharedModule_preBuild.ts
function getPrebuildResolutionSource(pkgName, shareItem) {
	return getConcreteSharedImportSource(pkgName, shareItem) || pkgName;
}
function tryResolveFromProjectRoot(source) {
	if (path.isAbsolute(source) || source.startsWith(".") || source.startsWith("/")) return source;
	const browserEntry = getInstalledPackageEntry(source, { cwd: getPackageDetectionCwd() });
	if (browserEntry) return browserEntry;
	try {
		return createRequire$1(new URL(`file://${path.join(getPackageDetectionCwd(), "package.json")}`)).resolve(source);
	} catch {
		return;
	}
}
function isBuildConfigImporter(importer) {
	if (!importer) return false;
	return /(^|\/)(?:nuxt|vite|vitest|webpack|rollup|rspack)\.config\.[cm]?[jt]sx?$/.test(importer.replace(/\\/g, "/"));
}
function matchesSharedSource(source, key) {
	const keyBase = key.endsWith("/") ? key.slice(0, -1) : key;
	if (key.endsWith("/")) return source === keyBase || source.startsWith(`${keyBase}/`);
	if (getCommonSharedSubpaths(keyBase).includes(source)) return true;
	return source === keyBase;
}
function findSharedKey(source, shared) {
	const keys = Object.keys(shared || {});
	return keys.find((key) => source === key) ?? keys.find((key) => matchesSharedSource(source, key));
}
function findSharedKeyForSource(source, shared) {
	const key = findSharedKey(source, shared);
	if (key) return key;
	if (isNodeModulePath(source)) {
		const explicitSubpathKey = getMatchingNodeModuleSubpath(source, Object.keys(shared || {}).filter((sharedKey) => sharedKey.includes("/") && !sharedKey.endsWith("/")));
		if (explicitSubpathKey) return explicitSubpathKey;
	}
	const packageName = getPackageNameFromNodeModulePath(source);
	return packageName ? findSharedKey(packageName, shared) : void 0;
}
/**
* Reads the dependencies of an installed package from its package.json.
*/
function getPackageDependencies(pkg) {
	const packageName = getPackageName(pkg);
	const installed = getInstalledPackageJson(packageName, { packageName });
	return Object.keys(installed?.packageJson.dependencies || {});
}
/**
* In dev mode, detects shared packages that are sub-dependencies of other
* shared packages and removes them to avoid initialization order issues.
* For example, `lit` depends on `lit-html`, `lit-element`, and
* `@lit/reactive-element` — sharing them separately causes the child modules
* to load before their parent, resulting in `undefined` class extends errors.
*/
function excludeSharedSubDependencies(shared) {
	const sharedKeys = new Set(Object.keys(shared));
	const sharedKeyByBase = new Map(Object.keys(shared).map((key) => [key.endsWith("/") ? key.slice(0, -1) : key, key]));
	for (const parentKey of sharedKeys) {
		const deps = getPackageDependencies(parentKey);
		for (const dep of deps) {
			const depKey = sharedKeyByBase.get(dep);
			if (depKey && depKey !== parentKey) {
				if (shared[depKey]?.shareConfig.singleton === true || shared[depKey]?.shareConfig.import === false) continue;
				mfWarn(`"${dep}" is a dependency of shared package "${parentKey}" and is also shared separately. This may cause initialization order issues in dev mode. Consider sharing only "${parentKey}".\n  Auto-excluding "${dep}" from shared modules for dev mode.`);
				delete shared[depKey];
				sharedKeys.delete(depKey);
				sharedKeyByBase.delete(dep);
			}
		}
	}
}
function proxySharedModule(options) {
	const { shared = {} } = options;
	let _config;
	let _command = "serve";
	let useDirectReactImport = false;
	let useRolldown = false;
	const savePrebuild = new PromiseStore();
	let devServer;
	return [
		{
			name: "generateLocalSharedImportMap",
			enforce: "post",
			configureServer(server) {
				devServer = server;
				setLocalSharedImportMapInvalidator(() => {
					const module = server.moduleGraph.getModuleById(getResolvedLocalSharedImportMapId());
					if (module) server.moduleGraph.invalidateModule(module);
				});
			},
			resolveId(source) {
				if (source === getLocalSharedImportMapPath()) return getResolvedLocalSharedImportMapId();
			},
			load(id) {
				if (id === getResolvedLocalSharedImportMapId()) return parsePromise.then((_) => generateLocalSharedImportMap());
			},
			closeBundle() {
				if (devServer) return;
				setLocalSharedImportMapInvalidator(void 0);
			}
		},
		{
			name: "proxyPreBuildShared",
			enforce: "post",
			config(config, { command }) {
				setPackageDetectionCwd(config.root || process.cwd());
				const isVinext = hasPackageDependency("vinext");
				const isAstro = hasPackageDependency("astro");
				const isRolldown = getIsRolldown(this);
				_command = command;
				useRolldown = isRolldown;
				useDirectReactImport = isVinext || isAstro;
				if (command === "serve") excludeSharedSubDependencies(shared);
			},
			configResolved(config) {
				_config = config;
				const isRolldown = getIsRolldown(this);
				Object.keys(shared).forEach((key) => {
					if (key.endsWith("/")) return;
					if (useDirectReactImport && key === "react") {
						addUsedShares(key);
						return;
					}
					writeLoadShareModule(key, shared[key], _command, isRolldown);
					if (shared[key].shareConfig.import !== false) writePreBuildLibPath(key, shared[key]);
					addUsedShares(key);
				});
				writeLocalSharedImportMap();
				refreshHostAutoInit();
			}
		},
		{
			name: "proxyPreBuildShared:resolve-shared-loadShare",
			enforce: "pre",
			async resolveId(source, importer) {
				const key = findSharedKeyForSource(source, shared);
				if (!key) return;
				if (useDirectReactImport && key === "react") return;
				if (/\.css$/.test(source)) return;
				if (isBuildConfigImporter(importer)) return;
				if (useDirectReactImport && source === "react") return;
				if (importer && importer.includes("localSharedImportMap")) return;
				if (importer && (importer.includes("hostAutoInit") || importer.includes("__H_A_I__"))) return;
				if (importer && importer.includes("__loadShare__")) return;
				if (importer && importer.includes("__prebuild__")) return;
				const shareSource = isNodeModulePath(source) ? getCommonSharedSubpathFromNodeModulePath(source, key) || key : source;
				const loadSharePath = getLoadShareModulePath(shareSource, useRolldown);
				writeLoadShareModule(shareSource, shared[key], _command, useRolldown);
				if (shared[key].shareConfig.import !== false) writePreBuildLibPath(shareSource, shared[key]);
				addUsedShares(shareSource);
				writeLocalSharedImportMap();
				refreshHostAutoInit();
				return this.resolve(loadSharePath, importer, { skipSelf: true });
			}
		},
		{
			name: "proxyPreBuildShared:resolve-prebuild",
			enforce: "pre",
			async resolveId(source, importer) {
				if (!source.includes("__prebuild__")) return;
				if (source.startsWith(".")) return;
				const pkgName = assertModuleFound(PREBUILD_TAG, source).name;
				const importSource = getPrebuildResolutionSource(pkgName, getPreBuildShareItem(pkgName));
				if (_command === "build") return this.resolve(importSource, importer, { skipSelf: true });
				const direct = tryResolveFromProjectRoot(importSource);
				const directSource = direct && !isNodeModulePath(direct) ? direct : void 0;
				const resolved = await this.resolve(directSource || importSource, importer, { skipSelf: true });
				if (!resolved?.id) return;
				const result = resolved.id;
				if (!_config || result.includes(_config.cacheDir)) {
					if (directSource) return await this.resolve(directSource, importer, { skipSelf: true }) || { id: directSource };
					return resolved;
				}
				savePrebuild.set(pkgName, Promise.resolve(result));
				return await this.resolve(await savePrebuild.get(pkgName), importer, { skipSelf: true });
			}
		}
	];
}
//#endregion
//#region src/utils/loadWalk.ts
let walkPromise = null;
function loadWalk() {
	walkPromise ||= import("estree-walker").then(({ walk }) => walk);
	return walkPromise;
}
//#endregion
//#region src/plugins/pluginRemoteNamedExports.ts
/**
* Transforms consumer-side imports of remote modules so that named exports
* are accessible even when the bundler does not support syntheticNamedExports
* (Rolldown / Vite 8+).
*
* The remote proxy module exports:
*   export const __moduleExports = exportModule;   // full namespace
*   export default exportModule.default ?? exportModule;  // unwrapped default
*
* This plugin rewrites consumer code:
*   import { foo } from "remote/xxx"
*     → import { __moduleExports as __mf_ns_0 } from "remote/xxx"; const { foo } = __mf_ns_0;
*
*   import("remote/xxx")
*     → import("remote/xxx").then(…)  // spreads __moduleExports into namespace
*
* NOTE: `export * from "remote/xxx"` is not supported — Rolldown cannot
* statically resolve the set of exported names from a federated remote at
* build time.  Use explicit named re-exports instead.
*/
const JS_EXTENSIONS_RE = /\.(?:[mc]?[jt]sx?|vue|svelte)(?:\?|$)/;
const REGEX_FALLBACK_EXTENSIONS_RE = /\.(?:[mc]?[jt]sx?)(?:\?|$)/;
function parseNamedSpecifiers(specifiersRaw, kind) {
	return specifiersRaw.split(",").map((s) => s.trim()).filter((s) => s.length > 0 && !s.startsWith("type ")).map((s) => {
		const asMatch = s.match(/^(\w+)\s+as\s+(\w+)$/);
		const sourceName = asMatch ? asMatch[1] : s;
		const targetName = asMatch ? asMatch[2] : s;
		return kind === "import" ? {
			imported: sourceName,
			local: targetName
		} : {
			local: sourceName,
			exported: targetName
		};
	});
}
function wrapDynamicImport(original) {
	return `${original}.then(function(__mf_m__) {\n  var __mf_ready__ = __mf_m__ && __mf_m__.__mf_remote_pending ? __mf_m__.__mf_remote_pending.then(function(__mf_resolved__) { return __mf_resolved__ || __mf_m__; }) : __mf_m__;\n  return Promise.resolve(__mf_ready__).then(function(__mf_m__) {\n  if (!__mf_m__ || !__mf_m__.__moduleExports) {\n    if (__mf_m__ && __mf_m__.default && typeof __mf_m__.default === "object" && __mf_m__.default.__esModule) {\n      var __mf_nested_e__ = __mf_m__.default;\n      var __mf_nested_ns__ = Object.create(null);\n      Object.defineProperty(__mf_nested_ns__, Symbol.toStringTag, { value: "Module" });\n      Object.keys(__mf_nested_e__).forEach(function(k) { if (k !== "__esModule") __mf_nested_ns__[k] = __mf_nested_e__[k] });\n      if ("default" in __mf_nested_e__) __mf_nested_ns__.default = __mf_nested_e__.default;\n      return __mf_nested_ns__;\n    }\n    return __mf_m__;\n  }\n  var __mf_ns__ = Object.create(null);\n  Object.defineProperty(__mf_ns__, Symbol.toStringTag, { value: "Module" });\n  var __mf_e__ = __mf_m__.__moduleExports;\n  if (__mf_e__ && __mf_e__.default && typeof __mf_e__.default === "object" && __mf_e__.default.__esModule) __mf_e__ = __mf_e__.default;\n  Object.keys(__mf_e__).forEach(function(k) { if (k !== "__esModule") __mf_ns__[k] = __mf_e__[k] });\n  if ("default" in __mf_e__) __mf_ns__.default = __mf_e__.default;\n  else if ("default" in __mf_m__) __mf_ns__.default = __mf_m__.default;\n  return __mf_ns__;\n  });\n})`;
}
function applyRewrites(code, imports, id) {
	if (imports.length === 0) return;
	const ms = new CodeRewriter(code);
	let changed = false;
	let counter = 0;
	for (const imp of imports) switch (imp.kind) {
		case "static": {
			const src = JSON.stringify(imp.source);
			if (imp.namespaceLocal && !imp.defaultLocal && imp.named.length === 0) ms.overwrite(imp.start, imp.end, `import { __moduleExports as ${imp.namespaceLocal} } from ${src};`);
			else {
				const nsId = `__mf_ns_${counter++}`;
				const importParts = [];
				if (imp.defaultLocal) importParts.push(`default as ${imp.defaultLocal}`);
				importParts.push(`__moduleExports as ${nsId}`);
				const destructParts = imp.named.map((s) => s.imported === s.local ? s.local : `${s.imported}: ${s.local}`);
				let rewrite = `import { ${importParts.join(", ")} } from ${src};`;
				if (destructParts.length > 0) rewrite += `\nconst { ${destructParts.join(", ")} } = ${nsId};`;
				ms.overwrite(imp.start, imp.end, rewrite);
			}
			changed = true;
			break;
		}
		case "reexport": {
			const src = JSON.stringify(imp.source);
			const nsId = `__mf_ns_${counter++}`;
			const vars = imp.specifiers.map((s) => {
				const tmp = `__mf_re_${counter++}`;
				return {
					...s,
					tmp
				};
			});
			const importLine = `import { __moduleExports as ${nsId} } from ${src};`;
			const varLines = vars.map((v) => `const ${v.tmp} = ${nsId}[${JSON.stringify(v.local)}];`).join("\n");
			const exportLine = `export { ${vars.map((v) => `${v.tmp} as ${v.exported}`).join(", ")} };`;
			ms.overwrite(imp.start, imp.end, `${importLine}\n${varLines}\n${exportLine}`);
			changed = true;
			break;
		}
		case "export-all":
			console.warn(`[module-federation] "export * from '${imp.source}'" is not supported with Rolldown — use explicit named re-exports instead. (${id})`);
			break;
		case "dynamic":
			ms.overwrite(imp.start, imp.end, wrapDynamicImport(imp.originalText));
			changed = true;
			break;
	}
	if (!changed) return;
	return {
		code: ms.toString(),
		map: ms.generateMap(id)
	};
}
async function collectFromAST(ast, code, isRemoteImport) {
	const walk = await loadWalk();
	const result = [];
	walk(ast, { enter(node) {
		if (node.type === "ImportDeclaration" && node.source?.value) {
			if (!isRemoteImport(node.source.value)) return;
			const specifiers = node.specifiers || [];
			const named = specifiers.filter((s) => s.type === "ImportSpecifier" && s.importKind !== "type").map((s) => ({
				imported: s.imported.name ?? s.imported.value,
				local: s.local.name
			}));
			const defaultSpec = specifiers.find((s) => s.type === "ImportDefaultSpecifier");
			const nsSpec = specifiers.find((s) => s.type === "ImportNamespaceSpecifier");
			if (named.length === 0 && !nsSpec) return;
			result.push({
				kind: "static",
				source: node.source.value,
				start: node.start,
				end: node.end,
				named,
				defaultLocal: defaultSpec?.local.name,
				namespaceLocal: nsSpec?.local.name
			});
		}
		if (node.type === "ExportNamedDeclaration" && node.source?.value && isRemoteImport(node.source.value)) {
			const specifiers = (node.specifiers || []).filter((s) => s.exportKind !== "type").map((s) => ({
				local: s.local.name ?? s.local.value,
				exported: s.exported.name ?? s.exported.value
			}));
			if (specifiers.length === 0) return;
			result.push({
				kind: "reexport",
				source: node.source.value,
				start: node.start,
				end: node.end,
				specifiers
			});
		}
		if (node.type === "ExportAllDeclaration" && node.source?.value && isRemoteImport(node.source.value)) {
			this.skip();
			result.push({
				kind: "export-all",
				source: node.source.value,
				start: node.start,
				end: node.end
			});
		}
		if (node.type === "ImportExpression") {
			const source = node.source;
			if (source.type !== "Literal" && source.type !== "StringLiteral" && source.type !== "TemplateLiteral") return;
			const value = source.type === "TemplateLiteral" ? source.quasis?.length === 1 ? source.quasis[0].value?.cooked : void 0 : source.value;
			if (!value || !isRemoteImport(value)) return;
			result.push({
				kind: "dynamic",
				start: node.start,
				end: node.end,
				originalText: code.slice(node.start, node.end)
			});
		}
	} });
	return result;
}
async function collectFromEsLexer(code, isRemoteImport) {
	await init;
	let imports;
	try {
		[imports] = parse$1(code);
	} catch {
		return;
	}
	const result = [];
	for (const imp of imports) {
		if (imp.d === -2) continue;
		if (!imp.n || !isRemoteImport(imp.n)) continue;
		const stmtText = code.slice(imp.ss, imp.se);
		if (imp.d >= 0) {
			result.push({
				kind: "dynamic",
				start: imp.ss,
				end: imp.se,
				originalText: stmtText
			});
			continue;
		}
		if (/^\s*export\s*\*\s/.test(stmtText)) {
			result.push({
				kind: "export-all",
				source: imp.n,
				start: imp.ss,
				end: imp.se
			});
			continue;
		}
		if (/^\s*export\s/.test(stmtText)) {
			const braceMatch = stmtText.match(/\{([^}]*)\}/);
			if (!braceMatch) continue;
			const specifiers = parseNamedSpecifiers(braceMatch[1], "export");
			if (specifiers.length === 0) continue;
			result.push({
				kind: "reexport",
				source: imp.n,
				start: imp.ss,
				end: imp.se,
				specifiers
			});
			continue;
		}
		const importMatch = stmtText.match(/^import\s+([\s\S]*?)\s+from\s/);
		if (!importMatch) continue;
		const specifiersPart = importMatch[1].trim();
		if (/^type\s/.test(specifiersPart)) continue;
		const nsMatch = specifiersPart.match(/^\*\s+as\s+(\w+)$/);
		if (nsMatch) {
			result.push({
				kind: "static",
				source: imp.n,
				start: imp.ss,
				end: imp.se,
				named: [],
				namespaceLocal: nsMatch[1]
			});
			continue;
		}
		const braceMatch = specifiersPart.match(/\{([^}]*)\}/);
		if (!braceMatch) continue;
		const named = parseNamedSpecifiers(braceMatch[1], "import");
		if (named.length === 0) continue;
		const defaultMatch = specifiersPart.match(/^(\w+)\s*,/);
		result.push({
			kind: "static",
			source: imp.n,
			start: imp.ss,
			end: imp.se,
			named,
			defaultLocal: defaultMatch?.[1]
		});
	}
	return result;
}
function collectFromRegex(code, isRemoteImport) {
	const result = [];
	for (const match of code.matchAll(/^\s*import\s+([\s\S]*?)\s+from\s+(['"])([^'"]+)\2\s*;?/gm)) {
		const [full, specifiersPartRaw, , source] = match;
		if (!isRemoteImport(source)) continue;
		const specifiersPart = specifiersPartRaw.trim();
		if (/^type\s/.test(specifiersPart)) continue;
		const nsMatch = specifiersPart.match(/^\*\s+as\s+(\w+)$/);
		if (nsMatch) {
			result.push({
				kind: "static",
				source,
				start: match.index,
				end: match.index + full.length,
				named: [],
				namespaceLocal: nsMatch[1]
			});
			continue;
		}
		const braceMatch = specifiersPart.match(/\{([^}]*)\}/);
		if (!braceMatch) continue;
		const named = parseNamedSpecifiers(braceMatch[1], "import");
		if (named.length === 0) continue;
		const defaultMatch = specifiersPart.match(/^(\w+)\s*,/);
		result.push({
			kind: "static",
			source,
			start: match.index,
			end: match.index + full.length,
			named,
			defaultLocal: defaultMatch?.[1]
		});
	}
	for (const match of code.matchAll(/^\s*export\s+\{([\s\S]*?)\}\s+from\s+(['"])([^'"]+)\2\s*;?/gm)) {
		const [full, specifiersRaw, , source] = match;
		if (!isRemoteImport(source)) continue;
		const specifiers = parseNamedSpecifiers(specifiersRaw, "export");
		if (specifiers.length === 0) continue;
		result.push({
			kind: "reexport",
			source,
			start: match.index,
			end: match.index + full.length,
			specifiers
		});
	}
	for (const match of code.matchAll(/^\s*export\s+\*\s+from\s+(['"])([^'"]+)\1\s*;?/gm)) {
		const [full, , source] = match;
		if (!isRemoteImport(source)) continue;
		result.push({
			kind: "export-all",
			source,
			start: match.index,
			end: match.index + full.length
		});
	}
	for (const match of code.matchAll(/import\(\s*(['"])([^'"]+)\1\s*\)/g)) {
		const [full, , source] = match;
		if (!isRemoteImport(source)) continue;
		result.push({
			kind: "dynamic",
			start: match.index,
			end: match.index + full.length,
			originalText: full
		});
	}
	return result.length > 0 ? result : void 0;
}
function pluginRemoteNamedExports(options) {
	const remoteNames = Object.keys(options.remotes);
	const isNodeModulesId = (id) => id.includes("/node_modules/") || id.includes("\\node_modules\\");
	function isRemoteImport(source, importerId) {
		return remoteNames.some((name) => {
			if (source.startsWith(name + "/")) return true;
			if (source !== name) return false;
			return !isNodeModulesId(importerId);
		}) || source.includes("__loadRemote__");
	}
	return {
		name: "module-federation-remote-named-exports",
		enforce: "post",
		async transform(code, id) {
			if (remoteNames.length === 0) return;
			if (id.includes("__loadRemote__") || id.includes("__loadShare__")) return;
			if (!JS_EXTENSIONS_RE.test(id)) return;
			if (!remoteNames.some((name) => code.includes(name))) return;
			const matchesRemoteImport = (source) => isRemoteImport(source, id);
			let imports;
			try {
				imports = await collectFromAST(this.parse(code), code, matchesRemoteImport);
			} catch {
				imports = await collectFromEsLexer(code, matchesRemoteImport);
			}
			if ((!imports || imports.length === 0) && REGEX_FALLBACK_EXTENSIONS_RE.test(id)) imports = collectFromRegex(code, matchesRemoteImport);
			if (!imports) return;
			return applyRewrites(code, imports, id);
		}
	};
}
//#endregion
//#region src/plugins/pluginVarRemoteEntry.ts
const VarRemoteEntry = () => {
	const mfOptions = getNormalizeModuleFederationOptions();
	const { name, varFilename, filename } = mfOptions;
	let viteConfig;
	return [{
		name: "module-federation-var-remote-entry",
		apply: "serve",
		configResolved(config) {
			viteConfig = config;
		},
		configureServer(server) {
			server.middlewares.use((req, res, next) => {
				if (!varFilename) {
					next();
					return;
				}
				if (req.url?.replace(/\?.*/, "") === (viteConfig.base + varFilename).replace(/^\/?/, "/")) {
					res.setHeader("Content-Type", "text/javascript");
					res.setHeader("Access-Control-Allow-Origin", "*");
					res.end(generateVarRemoteEntry(filename));
				} else next();
			});
		}
	}, {
		name: "module-federation-var-remote-entry",
		enforce: "post",
		config(config) {
			if (!config.build) config.build = {};
		},
		configResolved(config) {
			viteConfig = config;
		},
		async generateBundle(_options, bundle) {
			if (!varFilename) return;
			if (!isValidVarName(name)) mfWarn(`Provided remote name "${name}" is not valid for "var" remoteEntry type, thus it's placed in globalThis['${name}'].\nIt may cause problems, so you would better want to use valid var name (see https://www.w3schools.com/js/js_variables.asp).`);
			const remoteEntryFile = findRemoteEntryFile(mfOptions.filename, bundle);
			if (!remoteEntryFile) throw createModuleFederationError(`Couldn't find a remoteEntry chunk file for ${mfOptions.filename}, can't generate varRemoteEntry file`);
			this.emitFile({
				type: "asset",
				fileName: varFilename,
				source: generateVarRemoteEntry(remoteEntryFile)
			});
		}
	}];
	function isValidVarName(name) {
		return /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(name);
	}
	/**
	* Generates the final "var" remote entry file
	* @param remoteEntryFile - Path to esm remote entry file
	* @returns Complete "var" remoteEntry.js file source
	*/
	function generateVarRemoteEntry(remoteEntryFile) {
		const { name, varFilename } = getNormalizeModuleFederationOptions();
		const isValidName = isValidVarName(name);
		return `
  ${isValidName ? `var ${name};` : ""}
  ${isValidName ? name : `globalThis['${name}']`} = (function () {
    function getScriptUrl() {
      const currentScript = document.currentScript;
      if (!currentScript) {
        console.error("[Module Federation] ${varFilename} script should be called from sync <script> tag (document.currentScript is undefined)")
        return '/';
      }
      return document.currentScript.src.replace(/\\/[^/]*$/, '/');
    }

    const entry = getScriptUrl() + '${remoteEntryFile}';

    return {
      get: (...args) => import(entry).then(m => m.get(...args)),
      init: (...args) => import(entry).then(m => m.init(...args)),
    };
  })();
  `;
	}
};
//#endregion
//#region src/utils/aliasToArrayPlugin.ts
var aliasToArrayPlugin_default = {
	name: "alias-transform-plugin",
	config: (config) => {
		if (!config.resolve) config.resolve = {};
		if (!config.resolve.alias) config.resolve.alias = [];
		const { alias } = config.resolve;
		if (typeof alias === "object" && !Array.isArray(alias)) config.resolve.alias = Object.entries(alias).map(([find, replacement]) => ({
			find,
			replacement
		}));
	}
};
//#endregion
//#region src/utils/controlChunkSanitizer.ts
const FEDERATION_CONTROL_CHUNK_HINTS = [
	"hostInit",
	"virtualExposes",
	"localSharedImportMap"
];
function stripEmptyPreloadCalls(code) {
	const helperImportRegex = /import\s*\{\s*_\s*as\s*(\w+)\s*\}\s*from\s*["'][^"']+["']\s*;?/g;
	const helperAliases = [];
	let helperImportMatch;
	while ((helperImportMatch = helperImportRegex.exec(code)) !== null) helperAliases.push(helperImportMatch[1]);
	let nextCode = code;
	for (const alias of helperAliases) {
		const marker = `${alias}(()=>`;
		let start = nextCode.indexOf(marker);
		while (start !== -1) {
			const exprStart = start + marker.length;
			let depth = 0;
			let cursor = exprStart;
			let replacementEnd = -1;
			while (cursor < nextCode.length) {
				const char = nextCode[cursor];
				if (char === "(") depth++;
				else if (char === ")") {
					depth--;
					if (depth < 0) break;
				} else if (depth === 0 && nextCode.startsWith(",[],import.meta.url)", cursor)) {
					replacementEnd = cursor;
					break;
				}
				cursor++;
			}
			if (replacementEnd === -1) {
				start = nextCode.indexOf(marker, start + marker.length);
				continue;
			}
			const expression = nextCode.slice(exprStart, replacementEnd);
			nextCode = nextCode.slice(0, start) + expression + nextCode.slice(replacementEnd + 20);
			start = nextCode.indexOf(marker, start + expression.length);
		}
	}
	nextCode = nextCode.replace(/import\s*["'][^"']*__loadShare__[^"']*["']\s*;?/g, "");
	nextCode = nextCode.replace(helperImportRegex, (statement, local) => {
		return new RegExp(`\\b${local}\\s*\\(`).test(nextCode.replace(statement, "")) ? statement : "";
	});
	return nextCode;
}
function isFederationControlChunk(fileName, filename) {
	return fileName.includes(filename) || FEDERATION_CONTROL_CHUNK_HINTS.some((hint) => fileName.includes(hint));
}
function sanitizeFederationControlChunk(code, fileName, filename) {
	let nextCode = stripEmptyPreloadCalls(code);
	if (fileName.includes("localSharedImportMap")) {
		const remoteEntryImportRegex = new RegExp(`import\\s*["'][^"']*${filename.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}["']\\s*;?`, "g");
		nextCode = nextCode.replace(remoteEntryImportRegex, "");
	}
	return nextCode;
}
//#endregion
//#region src/utils/isTestEnv.ts
/**
* Detects whether the current process is running in a test environment
* Set `MFE_VITE_NO_TEST_ENV_CHECK=true` to load federation plugins during tests.
*/
function isTestEnv() {
	if (process.env.MFE_VITE_NO_TEST_ENV_CHECK === "true") return false;
	return process.env.NODE_ENV === "test" || process.env.VITEST != null || process.env.JEST_WORKER_ID != null;
}
//#endregion
//#region src/utils/normalizeOptimizeDeps.ts
var normalizeOptimizeDeps_default = {
	name: "normalizeOptimizeDeps",
	config: (config) => {
		let { optimizeDeps } = config;
		if (!optimizeDeps) {
			config.optimizeDeps = {};
			optimizeDeps = config.optimizeDeps;
		}
		if (!optimizeDeps.include) optimizeDeps.include = [];
		if (!optimizeDeps.exclude) optimizeDeps.exclude = [];
		if (!optimizeDeps.needsInterop) optimizeDeps.needsInterop = [];
	}
};
//#endregion
//#region src/index.ts
const patchedManualChunks = /* @__PURE__ */ new WeakSet();
function normalizeVinextRscPreloadHints(code) {
	return code.replace(/(:HL\[[^\]\n]*?,)"stylesheet"/g, "$1\"style\"").replace(/(:HL\[[^\]\n]*?,)\\"stylesheet\\"/g, "$1\\\"style\\\"");
}
function ignoreFederationGeneratedFiles(config, options) {
	config.server ??= {};
	config.server.watch ??= {};
	const federationIgnore = (file) => shouldIgnoreFile(file, options);
	const ignored = config.server.watch.ignored;
	if (!ignored) {
		config.server.watch.ignored = federationIgnore;
		return;
	}
	if (Array.isArray(ignored)) {
		ignored.push(federationIgnore);
		return;
	}
	config.server.watch.ignored = [ignored, federationIgnore];
}
function isSharedResolverInternalImporter(importer) {
	return !!importer && (importer.includes("__loadShare__") || importer.includes("__prebuild__"));
}
function isCommonJsImporter(importer) {
	return !!importer && (importer.endsWith(".cjs") || importer.includes("/cjs/"));
}
function isOutputChunk(chunk) {
	return chunk.type === "chunk";
}
function appendResolveAlias(config, alias) {
	const resolve = config.resolve ??= {};
	const existingAlias = resolve.alias;
	if (!existingAlias) {
		resolve.alias = [alias];
		return;
	}
	if (Array.isArray(existingAlias)) {
		existingAlias.push(alias);
		return;
	}
	resolve.alias = [...Object.entries(existingAlias).map(([find, replacement]) => ({
		find,
		replacement
	})), alias];
}
const UNSAFE_JS_SOURCE_CHAR_MAP = {
	"<": "\\u003C",
	">": "\\u003E",
	"/": "\\u002F",
	"\\": "\\\\",
	"\b": "\\b",
	"\f": "\\f",
	"\n": "\\n",
	"\r": "\\r",
	"	": "\\t",
	"\0": "\\0",
	"\u2028": "\\u2028",
	"\u2029": "\\u2029"
};
function escapeUnsafeJsSourceChars(str) {
	return str.replace(/[<>/\\\b\f\n\r\t\0\u2028\u2029]/g, (char) => {
		return UNSAFE_JS_SOURCE_CHAR_MAP[char] ?? char;
	});
}
function isFederationHtmlPreloadDependency(dep, includeSharedRuntime = false) {
	const file = path.basename(dep);
	if (file.includes("__mfe_internal__") || file.includes("virtual_mf-") || file.includes("virtualExposes") || file.includes("localSharedImportMap") || file.includes("hostInit")) return true;
	return includeSharedRuntime && (file.includes("preload-helper") || file.includes("rolldown-runtime") || file.startsWith("dist-"));
}
/**
* Plugin that runs FIRST to create virtual module files in the config hook.
* This prevents 504 "Outdated Optimize Dep" errors by ensuring files exist
* before Vite's optimization phase.
*/
function createEarlyVirtualModulesPlugin(options) {
	const { shared, remotes, virtualModuleDir } = options;
	const isLitShare = (pkg) => pkg === "lit" || pkg.startsWith("lit/");
	return {
		name: "vite:module-federation-early-init",
		enforce: "pre",
		config(config, { command: _command }) {
			if (_command === "serve") ignoreFederationGeneratedFiles(config, options);
			const root = config.root || process.cwd();
			setPackageDetectionCwd(root);
			const isVinext = hasPackageDependency("vinext");
			initVirtualModuleInfrastructure(root, virtualModuleDir);
			VirtualModule.setRoot(root);
			VirtualModule.ensureVirtualPackageExists();
			initVirtualModules(_command, getRemoteEntryId(options));
			const isRolldown = getIsRolldown(this);
			if (remotes && Object.keys(remotes).length > 0) {
				for (const key of Object.keys(remotes)) addUsedRemote(key, key);
				if (_command === "serve") {
					config.optimizeDeps = config.optimizeDeps || {};
					config.optimizeDeps.exclude = config.optimizeDeps.exclude || [];
					config.optimizeDeps.include = config.optimizeDeps.include || [];
					config.optimizeDeps.exclude.push(...Object.keys(remotes || {}));
				}
			}
			if (shared && Object.keys(shared).length > 0) {
				if (_command === "serve") {
					config.optimizeDeps = config.optimizeDeps || {};
					config.optimizeDeps.include = config.optimizeDeps.include || [];
					const optimizeDeps = config.optimizeDeps;
					if (isRolldown) {
						optimizeDeps.rolldownOptions ??= {};
						optimizeDeps.rolldownOptions.plugins ??= [];
						optimizeDeps.rolldownOptions.plugins.push({
							name: "module-federation:optimize-shared-resolver",
							resolveId(source, importer, options) {
								if (options?.kind?.startsWith("require")) return;
								if (isSharedResolverInternalImporter(importer)) return;
								if (isCommonJsImporter(importer)) return;
								const key = findSharedKey(source, shared);
								if (!key) return;
								if (source.endsWith(".css")) return;
								const shareItem = shared[key];
								const loadSharePath = getLoadShareModulePath(source, isRolldown);
								writeLoadShareModule(source, shareItem, _command, isRolldown);
								if (shareItem.shareConfig?.import !== false) writePreBuildLibPath(source, shareItem);
								addUsedShares(source);
								return {
									id: loadSharePath,
									external: true
								};
							}
						});
					} else {
						optimizeDeps.esbuildOptions ??= {};
						optimizeDeps.esbuildOptions.plugins ??= [];
						optimizeDeps.esbuildOptions.plugins.push({
							name: "module-federation:optimize-shared-proxy",
							setup(build) {
								build.onResolve({ filter: /.*/ }, (args) => {
									if (!args.importer || args.namespace === "mf-shared") return;
									if (isSharedResolverInternalImporter(args.importer)) return;
									if (!findSharedKey(args.path, shared) || args.path.endsWith(".css")) return;
									return {
										path: args.path,
										namespace: "mf-shared"
									};
								});
								build.onLoad({
									filter: /.*/,
									namespace: "mf-shared"
								}, (args) => {
									const key = findSharedKey(args.path, shared);
									if (!key) return;
									const shareItem = shared[key];
									const loadSharePath = getLoadShareModulePath(args.path, isRolldown);
									writeLoadShareModule(args.path, shareItem, _command, isRolldown);
									if (shareItem.shareConfig?.import !== false) writePreBuildLibPath(args.path, shareItem);
									addUsedShares(args.path);
									return {
										loader: "js",
										resolveDir: root,
										contents: `import * as __mfShared from ${JSON.stringify(loadSharePath)};
export * from ${JSON.stringify(loadSharePath)};
export default __mfShared.default ?? __mfShared;`
									};
								});
							}
						});
					}
					config.optimizeDeps.include.push(virtualRuntimeInitStatus.getImportId());
				}
				for (const key of Object.keys(shared)) {
					const shareItem = shared[key];
					if (key.endsWith("/")) {
						if (_command === "serve" && shareItem.shareConfig?.import !== false) {
							const optimizeDeps = config.optimizeDeps ??= {};
							optimizeDeps.include ??= [];
							for (const subpath of getCommonSharedSubpaths(key)) {
								writePreBuildLibPath(subpath, shareItem);
								optimizeDeps.include.push(subpath);
								optimizeDeps.include.push(getPreBuildLibImportId(subpath));
							}
						}
						continue;
					}
					if (isVinext && key === "react") {
						addUsedShares(key);
						continue;
					}
					getLoadShareModulePath(key, isRolldown);
					writeLoadShareModule(key, shareItem, _command, isRolldown);
					if (shareItem.shareConfig?.import !== false) writePreBuildLibPath(key, shareItem);
					addUsedShares(key);
					if (_command === "serve" && shareItem.shareConfig?.import !== false) {
						const optimizeDeps = config.optimizeDeps ??= {};
						optimizeDeps.include ??= [];
						optimizeDeps.exclude ??= [];
						const shouldBypassOptimizeDep = isLitShare(key);
						if (shouldBypassOptimizeDep) optimizeDeps.exclude.push(key);
						if (!isRolldown && !shouldBypassOptimizeDep) optimizeDeps.include.push(getLoadShareImportId(key, isRolldown));
						optimizeDeps.include.push(getPreBuildLibImportId(key));
						for (const subpath of getCommonSharedSubpaths(key)) {
							writePreBuildLibPath(subpath, shareItem);
							optimizeDeps.include.push(subpath);
							optimizeDeps.include.push(getPreBuildLibImportId(subpath));
						}
					}
				}
				writeLocalSharedImportMap();
			}
		}
	};
}
function federation(mfUserOptions) {
	if (isTestEnv()) return [];
	const options = normalizeModuleFederationOptions(mfUserOptions);
	const isVinext = hasPackageDependency("vinext");
	const { name, shared, filename, hostInitInjectLocation } = options;
	if (!name) throw createModuleFederationError("name is required");
	const remoteEntryId = getRemoteEntryId(options);
	const virtualExposesId = getVirtualExposesId(options);
	let command;
	let desiredRolldownOutput;
	return [
		createEarlyVirtualModulesPlugin(options),
		...isVinext ? [{
			name: "module-federation-vinext-react-server-build-alias",
			apply: "build",
			enforce: "pre",
			resolveId(id) {
				const reactServerEntryMap = {
					"react/jsx-runtime": "react/cjs/react-jsx-runtime.production.js",
					"react/jsx-dev-runtime": "react/cjs/react-jsx-dev-runtime.production.js"
				};
				if (!(id in reactServerEntryMap)) return;
				const environmentName = this.environment?.name;
				if (!environmentName || environmentName === "client") return;
				const target = reactServerEntryMap[id];
				const reactPackageJson = createRequire$1(new URL(`file://${process.cwd()}/package.json`)).resolve("react/package.json");
				return path.join(path.dirname(reactPackageJson), target.replace(/^react\//, ""));
			}
		}] : [],
		{
			name: "vite:module-federation-config",
			enforce: "pre",
			config(_config, env) {
				command = env.command;
			},
			configResolved(config) {
				VirtualModule.setRoot(config.root);
				VirtualModule.ensureVirtualPackageExists();
				initVirtualModules(command, remoteEntryId);
			}
		},
		aliasToArrayPlugin_default,
		checkAliasConflicts({ shared }),
		normalizeOptimizeDeps_default,
		...pluginDts(options),
		pluginDevRemoteHmr(options),
		...addEntry({
			entryName: "remoteEntry",
			entryPath: remoteEntryId,
			fileName: filename
		}),
		...addEntry({
			entryName: "hostInit",
			entryPath: () => getHostAutoInitPath(),
			inject: hostInitInjectLocation,
			forceClientInjected: Object.keys(options.exposes).length > 0
		}),
		...addEntry({
			entryName: "virtualExposes",
			entryPath: virtualExposesId
		}),
		pluginProxyRemoteEntry_default({
			options,
			remoteEntryId,
			virtualExposesId
		}),
		pluginProxyRemotes_default(options),
		pluginRemoteNamedExports(options),
		...pluginModuleParseEnd_default((id) => {
			return id.includes(getHostAutoInitImportId()) || id.includes(remoteEntryId) || id.includes(virtualExposesId) || id.includes(getLocalSharedImportMapPath());
		}, {
			moduleParseTimeout: options.moduleParseTimeout,
			moduleParseIdleTimeout: options.moduleParseIdleTimeout,
			virtualExposesId
		}),
		...proxySharedModule({ shared }),
		{
			name: "module-federation-esm-shims",
			enforce: "pre",
			apply: "build",
			config(config) {
				const runtimeInitId = virtualRuntimeInitStatus.getImportId();
				config.build = config.build || {};
				if (config.build.modulePreload !== false) {
					const currentModulePreload = config.build.modulePreload && typeof config.build.modulePreload === "object" ? config.build.modulePreload : {};
					const existingResolveDependencies = currentModulePreload.resolveDependencies;
					config.build.modulePreload = {
						...currentModulePreload,
						resolveDependencies(filename, deps, context) {
							const resolvedDeps = existingResolveDependencies ? existingResolveDependencies(filename, deps, context) : deps;
							const hostFile = path.basename(context.hostId);
							if (context.hostType === "js" && (hostFile === options.filename || hostFile.includes("hostInit") || hostFile.includes("virtualExposes") || hostFile.includes("localSharedImportMap"))) return [];
							const hasFederationHtmlDeps = context.hostType === "html" && resolvedDeps.some((dep) => isFederationHtmlPreloadDependency(dep));
							const hasFederationJsDeps = context.hostType === "js" && resolvedDeps.some((dep) => isFederationHtmlPreloadDependency(dep));
							return hasFederationHtmlDeps || hasFederationJsDeps ? resolvedDeps.filter((dep) => !isFederationHtmlPreloadDependency(dep, true)) : resolvedDeps;
						}
					};
				}
				let warnedAboutCodeSplitting = false;
				let warnedAboutCodeSplittingGroups = false;
				const ensureCodeSplitting = (output) => {
					if (output?.codeSplitting === false) {
						delete output.codeSplitting;
						if (warnedAboutCodeSplitting) return;
						warnedAboutCodeSplitting = true;
						mfWarn("Ignoring `output.codeSplitting = false` because module federation requires chunk splitting.");
						return;
					}
					if (!output?.codeSplitting || typeof output.codeSplitting !== "object") return;
					if (!("groups" in output.codeSplitting)) return;
					delete output.codeSplitting.groups;
					if (Object.keys(output.codeSplitting).length === 0) delete output.codeSplitting;
					if (warnedAboutCodeSplittingGroups) return;
					warnedAboutCodeSplittingGroups = true;
					mfWarn("Ignoring `output.codeSplitting.groups` because it conflicts with module federation. Grouping shared dependency init wrappers with their dependent modules can break runtime init order and cause standalone remotes to fail before mount.");
				};
				let warnedAboutManualChunks = false;
				const applyManualChunks = (output) => {
					ensureCodeSplitting(output);
					const isPatchedByPlugin = typeof output.manualChunks === "function" && patchedManualChunks.has(output.manualChunks);
					if (output.manualChunks && !isPatchedByPlugin && !warnedAboutManualChunks) {
						warnedAboutManualChunks = true;
						mfWarn("Ignoring `output.manualChunks` because it conflicts with module federation. Module federation transforms shared dependency imports with async init wrappers, and grouping these transformed modules into a single chunk creates circular async dependencies that cause the application to silently hang.");
					}
					const mfManualChunks = function(id) {
						if (id.includes(runtimeInitId)) return "runtimeInit";
						if (id.includes("__loadShare__")) {
							const match = id.match(/([^/\\]+__loadShare__[^/\\]+)/);
							return match ? match[1] : "loadShare";
						}
					};
					patchedManualChunks.add(mfManualChunks);
					output.manualChunks = mfManualChunks;
				};
				config.build.rollupOptions = config.build.rollupOptions || {};
				const rollupOutput = config.build.rollupOptions.output;
				if (Array.isArray(rollupOutput)) rollupOutput.forEach((output) => applyManualChunks(output));
				else applyManualChunks(config.build.rollupOptions.output ||= {});
				const buildWithRolldown = config.build;
				buildWithRolldown.rolldownOptions = buildWithRolldown.rolldownOptions || {};
				const rolldownOutput = buildWithRolldown.rolldownOptions.output;
				const snapshotRolldownOutput = (output) => ({
					entryFileNames: output.entryFileNames,
					chunkFileNames: output.chunkFileNames,
					assetFileNames: output.assetFileNames
				});
				if (Array.isArray(rolldownOutput)) {
					rolldownOutput.forEach((output) => applyManualChunks(output));
					desiredRolldownOutput = rolldownOutput.map((output) => snapshotRolldownOutput(output));
				} else {
					applyManualChunks(buildWithRolldown.rolldownOptions.output ||= {});
					desiredRolldownOutput = [snapshotRolldownOutput(buildWithRolldown.rolldownOptions.output)];
				}
			},
			async buildApp(builder) {
				const desiredOutput = desiredRolldownOutput;
				if (!desiredOutput) return;
				const applyRolldownOutput = (output, restoredOutput) => {
					if (!output || !restoredOutput) return;
					if (restoredOutput.entryFileNames !== void 0) output.entryFileNames = restoredOutput.entryFileNames;
					if (restoredOutput.chunkFileNames !== void 0) output.chunkFileNames = restoredOutput.chunkFileNames;
					if (restoredOutput.assetFileNames !== void 0) output.assetFileNames = restoredOutput.assetFileNames;
				};
				for (const environment of Object.values(builder.environments)) {
					const getRolldownOptions = environment.getRolldownOptions;
					if (typeof getRolldownOptions !== "function") continue;
					environment.getRolldownOptions = async () => {
						const rolldownOptions = await getRolldownOptions.call(environment);
						if (Array.isArray(rolldownOptions.output)) rolldownOptions.output.forEach((output, index) => {
							applyRolldownOutput(output, desiredOutput[index]);
						});
						else {
							rolldownOptions.output ||= {};
							applyRolldownOutput(rolldownOptions.output, desiredOutput[0]);
						}
						return rolldownOptions;
					};
				}
			},
			load(id) {
				if (id.startsWith("\0")) return;
				if (id.includes("__loadShare__") || id.includes("__loadRemote__")) {
					let code = readFileSync(id, "utf-8");
					code = code.replace(/import\s+["'][^"']*__prebuild__[^"']*["']\s*;?/g, "");
					code = code.replace(/export\s+\*\s+from\s+["'][^"']*__prebuild__[^"']*["']\s*;?/g, "");
					/**
					* Shared/remote shims only have `export default exportModule`.
					*
					* We add a second named export (__moduleExports) that holds the full
					* module namespace and point syntheticNamedExports at it.  This lets
					* Rollup resolve named imports (e.g. `import { useState } from 'react'`)
					* from the namespace while still applying its normal default-export
					* interop — which is needed for libraries like @emotion/styled where
					* `import styled from '@emotion/styled'` must receive the .default
					* function, not the raw namespace object.
					*
					* Using 'default' as the syntheticNamedExports key would skip the
					* interop and break default imports.
					*
					* @see https://rollupjs.org/plugin-development/#synthetic-named-exports
					*/
					if (!code.includes("__moduleExports")) code = code.replace("export default exportModule", "export const __moduleExports = exportModule;\nexport default exportModule.__esModule ? exportModule.default : exportModule");
					if (getIsRolldown(this)) return { code };
					return {
						code,
						syntheticNamedExports: "__moduleExports"
					};
				}
			},
			generateBundle(_outputOptions, bundle, _isWrite) {
				for (const [fileName, chunk] of Object.entries(bundle)) {
					if (!isOutputChunk(chunk)) continue;
					if (!isFederationControlChunk(fileName, filename)) continue;
					chunk.code = sanitizeFederationControlChunk(chunk.code, fileName, filename);
				}
				const proxyChunks = collectLoadShareProxyChunks(bundle, LOAD_SHARE_TAG);
				if (proxyChunks.size > 0) {
					const systemProxyInfo = collectSystemProxyInfos(proxyChunks, LOAD_SHARE_TAG);
					for (const [fileName, chunk] of Object.entries(bundle)) {
						if (!isOutputChunk(chunk)) continue;
						if (proxyChunks.has(fileName)) continue;
						let code = chunk.code;
						if (!fileName.includes("__loadShare__")) code = rewriteEsmProxyConsumers(code, proxyChunks);
						code = rewriteSystemProxyConsumers(code, systemProxyInfo);
						if (code !== chunk.code) chunk.code = code;
					}
				}
			}
		},
		{
			name: "module-federation-strip-empty-preload-helper",
			enforce: "post",
			apply: "build",
			renderChunk(code, chunk) {
				if (!isFederationControlChunk(chunk.fileName, filename)) return;
				const nextCode = sanitizeFederationControlChunk(code, chunk.fileName, filename);
				return nextCode === code ? null : {
					code: nextCode,
					map: null
				};
			},
			writeBundle(outputOptions, bundle) {
				if (!outputOptions.dir) return;
				for (const chunk of Object.values(bundle)) {
					if (!isOutputChunk(chunk)) continue;
					if (!isFederationControlChunk(chunk.fileName, filename)) continue;
					const outputPath = path.join(outputOptions.dir, chunk.fileName);
					writeFileSync(outputPath, sanitizeFederationControlChunk(readFileSync(outputPath, "utf-8"), chunk.fileName, filename));
				}
			}
		},
		{
			name: "module-federation-vite",
			enforce: "post",
			_options: options,
			config(config, { command: _command }) {
				const isRolldown = getIsRolldown(this);
				let implementation = options.implementation;
				if (isRolldown) implementation = implementation.replace(/\.cjs(\.js)?$/, ".js");
				appendResolveAlias(config, {
					find: "@module-federation/runtime",
					replacement: implementation
				});
				config.build ||= {};
				config.build.commonjsOptions ||= {};
				config.build.commonjsOptions.strictRequires ??= "auto";
				const virtualDir = options.virtualModuleDir;
				config.optimizeDeps ||= {};
				config.optimizeDeps.include ||= [];
				config.optimizeDeps.include.push("@module-federation/runtime");
				if (!isRolldown) config.optimizeDeps.include.push(virtualDir);
				config.ssr ||= {};
				config.ssr.noExternal ||= [];
				if (Array.isArray(config.ssr.noExternal)) config.ssr.noExternal.push(virtualDir);
				options.runtimePlugins.forEach((p) => {
					const pluginPath = typeof p === "string" ? p : p[0];
					if (pluginPath && !pluginPath.startsWith(".") && !pluginPath.startsWith("/") && !pluginPath.startsWith("\0") && !pluginPath.startsWith("virtual:")) config.optimizeDeps.include.push(pluginPath);
				});
				if (isRolldown) {
					config.build ??= {};
					config.build.target ??= "esnext";
				} else {
					config.optimizeDeps.needsInterop ||= [];
					config.optimizeDeps.needsInterop.push(virtualDir);
				}
				const isAstro = hasPackageDependency("astro");
				const resolvedTarget = options.target ?? (config.build?.ssr ? "node" : "web");
				const envTargetDefineValue = !options.target && isAstro ? "undefined" : JSON.stringify(resolvedTarget);
				if (!config.define) config.define = {};
				if (!("ENV_TARGET" in config.define)) config.define["ENV_TARGET"] = envTargetDefineValue;
				if (options.target && "ENV_TARGET" in config.define && config.define["ENV_TARGET"] !== JSON.stringify(options.target)) mfWarn(`ENV_TARGET define (${config.define["ENV_TARGET"]}) differs from target option ("${options.target}"). ENV_TARGET will not be overridden.`);
			}
		},
		...Manifest(),
		...VarRemoteEntry(),
		{
			name: "module-federation-vinext-fix-rsc-preload-as",
			enforce: "post",
			configureServer(server) {
				if (!hasPackageDependency("vinext")) return;
				server.middlewares.use((req, res, next) => {
					if (!req.headers.accept?.includes("text/html")) {
						next();
						return;
					}
					const chunks = [];
					const end = res.end.bind(res);
					res.write = (chunk) => {
						if (chunk) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
						return true;
					};
					res.end = (chunk, ...args) => {
						if (chunk) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
						return end(normalizeVinextRscPreloadHints(Buffer.concat(chunks).toString()), ...args);
					};
					next();
				});
			},
			generateBundle(_, bundle, _isWrite) {
				if (!hasPackageDependency("vinext")) return;
				for (const chunk of Object.values(bundle)) {
					if (!isOutputChunk(chunk)) continue;
					if (!chunk.code.includes("case\"L\"")) continue;
					chunk.code = chunk.code.replace(/case"L":(\w+)=(\w+)\[0\],(\w+)=\2\[1\],\2\.length===3\?(\w+)\.L\(\1,\3,\2\[2\]\):\4\.L\(\1,\3\)/g, "case\"L\":$1=$2[0],$3=$2[1],$3===\"stylesheet\"&&($3=\"style\"),$2.length===3?$4.L($1,$3,$2[2]):$4.L($1,$3)");
				}
			}
		},
		...(function() {
			let disablePreload = false;
			return Object.keys(options.exposes).length > 0 ? [{
				name: "module-federation-fix-preload",
				enforce: "post",
				apply: "build",
				config(_config, { command }) {
					const manifest = options.manifest;
					const getDefaultDisableAssetsAnalyze = (cfgCommand) => cfgCommand === "serve" && (typeof manifest !== "object" || !Object.prototype.hasOwnProperty.call(manifest, "disableAssetsAnalyze"));
					const getConfiguredDisableAssetsAnalyze = (cfgCommand) => {
						if (typeof manifest === "object" && manifest !== null) {
							if (Object.prototype.hasOwnProperty.call(manifest, "disableAssetsAnalyze")) return manifest.disableAssetsAnalyze === true;
						}
						return getDefaultDisableAssetsAnalyze(cfgCommand);
					};
					disablePreload = getConfiguredDisableAssetsAnalyze(command);
				},
				generateBundle(_outputOptions, bundle, _isWrite) {
					if (disablePreload) return;
					for (const chunk of Object.values(bundle)) {
						if (!isOutputChunk(chunk)) continue;
						if (!chunk.code.includes("modulepreload")) continue;
						const chunkDir = path.dirname(chunk.fileName);
						const prefixToRoot = chunkDir === "." ? "" : `${path.relative(chunkDir, ".")}/`;
						const replacementExpr = prefixToRoot ? `${escapeUnsafeJsSourceChars(JSON.stringify(prefixToRoot))}+$1` : "$1";
						const replacement = `=function($1){return new URL(${replacementExpr},import.meta.url).href}`;
						const replaced = chunk.code.replace(/=\(?(\w+)(?:,\w+)?\)?\s*=>\s*[`"'][./][^`"']*[`"']\s*\+\s*\1/, replacement);
						if (replaced !== chunk.code) {
							chunk.code = replaced;
							continue;
						}
						chunk.code = chunk.code.replace(/=function\((\w+)(?:,\w+)?\)\{return\s*[`"'][./][^`"']*[`"']\s*\+\s*\1\s*\}/, replacement);
						chunk.code = chunk.code.replace(/=function\((\w+)(?:,\w+)?\)\{return new URL\("\.\.\/"\+\1,import\.meta\.url\)\.href\}/, replacement);
						chunk.code = chunk.code.replace(/new URL\("\.\.\/"\+(\w+),import\.meta\.url\)\.href/g, `new URL(${replacementExpr},import.meta.url).href`);
					}
				}
			}] : [];
		})()
	];
}
//#endregion
export { federation };
