// __mf__virtual/__mfe_internal__vis2DemoWidget__H_A_I__hostAutoInit__H_A_I__.js

    
const __mfCacheGlobalKey = "__mf_module_cache__";
globalThis[__mfCacheGlobalKey] ||= { share: {}, remote: {} };
globalThis[__mfCacheGlobalKey].share ||= {};
globalThis[__mfCacheGlobalKey].remote ||= {};
const __mfModuleCache = globalThis[__mfCacheGlobalKey];

    let hostInitPromise;
    async function initHost() {
      if (!hostInitPromise) {
        hostInitPromise = (async () => {
          
          const remoteEntry = await import("virtual:mf-REMOTE_ENTRY_ID:__mfe_internal__vis2DemoWidget__customWidgets_js");
          const runtime = await remoteEntry.init();
          const usedShared = {
      "react": {
            shareConfig: {
              singleton: true,
              requiredVersion: "*",
              
            }
          },
"react-dom": {
            shareConfig: {
              singleton: true,
              requiredVersion: "*",
              
            }
          }
    };
          for (const [pkg, share] of Object.entries(usedShared)) {
            if (__mfModuleCache.share[pkg] !== undefined) {
              continue;
            }
            await runtime.loadShare(pkg, {
              customShareInfo: { shareConfig: share.shareConfig }
            }).then((factory) => {
              const mod = typeof factory === "function" ? factory() : factory;
              return Promise.resolve(mod).then((resolved) => {
                __mfModuleCache.share[pkg] = resolved;
              });
            });
          }
          const __mfRemotePreloads = [];
          await Promise.all(__mfRemotePreloads);
          return runtime;
        })();
      }
      return hostInitPromise;
    }
    hostInitPromise = initHost();
    export { initHost, hostInitPromise };
