// __mf__virtual/__mfe_internal__vis2DemoWidget__loadShare__react_mf_2_dom__loadShare__.mjs

    import * as __mfLocalShare from "__mf__virtual/__mfe_internal__vis2DemoWidget__prebuild__react_mf_2_dom__prebuild__.js";
    
    
const __mfCacheGlobalKey = "__mf_module_cache__";
globalThis[__mfCacheGlobalKey] ||= { share: {}, remote: {} };
globalThis[__mfCacheGlobalKey].share ||= {};
globalThis[__mfCacheGlobalKey].remote ||= {};
const __mfModuleCache = globalThis[__mfCacheGlobalKey];

    let exportModule = __mfModuleCache.share["react-dom"]
    if (exportModule === undefined) {
      exportModule = __mfLocalShare;
      __mfModuleCache.share["react-dom"] = exportModule;
    }
    export default exportModule.default ?? exportModule;
    const { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: __mf_0, createPortal: __mf_1, createRoot: __mf_2, findDOMNode: __mf_3, flushSync: __mf_4, hydrate: __mf_5, hydrateRoot: __mf_6, render: __mf_7, unmountComponentAtNode: __mf_8, unstable_batchedUpdates: __mf_9, unstable_renderSubtreeIntoContainer: __mf_10, version: __mf_11 } = exportModule;
    export { __mf_0 as __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, __mf_1 as createPortal, __mf_2 as createRoot, __mf_3 as findDOMNode, __mf_4 as flushSync, __mf_5 as hydrate, __mf_6 as hydrateRoot, __mf_7 as render, __mf_8 as unmountComponentAtNode, __mf_9 as unstable_batchedUpdates, __mf_10 as unstable_renderSubtreeIntoContainer, __mf_11 as version };
